<!DOCTYPE html>
<html class="no-js" lang="en"><head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://biodatageeks.org/polars-bio/blog/2026/02/20/interval-operations-benchmark--update-february-2026/" rel="canonical"/>
<link href="../../14/benchmarking-genomic-format-readers-in-python-with-polars/" rel="prev"/>
<link href="../../../../../assets/logo.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.7.1" name="generator"/>
<title>Interval operations benchmark â€” update February 2026 - polars-bio</title>
<link href="../../../../../assets/stylesheets/main.484c7ddc.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../../../assets/_markdown_exec_pyodide.css" rel="stylesheet"/>
<link href="../../../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-1M9PLLGBYS"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-1M9PLLGBYS",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-1M9PLLGBYS",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
<script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
<meta content="website" property="og:type"/>
<meta content="Interval operations benchmark â€” update February 2026 - polars-bio" property="og:title"/>
<meta content="https://biodatageeks.org/polars-bio/assets/images/social/blog/2026/02/20/interval-operations-benchmark--update-february-2026.png" property="og:image"/>
<meta content="image/png" property="og:image:type"/>
<meta content="1200" property="og:image:width"/>
<meta content="630" property="og:image:height"/>
<meta content="https://biodatageeks.org/polars-bio/blog/2026/02/20/interval-operations-benchmark--update-february-2026/" property="og:url"/>
<meta content="summary_large_image" property="twitter:card"/>
<meta content="Interval operations benchmark â€” update February 2026 - polars-bio" property="twitter:title"/>
<meta content="https://biodatageeks.org/polars-bio/assets/images/social/blog/2026/02/20/interval-operations-benchmark--update-february-2026.png" property="twitter:image"/>
<link href="../../../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><script src="../../../../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#interval-operations-benchmark-update-february-2026">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<div data-md-color-scheme="default" data-md-component="outdated" hidden="">
</div>
<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="polars-bio" class="md-header__button md-logo" data-md-component="logo" href="../../../../.." title="polars-bio">
<img alt="logo" src="../../../../../assets/logo.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            polars-bio
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              Interval operations benchmark â€” update February 2026
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/biodatageeks/polars-bio" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"></path></svg>
</div>
<div class="md-source__repository">
    biodatageeks/polars-bio
  </div>
</a>
</div>
</nav>
<nav aria-label="Tabs" class="md-tabs" data-md-component="tabs">
<div class="md-grid">
<ul class="md-tabs__list">
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../..">
          
  
  
  Home

        </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../quickstart/">
        
  
  
    
  
  ğŸƒğŸ¼â€â™‚ï¸ Quick start

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../features/">
        
  
  
    
  
  ğŸ”¨Features

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../notebooks/tutorial/">
        
  
  
    
  
  ğŸ“šTutorial

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../api/">
        
  
  
    
  
  âš™ï¸ API reference

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../faq/">
        
  
  
    
  
  â“ FAQ

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../supplement/">
        
  
  
    
  
  ğŸš€ Performance

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../developers/">
        
  
  
    
  
  ğŸ› ï¸ Developers guide

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../../../contact/">
        
  
  
    
  
  ğŸ“¡ Contact

      </a>
</li>
<li class="md-tabs__item md-tabs__item--active">
<a class="md-tabs__link" href="../../../../">
          
  
  
  ğŸ“– Blog

        </a>
</li>
</ul>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" hidden="">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary md-nav--lifted" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="polars-bio" class="md-nav__button md-logo" data-md-component="logo" href="../../../../.." title="polars-bio">
<img alt="logo" src="../../../../../assets/logo.png"/>
</a>
    polars-bio
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/biodatageeks/polars-bio" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"></path></svg>
</div>
<div class="md-source__repository">
    biodatageeks/polars-bio
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
<span class="md-ellipsis">
    
  
    Home
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_1_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_1">
<span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../..">
<span class="md-ellipsis">
    
  
    polars-bio
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../quickstart/">
<span class="md-ellipsis">
    
  
    ğŸƒğŸ¼â€â™‚ï¸ Quick start
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../features/">
<span class="md-ellipsis">
    
  
    ğŸ”¨Features
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../notebooks/tutorial/">
<span class="md-ellipsis">
    
  
    ğŸ“šTutorial
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../api/">
<span class="md-ellipsis">
    
  
    âš™ï¸ API reference
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../faq/">
<span class="md-ellipsis">
    
  
    â“ FAQ
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../supplement/">
<span class="md-ellipsis">
    
  
    ğŸš€ Performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../developers/">
<span class="md-ellipsis">
    
  
    ğŸ› ï¸ Developers guide
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../contact/">
<span class="md-ellipsis">
    
  
    ğŸ“¡ Contact
  

    
  </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../quickstart/">
<span class="md-ellipsis">
    
  
    ğŸƒğŸ¼â€â™‚ï¸ Quick start
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../features/">
<span class="md-ellipsis">
    
  
    ğŸ”¨Features
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../notebooks/tutorial/">
<span class="md-ellipsis">
    
  
    ğŸ“šTutorial
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../api/">
<span class="md-ellipsis">
    
  
    âš™ï¸ API reference
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../faq/">
<span class="md-ellipsis">
    
  
    â“ FAQ
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../supplement/">
<span class="md-ellipsis">
    
  
    ğŸš€ Performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../developers/">
<span class="md-ellipsis">
    
  
    ğŸ› ï¸ Developers guide
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../../contact/">
<span class="md-ellipsis">
    
  
    ğŸ“¡ Contact
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_10" type="checkbox"/>
<label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="">
<span class="md-ellipsis">
    
  
    ğŸ“– Blog
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="true" aria-labelledby="__nav_10_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_10">
<span class="md-nav__icon md-icon"></span>
            
  
    ğŸ“– Blog
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--active">
<a class="md-nav__link" href="../../../../">
<span class="md-ellipsis">
    
  
    Blog
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_10_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_10_2" id="__nav_10_2_label" tabindex="0">
<span class="md-ellipsis">
    
  
    Categories
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_10_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_10_2">
<span class="md-nav__icon md-icon"></span>
            
  
    Categories
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../category/benchmarks/">
<span class="md-ellipsis">
    
  
    benchmarks
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../category/file-formats/">
<span class="md-ellipsis">
    
  
    file formats
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../category/performance/">
<span class="md-ellipsis">
    
  
    performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../../category/releases/">
<span class="md-ellipsis">
    
  
    releases
  

    
  </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#introduction">
<span class="md-ellipsis">
      
        Introduction
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setup">
<span class="md-ellipsis">
      
        Setup
      
    </span>
</a>
<nav aria-label="Setup" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#software-versions">
<span class="md-ellipsis">
      
        Software versions
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#benchmark-test-cases">
<span class="md-ellipsis">
      
        Benchmark test cases
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#operations-and-tool-support">
<span class="md-ellipsis">
      
        Operations and tool support
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#results">
<span class="md-ellipsis">
      
        Results
      
    </span>
</a>
<nav aria-label="Results" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#speedup-comparison-across-all-operations">
<span class="md-ellipsis">
      
        Speedup comparison across all operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#thread-scalability-dataset-8-7">
<span class="md-ellipsis">
      
        Thread scalability (dataset 8-7)
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#summary">
<span class="md-ellipsis">
      
        Summary
      
    </span>
</a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content md-content--post" data-md-component="content">
<div class="md-sidebar md-sidebar--post" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner md-post">
<nav class="md-nav md-nav--primary">
<div class="md-post__back">
<div class="md-nav__title md-nav__container">
<a class="md-nav__link" href="../../../../">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
<span class="md-ellipsis">
                    Back to index
                  </span>
</a>
</div>
</div>
<ul class="md-post__meta md-nav__list">
<li class="md-nav__item md-nav__item--section">
<div class="md-post__title">
<span class="md-ellipsis">
                    Metadata
                  </span>
</div>
<nav class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<div class="md-nav__link">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 19H5V8h14m-3-7v2H8V1H6v2H5c-1.11 0-2 .89-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2h-1V1m-1 11h-5v5h5z"></path></svg>
<time class="md-ellipsis" datetime="2026-02-20 00:00:00+00:00">Friday, February 20, 2026</time>
</div>
</li>
<li class="md-nav__item">
<div class="md-nav__link">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9 3v15h3V3zm3 2 4 13 3-1-4-13zM5 5v13h3V5zM3 19v2h18v-2z"></path></svg>
<span class="md-ellipsis">
                            in
                            
                              <a href="../../../../category/performance/">performance</a>, 
                              <a href="../../../../category/benchmarks/">benchmarks</a></span>
</div>
</li>
<li class="md-nav__item">
<div class="md-nav__link">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M12 20a8 8 0 0 0 8-8 8 8 0 0 0-8-8 8 8 0 0 0-8 8 8 8 0 0 0 8 8m0-18a10 10 0 0 1 10 10 10 10 0 0 1-10 10C6.47 22 2 17.5 2 12A10 10 0 0 1 12 2m.5 5v5.25l4.5 2.67-.75 1.23L11 13V7z"></path></svg>
<span class="md-ellipsis">
                            
                              4 min read
                            
                          </span>
</div>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<article class="md-content__inner md-typeset">
<a class="md-content__button md-icon" href="https://github.com/biodatageeks/polars-bio/edit/master/docs/blog/posts/benchmark-operations-2026-02.md" rel="edit" title="Edit this page">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"></path></svg>
</a>
<h1 id="interval-operations-benchmark-update-february-2026">Interval operations benchmark â€” update February 2026<a class="headerlink" href="#interval-operations-benchmark-update-february-2026" title="Permanent link"></a></h1>
<h2 id="introduction">Introduction<a class="headerlink" href="#introduction" title="Permanent link"></a></h2>
<p>Back in <a href="../../../../2025/09/05/interval-operations-benchmark--update-september-2025/">September 2025</a> we benchmarked three libraries across three operations. A lot has changed since then. In December 2025, pyranges1 published a <a href="https://www.biorxiv.org/content/10.64898/2025.12.11.693639v1">preprint</a> describing its Rust-powered backend (<a href="https://github.com/pyranges/ruranges">ruranges</a>) and an expanded set of interval operations. On the <a href="https://github.com/biodatageeks/polars-bio">polars-bio</a> side, version 0.24.0 ships a fully rewritten range-operations engine built on upstream DataFusion UDTF providers (OverlapProvider, NearestProvider, and the new coverage/cluster/complement/merge/subtract providers from <a href="https://github.com/biodatageeks/datafusion-bio-functions">datafusion-bio-function-ranges</a>), replacing the earlier sequila-native backend.</p>
<!-- more -->
<p>This rewrite also expanded the operation set from three to eight. In addition to <strong>overlap</strong>, <strong>nearest</strong>, and <strong>count_overlaps</strong>, polars-bio 0.24.0 supports <strong>coverage</strong>, <strong>cluster</strong>, <strong>complement</strong>, <strong>merge</strong>, and <strong>subtract</strong> â€” covering all the everyday interval manipulation tasks that genomics workflows depend on.</p>
<p>We also added a fourth contender: <a href="https://github.com/open2c/bioframe">Bioframe</a>, a pandas-based genomic interval library widely used in the 3D genomics community. This gives us a broader view of the Python genomic interval landscape.</p>
<p>For comparability with our previous benchmarks, we continue to use the same <a href="/polars-bio/supplement/#real-dataset">AIList</a> dataset. All benchmark code and raw results are available in the <a href="https://github.com/biodatageeks/polars-bio-bench">polars-bio-bench</a> repository.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link"></a></h2>
<h3 id="software-versions">Software versions<a class="headerlink" href="#software-versions" title="Permanent link"></a></h3>
<table>
<thead>
<tr>
<th>Library</th>
<th>Version</th>
</tr>
</thead>
<tbody>
<tr>
<td>polars-bio</td>
<td>0.24.0</td>
</tr>
<tr>
<td>pyranges1</td>
<td>1.2.0</td>
</tr>
<tr>
<td>GenomicRanges</td>
<td>0.8.4</td>
</tr>
<tr>
<td>bioframe</td>
<td>0.8.0</td>
</tr>
</tbody>
</table>
<h3 id="benchmark-test-cases">Benchmark test cases<a class="headerlink" href="#benchmark-test-cases" title="Permanent link"></a></h3>
<p><strong>Binary operations</strong> (overlap, nearest, count_overlaps, coverage):</p>
<table>
<thead>
<tr>
<th>Dataset pairs</th>
<th>Size</th>
<th># of overlaps (1-based)</th>
</tr>
</thead>
<tbody>
<tr>
<td>1-2 &amp; 2-1</td>
<td>Small</td>
<td>54,246</td>
</tr>
<tr>
<td>3-7 &amp; 7-3</td>
<td>Medium</td>
<td>4,408,383</td>
</tr>
<tr>
<td>7-8 &amp; 8-7</td>
<td>Large</td>
<td>307,184,634</td>
</tr>
</tbody>
</table>
<p><strong>Unary operations</strong> (cluster, complement, merge, subtract):</p>
<table>
<thead>
<tr>
<th>Dataset</th>
<th>Size</th>
<th>Name (intervals)</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>Small</td>
<td>fBrain (199K)</td>
</tr>
<tr>
<td>2</td>
<td>Small</td>
<td>exons (439K)</td>
</tr>
<tr>
<td>7</td>
<td>Medium</td>
<td>ex-anno (1,194K)</td>
</tr>
<tr>
<td>3</td>
<td>Medium</td>
<td>chainOrnAna1 (1,957K)</td>
</tr>
<tr>
<td>8</td>
<td>Large</td>
<td>ex-rna (9,945K)</td>
</tr>
<tr>
<td>5</td>
<td>Large</td>
<td>chainXenTro3Link (50,981K)</td>
</tr>
</tbody>
</table>
<h3 id="operations-and-tool-support">Operations and tool support<a class="headerlink" href="#operations-and-tool-support" title="Permanent link"></a></h3>
<table>
<thead>
<tr>
<th>Operation</th>
<th>polars-bio</th>
<th>PyRanges1</th>
<th>GenomicRanges</th>
<th>Bioframe</th>
</tr>
</thead>
<tbody>
<tr>
<td>overlap</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>nearest</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>count_overlaps</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>coverage</td>
<td>yes</td>
<td>--</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>cluster</td>
<td>yes</td>
<td>yes</td>
<td>--</td>
<td>yes</td>
</tr>
<tr>
<td>complement</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>merge</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
<td>yes</td>
</tr>
<tr>
<td>subtract</td>
<td>yes</td>
<td>yes</td>
<td>--</td>
<td>yes</td>
</tr>
</tbody>
</table>
<h2 id="results">Results<a class="headerlink" href="#results" title="Permanent link"></a></h2>
<h3 id="speedup-comparison-across-all-operations">Speedup comparison across all operations<a class="headerlink" href="#speedup-comparison-across-all-operations" title="Permanent link"></a></h3>
<p><a class="glightbox" data-desc-position="bottom" data-height="auto" data-type="image" data-width="auto" href="../../../../figures/benchmark-operations-2026-02/all_operations_speedup_comparison.png"><img alt="all_operations_speedup_comparison.png" src="../../../../figures/benchmark-operations-2026-02/all_operations_speedup_comparison.png"/></a></p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<ol>
<li>Missing bars indicate that the operation is not supported by the library for the given dataset.</li>
<li>Crash bars indicate that the library failed to complete.</li>
</ol>
</div>
<p>Key takeaways:</p>
<ul>
<li><strong>polars-bio</strong> is the fastest library in 7 out of 8 operations on the large dataset (8-7). The sole exception is <strong>nearest</strong>, where <strong>GenomicRanges</strong> holds a 1.63x advantage.</li>
<li>On small datasets (1-2, 2-1), <strong>GenomicRanges</strong> leads in overlap (1.74x) and count_overlaps, reflecting lower per-call overhead for small inputs.</li>
<li><strong>Bioframe</strong> is consistently the slowest library, falling 5-50x behind polars-bio depending on the operation and dataset size.</li>
<li>For the new operations (coverage, cluster, complement, merge, subtract), <strong>polars-bio</strong> leads across the board with <strong>PyRanges1</strong> a respectable second (0.23-0.61x relative to polars-bio on the large dataset).</li>
</ul>
<p><strong>Summary â€” dataset 8-7 speedup relative to polars-bio (higher is better for polars-bio):</strong></p>
<table>
<thead>
<tr>
<th>Operation</th>
<th>polars-bio</th>
<th>PyRanges1</th>
<th>GenomicRanges</th>
<th>Bioframe</th>
</tr>
</thead>
<tbody>
<tr>
<td>overlap</td>
<td>1.00x</td>
<td>0.18x</td>
<td>0.73x</td>
<td>0.13x</td>
</tr>
<tr>
<td>nearest</td>
<td>1.00x</td>
<td>0.06x</td>
<td>1.63x</td>
<td>0.04x</td>
</tr>
<tr>
<td>count_overlaps</td>
<td>1.00x</td>
<td>0.24x</td>
<td>0.19x</td>
<td>0.02x</td>
</tr>
<tr>
<td>coverage</td>
<td>1.00x</td>
<td>--</td>
<td>0.36x</td>
<td>0.05x</td>
</tr>
<tr>
<td>cluster</td>
<td>1.00x</td>
<td>0.61x</td>
<td>--</td>
<td>0.20x</td>
</tr>
<tr>
<td>complement</td>
<td>1.00x</td>
<td>0.53x</td>
<td>0.02x</td>
<td>0.06x</td>
</tr>
<tr>
<td>merge</td>
<td>1.00x</td>
<td>0.51x</td>
<td>0.02x</td>
<td>0.12x</td>
</tr>
<tr>
<td>subtract</td>
<td>1.00x</td>
<td>0.23x</td>
<td>--</td>
<td>0.05x</td>
</tr>
</tbody>
</table>
<h3 id="thread-scalability-dataset-8-7">Thread scalability (dataset 8-7)<a class="headerlink" href="#thread-scalability-dataset-8-7" title="Permanent link"></a></h3>
<p><a class="glightbox" data-desc-position="bottom" data-height="auto" data-type="image" data-width="auto" href="../../../../figures/benchmark-operations-2026-02/polars_bio_scalability_8_7.png"><img alt="polars_bio_scalability_8_7.png" src="../../../../figures/benchmark-operations-2026-02/polars_bio_scalability_8_7.png"/></a>
One of polars-bio's key advantages is transparent multithreaded execution. The table below shows wall-clock times for all eight operations on the large dataset (8-7) as thread count increases:</p>
<table>
<thead>
<tr>
<th>Operation</th>
<th>1 thread</th>
<th>8 threads</th>
<th>Speedup</th>
</tr>
</thead>
<tbody>
<tr>
<td>overlap</td>
<td>4.03s</td>
<td>0.74s</td>
<td>5.43x</td>
</tr>
<tr>
<td>nearest</td>
<td>2.55s</td>
<td>0.44s</td>
<td>5.77x</td>
</tr>
<tr>
<td>count_overlaps</td>
<td>1.55s</td>
<td>0.27s</td>
<td>5.75x</td>
</tr>
<tr>
<td>coverage</td>
<td>1.20s</td>
<td>0.22s</td>
<td>5.48x</td>
</tr>
<tr>
<td>subtract</td>
<td>1.15s</td>
<td>0.17s</td>
<td>6.59x</td>
</tr>
<tr>
<td>merge</td>
<td>0.29s</td>
<td>0.05s</td>
<td>5.34x</td>
</tr>
<tr>
<td>complement</td>
<td>0.29s</td>
<td>0.06s</td>
<td>4.99x</td>
</tr>
<tr>
<td>cluster</td>
<td>0.48s</td>
<td>0.15s</td>
<td>3.27x</td>
</tr>
</tbody>
</table>
<p>Key takeaways:</p>
<ul>
<li>Most operations achieve near-linear scaling, with 5-6x speedup at 8 threads.</li>
<li><strong>subtract</strong> scales best at 6.59x, while <strong>cluster</strong> shows more modest scaling (3.27x) â€” expected, as clustering is inherently sequential within each chromosome.</li>
<li>At 8 threads, <strong>overlap</strong> on 307M result pairs completes in just 0.74 seconds.</li>
</ul>
<h2 id="summary">Summary<a class="headerlink" href="#summary" title="Permanent link"></a></h2>
<ul>
<li><strong>polars-bio</strong> is the fastest single-threaded library for 7 out of 8 operations at scale, with speedups ranging from 1.5x to 25x over alternatives.</li>
<li><strong>GenomicRanges</strong> wins the nearest operation and leads on small datasets, making it a solid choice when input sizes are modest.</li>
<li>polars-bio delivers excellent thread scaling (5-6x at 8 threads), turning already-fast single-threaded times into sub-second performance on large datasets.</li>
<li><strong>PyRanges1</strong>, despite its recent <a href="https://www.biorxiv.org/content/10.64898/2025.12.11.693639v1">preprint</a> claiming "ultrafast" performance, is slower than polars-bio in every operation tested on large datasets (0.06-0.61x). While it is a solid improvement over its predecessor, the "ultrafast" characterization does not hold when compared against polars-bio's <a href="https://datafusion.apache.org/">DataFusion</a>-based engine.</li>
<li><strong>Bioframe</strong> is consistently the slowest across all operations and dataset sizes.</li>
<li>The expanded eight-operation benchmark confirms that polars-bio's advantage extends beyond overlap and nearest to all standard range operation categories â€” coverage, cluster, complement, merge, and subtract.</li>
</ul>
<p>All benchmark code, raw results, and additional figures are available at <a href="https://github.com/biodatageeks/polars-bio-bench">polars-bio-bench</a>.</p>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
<button class="md-top md-icon" data-md-component="top" hidden="" type="button">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
</main>
<footer class="md-footer">
<nav aria-label="Footer" class="md-footer__inner md-grid">
<a aria-label="Previous: Benchmarking Genomic Format Readers in Python with Polars" class="md-footer__link md-footer__link--prev" href="../../14/benchmarking-genomic-format-readers-in-python-with-polars/">
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</div>
<div class="md-footer__title">
<span class="md-footer__direction">
                Previous
              </span>
<div class="md-ellipsis">
                Benchmarking Genomic Format Readers in Python with Polars
              </div>
</div>
</a>
</nav>
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
<div class="md-social">
<a class="md-social__link" href="https://discord.gg/bpxQ4Yxhk5" rel="noopener" target="_blank" title="Discord">
<svg viewbox="0 0 576 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M492.5 69.8c-.2-.3-.4-.6-.8-.7-38.1-17.5-78.4-30-119.7-37.1-.4-.1-.8 0-1.1.1s-.6.4-.8.8c-5.5 9.9-10.5 20.2-14.9 30.6-44.6-6.8-89.9-6.8-134.4 0-4.5-10.5-9.5-20.7-15.1-30.6-.2-.3-.5-.6-.8-.8s-.7-.2-1.1-.2C162.5 39 122.2 51.5 84.1 69c-.3.1-.6.4-.8.7C7.1 183.5-13.8 294.6-3.6 404.2c0 .3.1.5.2.8s.3.4.5.6c44.4 32.9 94 58 146.8 74.2.4.1.8.1 1.1 0s.7-.4.9-.7c11.3-15.4 21.4-31.8 30-48.8.1-.2.2-.5.2-.8s0-.5-.1-.8-.2-.5-.4-.6-.4-.3-.7-.4c-15.8-6.1-31.2-13.4-45.9-21.9-.3-.2-.5-.4-.7-.6s-.3-.6-.3-.9 0-.6.2-.9.3-.5.6-.7c3.1-2.3 6.2-4.7 9.1-7.1.3-.2.6-.4.9-.4s.7 0 1 .1c96.2 43.9 200.4 43.9 295.5 0 .3-.1.7-.2 1-.2s.7.2.9.4c2.9 2.4 6 4.9 9.1 7.2.2.2.4.4.6.7s.2.6.2.9-.1.6-.3.9-.4.5-.6.6c-14.7 8.6-30 15.9-45.9 21.8-.2.1-.5.2-.7.4s-.3.4-.4.7-.1.5-.1.8.1.5.2.8c8.8 17 18.8 33.3 30 48.8.2.3.6.6.9.7s.8.1 1.1 0c52.9-16.2 102.6-41.3 147.1-74.2.2-.2.4-.4.5-.6s.2-.5.2-.8c12.3-126.8-20.5-236.9-86.9-334.5zm-302 267.7c-29 0-52.8-26.6-52.8-59.2s23.4-59.2 52.8-59.2c29.7 0 53.3 26.8 52.8 59.2 0 32.7-23.4 59.2-52.8 59.2m195.4 0c-29 0-52.8-26.6-52.8-59.2s23.4-59.2 52.8-59.2c29.7 0 53.3 26.8 52.8 59.2 0 32.7-23.2 59.2-52.8 59.2"></path></svg>
</a>
<a class="md-social__link" href="mailto:marek.wiewiorka@pw.edu.pl" rel="noopener" target="_blank" title="Email">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m20 8-8 5-8-5V6l8 5 8-5m0-2H4c-1.11 0-2 .89-2 2v12a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2"></path></svg>
</a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"annotate": null, "base": "../../../../..", "features": ["content.code.copy", "content.code.select", "content.code.annotate", "content.tabs.link", "content.action.edit", "content.tooltips", "toc.follow", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "navigation.footer", "navigation.tracking", "navigation.instant", "content.footnote.tooltips"], "search": "../../../../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"default": ["latest"], "provider": "mike"}}</script>
<script src="../../../../../assets/javascripts/bundle.79ae519e.min.js"></script>
<script src="../../../../../assets/_markdown_exec_pyodide.js"></script>
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>