<!DOCTYPE html>
<html class="no-js" lang="en"><head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://biodatageeks.org/polars-bio/features/" rel="canonical"/>
<link href="../quickstart/" rel="prev"/>
<link href="../notebooks/tutorial/" rel="next"/>
<link href="../assets/logo.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.7.1" name="generator"/>
<title>üî®Features - polars-bio</title>
<link href="../assets/stylesheets/main.484c7ddc.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../assets/_markdown_exec_pyodide.css" rel="stylesheet"/>
<link href="../assets/_mkdocstrings.css" rel="stylesheet"/>
<script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-1M9PLLGBYS"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-1M9PLLGBYS",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-1M9PLLGBYS",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
<script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
<meta content="website" property="og:type"/>
<meta content="üî®Features - polars-bio" property="og:title"/>
<meta content="https://biodatageeks.org/polars-bio/assets/images/social/features.png" property="og:image"/>
<meta content="image/png" property="og:image:type"/>
<meta content="1200" property="og:image:width"/>
<meta content="630" property="og:image:height"/>
<meta content="https://biodatageeks.org/polars-bio/features/" property="og:url"/>
<meta content="summary_large_image" property="twitter:card"/>
<meta content="üî®Features - polars-bio" property="twitter:title"/>
<meta content="https://biodatageeks.org/polars-bio/assets/images/social/features.png" property="twitter:image"/>
<link href="../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><script src="../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#genomic-ranges-operations">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<div data-md-color-scheme="default" data-md-component="outdated" hidden="">
</div>
<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="polars-bio" class="md-header__button md-logo" data-md-component="logo" href=".." title="polars-bio">
<img alt="logo" src="../assets/logo.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            polars-bio
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              üî®Features
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/biodatageeks/polars-bio" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"></path></svg>
</div>
<div class="md-source__repository">
    biodatageeks/polars-bio
  </div>
</a>
</div>
</nav>
<nav aria-label="Tabs" class="md-tabs" data-md-component="tabs">
<div class="md-grid">
<ul class="md-tabs__list">
<li class="md-tabs__item md-tabs__item--active">
<a class="md-tabs__link" href="..">
          
  
  
  Home

        </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../quickstart/">
        
  
  
    
  
  üèÉüèº‚Äç‚ôÇÔ∏è Quick start

      </a>
</li>
<li class="md-tabs__item md-tabs__item--active">
<a class="md-tabs__link" href="./">
        
  
  
    
  
  üî®Features

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../notebooks/tutorial/">
        
  
  
    
  
  üìöTutorial

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../api/">
        
  
  
    
  
  ‚öôÔ∏è API reference

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../faq/">
        
  
  
    
  
  ‚ùì FAQ

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../supplement/">
        
  
  
    
  
  üöÄ Performance

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../developers/">
        
  
  
    
  
  üõ†Ô∏è Developers guide

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../contact/">
        
  
  
    
  
  üì° Contact

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../blog/">
          
  
  
  üìñ Blog

        </a>
</li>
</ul>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary md-nav--lifted" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="polars-bio" class="md-nav__button md-logo" data-md-component="logo" href=".." title="polars-bio">
<img alt="logo" src="../assets/logo.png"/>
</a>
    polars-bio
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/biodatageeks/polars-bio" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"></path></svg>
</div>
<div class="md-source__repository">
    biodatageeks/polars-bio
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="">
<span class="md-ellipsis">
    
  
    Home
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="true" aria-labelledby="__nav_1_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_1">
<span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="..">
<span class="md-ellipsis">
    
  
    polars-bio
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../quickstart/">
<span class="md-ellipsis">
    
  
    üèÉüèº‚Äç‚ôÇÔ∏è Quick start
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--active">
<input class="md-nav__toggle md-toggle" id="__toc" type="checkbox"/>
<label class="md-nav__link md-nav__link--active" for="__toc">
<span class="md-ellipsis">
    
  
    üî®Features
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<a class="md-nav__link md-nav__link--active" href="./">
<span class="md-ellipsis">
    
  
    üî®Features
  

    
  </span>
</a>
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#genomic-ranges-operations">
<span class="md-ellipsis">
      
        Genomic ranges operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#pileup-operations">
<span class="md-ellipsis">
      
        Pileup operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#coordinate-systems-support">
<span class="md-ellipsis">
      
        Coordinate systems support
      
    </span>
</a>
<nav aria-label="Coordinate systems support" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#how-it-works">
<span class="md-ellipsis">
      
        How it works
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#session-parameters">
<span class="md-ellipsis">
      
        Session parameters
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#reading-files-with-coordinate-system-metadata">
<span class="md-ellipsis">
      
        Reading files with coordinate system metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setting-metadata-on-dataframes">
<span class="md-ellipsis">
      
        Setting metadata on DataFrames
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#error-handling">
<span class="md-ellipsis">
      
        Error handling
      
    </span>
</a>
<nav aria-label="Error handling" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#missingcoordinatesystemerror">
<span class="md-ellipsis">
      
        MissingCoordinateSystemError
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#coordinatesystemmismatcherror">
<span class="md-ellipsis">
      
        CoordinateSystemMismatchError
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#default-behavior-lenient-validation">
<span class="md-ellipsis">
      
        Default behavior (lenient validation)
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#strict-mode">
<span class="md-ellipsis">
      
        Strict mode
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#migration-from-previous-versions">
<span class="md-ellipsis">
      
        Migration from previous versions
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-metadata">
<span class="md-ellipsis">
      
        File Metadata
      
    </span>
</a>
<nav aria-label="File Metadata" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#vcf-changes-since-0240">
<span class="md-ellipsis">
      
        VCF changes since 0.24.0
      
    </span>
</a>
<nav aria-label="VCF changes since 0.24.0" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#migration-examples">
<span class="md-ellipsis">
      
        Migration examples
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#metadata-structure">
<span class="md-ellipsis">
      
        Metadata Structure
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#accessing-metadata">
<span class="md-ellipsis">
      
        Accessing Metadata
      
    </span>
</a>
<nav aria-label="Accessing Metadata" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#1-get-all-metadata-as-a-dictionary">
<span class="md-ellipsis">
      
        1. Get all metadata as a dictionary
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#2-print-metadata-as-formatted-json">
<span class="md-ellipsis">
      
        2. Print metadata as formatted JSON
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#3-print-human-readable-summary">
<span class="md-ellipsis">
      
        3. Print human-readable summary
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#format-specific-metadata">
<span class="md-ellipsis">
      
        Format-Specific Metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setting-custom-metadata">
<span class="md-ellipsis">
      
        Setting Custom Metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#metadata-preservation">
<span class="md-ellipsis">
      
        Metadata Preservation
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#using-metadata-for-debugging">
<span class="md-ellipsis">
      
        Using Metadata for Debugging
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#api-reference">
<span class="md-ellipsis">
      
        API Reference
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-formats-support">
<span class="md-ellipsis">
      
        File formats support
      
    </span>
</a>
<nav aria-label="File formats support" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#indexed-reads-predicate-pushdown">
<span class="md-ellipsis">
      
        Indexed reads &amp; predicate pushdown
      
    </span>
</a>
<nav aria-label="Indexed reads &amp; predicate pushdown" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#supported-index-formats">
<span class="md-ellipsis">
      
        Supported index formats
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usage-with-the-scanread-api">
<span class="md-ellipsis">
      
        Usage with the scan/read API
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usage-with-the-sql-api">
<span class="md-ellipsis">
      
        Usage with the SQL API
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#automatic-parallel-partitioning">
<span class="md-ellipsis">
      
        Automatic parallel partitioning
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#record-level-filter-pushdown">
<span class="md-ellipsis">
      
        Record-level filter pushdown
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#projection-pushdown">
<span class="md-ellipsis">
      
        Projection pushdown
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#index-file-generation">
<span class="md-ellipsis">
      
        Index file generation
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-output">
<span class="md-ellipsis">
      
        File Output
      
    </span>
</a>
<nav aria-label="File Output" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#output-format-support">
<span class="md-ellipsis">
      
        Output Format Support
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#basic-usage">
<span class="md-ellipsis">
      
        Basic Usage
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#sorted-output-with-sort_on_write">
<span class="md-ellipsis">
      
        Sorted Output with sort_on_write
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#cram-output">
<span class="md-ellipsis">
      
        CRAM Output
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output-compression">
<span class="md-ellipsis">
      
        Output Compression
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#header-preservation">
<span class="md-ellipsis">
      
        Header Preservation
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#polars-extension-methods">
<span class="md-ellipsis">
      
        Polars Extension Methods
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#sql-powered-data-processing">
<span class="md-ellipsis">
      
        SQL-powered data processing
      
    </span>
</a>
<nav aria-label="SQL-powered data processing" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#accessing-registered-tables">
<span class="md-ellipsis">
      
        Accessing registered tables
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-inspection">
<span class="md-ellipsis">
      
        Schema Inspection
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#bam-optional-tags">
<span class="md-ellipsis">
      
        BAM Optional Tags
      
    </span>
</a>
<nav aria-label="BAM Optional Tags" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#usage">
<span class="md-ellipsis">
      
        Usage
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#common-tags">
<span class="md-ellipsis">
      
        Common Tags
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#performance">
<span class="md-ellipsis">
      
        Performance
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#parallel-engine">
<span class="md-ellipsis">
      
        Parallel engine üèéÔ∏è
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#cloud-storage">
<span class="md-ellipsis">
      
        Cloud storage ‚òÅÔ∏è
      
    </span>
</a>
<nav aria-label="Cloud storage ‚òÅÔ∏è" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#example">
<span class="md-ellipsis">
      
        Example
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#supported-features">
<span class="md-ellipsis">
      
        Supported features
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#aws-s3-configuration">
<span class="md-ellipsis">
      
        AWS S3 configuration
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#google-cloud-storage-configuration">
<span class="md-ellipsis">
      
        Google Cloud Storage configuration
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#azure-blob-storage-configuration">
<span class="md-ellipsis">
      
        Azure Blob Storage configuration
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#streaming">
<span class="md-ellipsis">
      
        Streaming üöÇ
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#compression">
<span class="md-ellipsis">
      
        Compression
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#dataframes-support">
<span class="md-ellipsis">
      
        DataFrames support
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#polars-integration">
<span class="md-ellipsis">
      
        Polars Integration
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../notebooks/tutorial/">
<span class="md-ellipsis">
    
  
    üìöTutorial
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../api/">
<span class="md-ellipsis">
    
  
    ‚öôÔ∏è API reference
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../faq/">
<span class="md-ellipsis">
    
  
    ‚ùì FAQ
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../supplement/">
<span class="md-ellipsis">
    
  
    üöÄ Performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../developers/">
<span class="md-ellipsis">
    
  
    üõ†Ô∏è Developers guide
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../contact/">
<span class="md-ellipsis">
    
  
    üì° Contact
  

    
  </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../quickstart/">
<span class="md-ellipsis">
    
  
    üèÉüèº‚Äç‚ôÇÔ∏è Quick start
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--active">
<input class="md-nav__toggle md-toggle" id="__toc" type="checkbox"/>
<label class="md-nav__link md-nav__link--active" for="__toc">
<span class="md-ellipsis">
    
  
    üî®Features
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<a class="md-nav__link md-nav__link--active" href="./">
<span class="md-ellipsis">
    
  
    üî®Features
  

    
  </span>
</a>
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#genomic-ranges-operations">
<span class="md-ellipsis">
      
        Genomic ranges operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#pileup-operations">
<span class="md-ellipsis">
      
        Pileup operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#coordinate-systems-support">
<span class="md-ellipsis">
      
        Coordinate systems support
      
    </span>
</a>
<nav aria-label="Coordinate systems support" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#how-it-works">
<span class="md-ellipsis">
      
        How it works
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#session-parameters">
<span class="md-ellipsis">
      
        Session parameters
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#reading-files-with-coordinate-system-metadata">
<span class="md-ellipsis">
      
        Reading files with coordinate system metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setting-metadata-on-dataframes">
<span class="md-ellipsis">
      
        Setting metadata on DataFrames
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#error-handling">
<span class="md-ellipsis">
      
        Error handling
      
    </span>
</a>
<nav aria-label="Error handling" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#missingcoordinatesystemerror">
<span class="md-ellipsis">
      
        MissingCoordinateSystemError
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#coordinatesystemmismatcherror">
<span class="md-ellipsis">
      
        CoordinateSystemMismatchError
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#default-behavior-lenient-validation">
<span class="md-ellipsis">
      
        Default behavior (lenient validation)
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#strict-mode">
<span class="md-ellipsis">
      
        Strict mode
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#migration-from-previous-versions">
<span class="md-ellipsis">
      
        Migration from previous versions
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-metadata">
<span class="md-ellipsis">
      
        File Metadata
      
    </span>
</a>
<nav aria-label="File Metadata" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#vcf-changes-since-0240">
<span class="md-ellipsis">
      
        VCF changes since 0.24.0
      
    </span>
</a>
<nav aria-label="VCF changes since 0.24.0" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#migration-examples">
<span class="md-ellipsis">
      
        Migration examples
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#metadata-structure">
<span class="md-ellipsis">
      
        Metadata Structure
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#accessing-metadata">
<span class="md-ellipsis">
      
        Accessing Metadata
      
    </span>
</a>
<nav aria-label="Accessing Metadata" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#1-get-all-metadata-as-a-dictionary">
<span class="md-ellipsis">
      
        1. Get all metadata as a dictionary
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#2-print-metadata-as-formatted-json">
<span class="md-ellipsis">
      
        2. Print metadata as formatted JSON
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#3-print-human-readable-summary">
<span class="md-ellipsis">
      
        3. Print human-readable summary
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#format-specific-metadata">
<span class="md-ellipsis">
      
        Format-Specific Metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setting-custom-metadata">
<span class="md-ellipsis">
      
        Setting Custom Metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#metadata-preservation">
<span class="md-ellipsis">
      
        Metadata Preservation
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#using-metadata-for-debugging">
<span class="md-ellipsis">
      
        Using Metadata for Debugging
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#api-reference">
<span class="md-ellipsis">
      
        API Reference
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-formats-support">
<span class="md-ellipsis">
      
        File formats support
      
    </span>
</a>
<nav aria-label="File formats support" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#indexed-reads-predicate-pushdown">
<span class="md-ellipsis">
      
        Indexed reads &amp; predicate pushdown
      
    </span>
</a>
<nav aria-label="Indexed reads &amp; predicate pushdown" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#supported-index-formats">
<span class="md-ellipsis">
      
        Supported index formats
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usage-with-the-scanread-api">
<span class="md-ellipsis">
      
        Usage with the scan/read API
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usage-with-the-sql-api">
<span class="md-ellipsis">
      
        Usage with the SQL API
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#automatic-parallel-partitioning">
<span class="md-ellipsis">
      
        Automatic parallel partitioning
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#record-level-filter-pushdown">
<span class="md-ellipsis">
      
        Record-level filter pushdown
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#projection-pushdown">
<span class="md-ellipsis">
      
        Projection pushdown
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#index-file-generation">
<span class="md-ellipsis">
      
        Index file generation
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-output">
<span class="md-ellipsis">
      
        File Output
      
    </span>
</a>
<nav aria-label="File Output" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#output-format-support">
<span class="md-ellipsis">
      
        Output Format Support
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#basic-usage">
<span class="md-ellipsis">
      
        Basic Usage
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#sorted-output-with-sort_on_write">
<span class="md-ellipsis">
      
        Sorted Output with sort_on_write
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#cram-output">
<span class="md-ellipsis">
      
        CRAM Output
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output-compression">
<span class="md-ellipsis">
      
        Output Compression
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#header-preservation">
<span class="md-ellipsis">
      
        Header Preservation
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#polars-extension-methods">
<span class="md-ellipsis">
      
        Polars Extension Methods
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#sql-powered-data-processing">
<span class="md-ellipsis">
      
        SQL-powered data processing
      
    </span>
</a>
<nav aria-label="SQL-powered data processing" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#accessing-registered-tables">
<span class="md-ellipsis">
      
        Accessing registered tables
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-inspection">
<span class="md-ellipsis">
      
        Schema Inspection
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#bam-optional-tags">
<span class="md-ellipsis">
      
        BAM Optional Tags
      
    </span>
</a>
<nav aria-label="BAM Optional Tags" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#usage">
<span class="md-ellipsis">
      
        Usage
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#common-tags">
<span class="md-ellipsis">
      
        Common Tags
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#performance">
<span class="md-ellipsis">
      
        Performance
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#parallel-engine">
<span class="md-ellipsis">
      
        Parallel engine üèéÔ∏è
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#cloud-storage">
<span class="md-ellipsis">
      
        Cloud storage ‚òÅÔ∏è
      
    </span>
</a>
<nav aria-label="Cloud storage ‚òÅÔ∏è" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#example">
<span class="md-ellipsis">
      
        Example
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#supported-features">
<span class="md-ellipsis">
      
        Supported features
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#aws-s3-configuration">
<span class="md-ellipsis">
      
        AWS S3 configuration
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#google-cloud-storage-configuration">
<span class="md-ellipsis">
      
        Google Cloud Storage configuration
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#azure-blob-storage-configuration">
<span class="md-ellipsis">
      
        Azure Blob Storage configuration
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#streaming">
<span class="md-ellipsis">
      
        Streaming üöÇ
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#compression">
<span class="md-ellipsis">
      
        Compression
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#dataframes-support">
<span class="md-ellipsis">
      
        DataFrames support
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#polars-integration">
<span class="md-ellipsis">
      
        Polars Integration
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../notebooks/tutorial/">
<span class="md-ellipsis">
    
  
    üìöTutorial
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../api/">
<span class="md-ellipsis">
    
  
    ‚öôÔ∏è API reference
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../faq/">
<span class="md-ellipsis">
    
  
    ‚ùì FAQ
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../supplement/">
<span class="md-ellipsis">
    
  
    üöÄ Performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../developers/">
<span class="md-ellipsis">
    
  
    üõ†Ô∏è Developers guide
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../contact/">
<span class="md-ellipsis">
    
  
    üì° Contact
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_10" type="checkbox"/>
<label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="0">
<span class="md-ellipsis">
    
  
    üìñ Blog
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_10_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_10">
<span class="md-nav__icon md-icon"></span>
            
  
    üìñ Blog
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../blog/">
<span class="md-ellipsis">
    
  
    Blog
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_10_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_10_2" id="__nav_10_2_label" tabindex="0">
<span class="md-ellipsis">
    
  
    Categories
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_10_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_10_2">
<span class="md-nav__icon md-icon"></span>
            
  
    Categories
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../blog/category/benchmarks/">
<span class="md-ellipsis">
    
  
    benchmarks
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../blog/category/file-formats/">
<span class="md-ellipsis">
    
  
    file formats
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../blog/category/performance/">
<span class="md-ellipsis">
    
  
    performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../blog/category/releases/">
<span class="md-ellipsis">
    
  
    releases
  

    
  </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
<label class="md-nav__title" for="__toc">
<span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
<ul class="md-nav__list" data-md-component="toc" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="#genomic-ranges-operations">
<span class="md-ellipsis">
      
        Genomic ranges operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#pileup-operations">
<span class="md-ellipsis">
      
        Pileup operations
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#coordinate-systems-support">
<span class="md-ellipsis">
      
        Coordinate systems support
      
    </span>
</a>
<nav aria-label="Coordinate systems support" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#how-it-works">
<span class="md-ellipsis">
      
        How it works
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#session-parameters">
<span class="md-ellipsis">
      
        Session parameters
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#reading-files-with-coordinate-system-metadata">
<span class="md-ellipsis">
      
        Reading files with coordinate system metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setting-metadata-on-dataframes">
<span class="md-ellipsis">
      
        Setting metadata on DataFrames
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#error-handling">
<span class="md-ellipsis">
      
        Error handling
      
    </span>
</a>
<nav aria-label="Error handling" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#missingcoordinatesystemerror">
<span class="md-ellipsis">
      
        MissingCoordinateSystemError
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#coordinatesystemmismatcherror">
<span class="md-ellipsis">
      
        CoordinateSystemMismatchError
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#default-behavior-lenient-validation">
<span class="md-ellipsis">
      
        Default behavior (lenient validation)
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#strict-mode">
<span class="md-ellipsis">
      
        Strict mode
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#migration-from-previous-versions">
<span class="md-ellipsis">
      
        Migration from previous versions
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-metadata">
<span class="md-ellipsis">
      
        File Metadata
      
    </span>
</a>
<nav aria-label="File Metadata" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#vcf-changes-since-0240">
<span class="md-ellipsis">
      
        VCF changes since 0.24.0
      
    </span>
</a>
<nav aria-label="VCF changes since 0.24.0" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#migration-examples">
<span class="md-ellipsis">
      
        Migration examples
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#metadata-structure">
<span class="md-ellipsis">
      
        Metadata Structure
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#accessing-metadata">
<span class="md-ellipsis">
      
        Accessing Metadata
      
    </span>
</a>
<nav aria-label="Accessing Metadata" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#1-get-all-metadata-as-a-dictionary">
<span class="md-ellipsis">
      
        1. Get all metadata as a dictionary
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#2-print-metadata-as-formatted-json">
<span class="md-ellipsis">
      
        2. Print metadata as formatted JSON
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#3-print-human-readable-summary">
<span class="md-ellipsis">
      
        3. Print human-readable summary
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#format-specific-metadata">
<span class="md-ellipsis">
      
        Format-Specific Metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#setting-custom-metadata">
<span class="md-ellipsis">
      
        Setting Custom Metadata
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#metadata-preservation">
<span class="md-ellipsis">
      
        Metadata Preservation
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#using-metadata-for-debugging">
<span class="md-ellipsis">
      
        Using Metadata for Debugging
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#api-reference">
<span class="md-ellipsis">
      
        API Reference
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-formats-support">
<span class="md-ellipsis">
      
        File formats support
      
    </span>
</a>
<nav aria-label="File formats support" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#indexed-reads-predicate-pushdown">
<span class="md-ellipsis">
      
        Indexed reads &amp; predicate pushdown
      
    </span>
</a>
<nav aria-label="Indexed reads &amp; predicate pushdown" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#supported-index-formats">
<span class="md-ellipsis">
      
        Supported index formats
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usage-with-the-scanread-api">
<span class="md-ellipsis">
      
        Usage with the scan/read API
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#usage-with-the-sql-api">
<span class="md-ellipsis">
      
        Usage with the SQL API
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#automatic-parallel-partitioning">
<span class="md-ellipsis">
      
        Automatic parallel partitioning
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#record-level-filter-pushdown">
<span class="md-ellipsis">
      
        Record-level filter pushdown
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#projection-pushdown">
<span class="md-ellipsis">
      
        Projection pushdown
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#index-file-generation">
<span class="md-ellipsis">
      
        Index file generation
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#file-output">
<span class="md-ellipsis">
      
        File Output
      
    </span>
</a>
<nav aria-label="File Output" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#output-format-support">
<span class="md-ellipsis">
      
        Output Format Support
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#basic-usage">
<span class="md-ellipsis">
      
        Basic Usage
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#sorted-output-with-sort_on_write">
<span class="md-ellipsis">
      
        Sorted Output with sort_on_write
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#cram-output">
<span class="md-ellipsis">
      
        CRAM Output
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#output-compression">
<span class="md-ellipsis">
      
        Output Compression
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#header-preservation">
<span class="md-ellipsis">
      
        Header Preservation
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#polars-extension-methods">
<span class="md-ellipsis">
      
        Polars Extension Methods
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#sql-powered-data-processing">
<span class="md-ellipsis">
      
        SQL-powered data processing
      
    </span>
</a>
<nav aria-label="SQL-powered data processing" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#accessing-registered-tables">
<span class="md-ellipsis">
      
        Accessing registered tables
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#schema-inspection">
<span class="md-ellipsis">
      
        Schema Inspection
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#bam-optional-tags">
<span class="md-ellipsis">
      
        BAM Optional Tags
      
    </span>
</a>
<nav aria-label="BAM Optional Tags" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#usage">
<span class="md-ellipsis">
      
        Usage
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#common-tags">
<span class="md-ellipsis">
      
        Common Tags
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#performance">
<span class="md-ellipsis">
      
        Performance
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#parallel-engine">
<span class="md-ellipsis">
      
        Parallel engine üèéÔ∏è
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#cloud-storage">
<span class="md-ellipsis">
      
        Cloud storage ‚òÅÔ∏è
      
    </span>
</a>
<nav aria-label="Cloud storage ‚òÅÔ∏è" class="md-nav">
<ul class="md-nav__list">
<li class="md-nav__item">
<a class="md-nav__link" href="#example">
<span class="md-ellipsis">
      
        Example
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#supported-features">
<span class="md-ellipsis">
      
        Supported features
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#aws-s3-configuration">
<span class="md-ellipsis">
      
        AWS S3 configuration
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#google-cloud-storage-configuration">
<span class="md-ellipsis">
      
        Google Cloud Storage configuration
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#azure-blob-storage-configuration">
<span class="md-ellipsis">
      
        Azure Blob Storage configuration
      
    </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#streaming">
<span class="md-ellipsis">
      
        Streaming üöÇ
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#compression">
<span class="md-ellipsis">
      
        Compression
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#dataframes-support">
<span class="md-ellipsis">
      
        DataFrames support
      
    </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="#polars-integration">
<span class="md-ellipsis">
      
        Polars Integration
      
    </span>
</a>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<article class="md-content__inner md-typeset">
<a class="md-content__button md-icon" href="https://github.com/biodatageeks/polars-bio/edit/master/docs/features.md" rel="edit" title="Edit this page">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"></path></svg>
</a>
<h1>üî®Features</h1>
<h2 id="genomic-ranges-operations">Genomic ranges operations<a class="headerlink" href="#genomic-ranges-operations" title="Permanent link"></a></h2>
<table>
<thead>
<tr>
<th>operation</th>
<th>Bioframe</th>
<th>polars-bio</th>
<th>PyRanges0</th>
<th>PyRanges1</th>
<th>Pybedtools</th>
<th>GenomicRanges</th>
</tr>
</thead>
<tbody>
<tr>
<td>overlap</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.overlap">overlap</a></td>
<td><a href="../api/#polars_bio.overlap">overlap</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.join">join</a><sup>1</sup></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_overlap.html#the-most-versatile-join-overlaps">join_overlaps</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html">intersect</a><sup>2</sup></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.find_overlaps">find_overlaps</a><sup>3</sup></td>
</tr>
<tr>
<td>nearest</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.closest">closest</a></td>
<td><a href="../api/#polars_bio.nearest">nearest</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.nearest">nearest</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_overlap.html#find-the-closest-interval-nearest-ranges">nearest_ranges</a></td>
<td><a href="https://daler.github.io/pybedtools/autodocs/pybedtools.bedtool.BedTool.closest.html">closest</a><sup>4</sup></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.nearest">nearest</a><sup>5</sup></td>
</tr>
<tr>
<td>count_overlaps</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.count_overlaps">count_overlaps</a></td>
<td><a href="../api/#polars_bio.count_overlaps">count_overlaps</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.count_overlaps">count_overlaps</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_overlap.html">count_overlaps</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html">intersect</a><sup>6</sup></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.count_overlaps">count_overlaps</a></td>
</tr>
<tr>
<td>cluster</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.cluster">cluster</a></td>
<td><a href="../api/#polars_bio.cluster">cluster</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.cluster">cluster</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_overlap.html#grouping-overlapping-intervals-cluster-overlaps">cluster_overlaps</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/cluster.html">cluster</a></td>
<td></td>
</tr>
<tr>
<td>merge</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.merge">merge</a></td>
<td><a href="../api/#polars_bio.merge">merge</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.merge">merge</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_overlap.html">merge_overlaps</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/merge.html">merge</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.reduce">reduce</a><sup>7</sup></td>
</tr>
<tr>
<td>complement</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.complement">complement</a></td>
<td><a href="../api/#polars_bio.complement">complement</a></td>
<td></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_genomic_ops.html#interval-complement">complement_ranges</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/complement.html">complement</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.gaps">gaps</a><sup>8</sup></td>
</tr>
<tr>
<td>subtract</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.subtract">subtract</a></td>
<td><a href="../api/#polars_bio.subtract">subtract</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.subtract">subtract</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/how_to_overlap.html#interval-manipulation-operations-intersect-overlaps-subtract-overlaps">subtract_overlaps</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/subtract.html">subtract</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.subtract">subtract</a></td>
</tr>
<tr>
<td>coverage</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.coverage">coverage</a></td>
<td><a href="../api/#polars_bio.coverage">coverage</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.coverage">coverage</a></td>
<td></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/coverage.html">coverage</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.coverage">coverage</a></td>
</tr>
<tr>
<td>expand</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.expand">expand</a></td>
<td><a href="../api/#polars_bio.LazyFrame.expand">expand</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.extend">extend</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/pyranges_objects.html">extend_ranges</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/slop.html">slop</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.resize">resize</a></td>
</tr>
<tr>
<td>sort</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-intervalops.html#bioframe.ops.sort_bedframe">sort_bedframe</a></td>
<td><a href="../api/#polars_bio.LazyFrame.sort_bedframe">sort</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.sort">sort</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/pyranges_objects.html">sort_ranges</a></td>
<td><a href="https://bedtools.readthedocs.io/en/latest/content/tools/sort.html">sort</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.sort">sort</a></td>
</tr>
<tr>
<td>read_table</td>
<td><a href="https://bioframe.readthedocs.io/en/latest/api-fileops.html#bioframe.io.fileops.read_table">read_table</a></td>
<td><a href="../api/#polars_bio.read_table">read_table</a></td>
<td><a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/readers/index.html#pyranges.readers.read_bed">read_bed</a></td>
<td><a href="https://pyranges1.readthedocs.io/en/latest/pyranges_module.html#pyranges.read_bed">read_bed</a></td>
<td><a href="https://daler.github.io/pybedtools/topical-create-a-bedtool.html#creating-a-bedtool">BedTool</a></td>
<td><a href="https://biocpy.github.io/GenomicRanges/tutorial.html#from-bioinformatic-file-formats">read_bed</a></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol>
<li>There is an <a href="https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/index.html#pyranges.PyRanges.overlap">overlap</a> method in PyRanges, but its output is only limited to indices of intervals from the other Dataframe that overlap.
In Bioframe's <a href="https://bioframe.readthedocs.io/en/latest/guide-performance.html#vs-pyranges-and-optionally-pybedtools">benchmark</a> also <strong>join</strong> method instead of overlap was used.</li>
<li><strong>wa</strong> and <strong>wb</strong> options used to obtain a comparable output.</li>
<li>Output contains only a list with the same length as query, containing hits to overlapping indices. Data transformation is required to obtain the same output as in other libraries.
  Since the performance was far worse than in more efficient libraries anyway, additional data transformation was not included in the benchmark.</li>
<li><strong>s=first</strong> was used to obtain a comparable output.</li>
<li><strong>select="arbitrary"</strong> was used to obtain a comparable output.</li>
<li><strong>-c</strong> flag used with <code>intersect</code> to count overlaps per feature.</li>
<li>GenomicRanges exposes merge as <a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.reduce">reduce()</a>.</li>
<li>GenomicRanges exposes complement as <a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.gaps">gaps()</a>.</li>
</ol>
</div>
<div class="admonition limitations">
<p class="admonition-title">Limitations</p>
<p>For now <em>polars-bio</em> uses <code>int32</code> positions encoding for interval operations (<a href="https://github.com/dcjones/coitrees/issues/18">issue</a>) meaning that it does not support operation on chromosomes longer than <strong>2Gb</strong>. <code>int64</code> support is planned for future releases (<a href="https://github.com/biodatageeks/polars-bio/issues/169">issue</a>).</p>
</div>
<h2 id="pileup-operations">Pileup operations<a class="headerlink" href="#pileup-operations" title="Permanent link"></a></h2>
<p>Per-base read depth computation from alignment files using CIGAR operations. Produces mosdepth-compatible coverage blocks.</p>
<table>
<thead>
<tr>
<th>Feature</th>
<th>mosdepth</th>
<th>samtools depth</th>
<th>polars-bio</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../api/#polars_bio.depth">depth</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
</tbody>
</table>
<div class="language-python highlight"><pre><span></span><code><span id="__span-0-1"><a href="#__codelineno-0-1" id="__codelineno-0-1" name="__codelineno-0-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-0-2"><a href="#__codelineno-0-2" id="__codelineno-0-2" name="__codelineno-0-2"></a>
</span><span id="__span-0-3"><a href="#__codelineno-0-3" id="__codelineno-0-3" name="__codelineno-0-3"></a><span class="c1"># Compute per-base depth from a BAM file</span>
</span><span id="__span-0-4"><a href="#__codelineno-0-4" id="__codelineno-0-4" name="__codelineno-0-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">depth</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-0-5"><a href="#__codelineno-0-5" id="__codelineno-0-5" name="__codelineno-0-5"></a>
</span><span id="__span-0-6"><a href="#__codelineno-0-6" id="__codelineno-0-6" name="__codelineno-0-6"></a><span class="c1"># With MAPQ filter (equivalent to samtools depth -q 20)</span>
</span><span id="__span-0-7"><a href="#__codelineno-0-7" id="__codelineno-0-7" name="__codelineno-0-7"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">depth</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">,</span> <span class="n">min_mapping_quality</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-0-8"><a href="#__codelineno-0-8" id="__codelineno-0-8" name="__codelineno-0-8"></a>
</span><span id="__span-0-9"><a href="#__codelineno-0-9" id="__codelineno-0-9" name="__codelineno-0-9"></a><span class="c1"># Via SQL</span>
</span><span id="__span-0-10"><a href="#__codelineno-0-10" id="__codelineno-0-10" name="__codelineno-0-10"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="s2">"SELECT * FROM depth('alignments.bam')"</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span></code></pre></div>
<h2 id="coordinate-systems-support">Coordinate systems support<a class="headerlink" href="#coordinate-systems-support" title="Permanent link"></a></h2>
<p>polars-bio supports both <strong>0-based half-open</strong> and <strong>1-based closed</strong> coordinate systems for genomic ranges operations. By <strong>default</strong>, it uses <strong>1-based closed</strong> coordinates, which is the native format for VCF, GFF, and SAM/BAM files.</p>
<h3 id="how-it-works">How it works<a class="headerlink" href="#how-it-works" title="Permanent link"></a></h3>
<p>The coordinate system is managed through <strong>DataFrame metadata</strong> that is set at I/O time and read by range operations. This ensures consistency throughout your analysis pipeline.</p>
<pre class="mermaid"><code>flowchart TB
    subgraph IO["I/O Layer"]
        scan["scan_vcf/gff/bam/cram/bed()"]
        read["read_vcf/gff/bam/cram/bed()"]
    end

    subgraph Config["Session Configuration"]
        zero_based["datafusion.bio.coordinate_system_zero_based&lt;br/&gt;(default: false = 1-based)"]
        check["datafusion.bio.coordinate_system_check&lt;br/&gt;(default: false = lenient)"]
    end

    subgraph DF["DataFrame with Metadata"]
        polars_meta["Polars DataFrame/LazyFrame&lt;br/&gt;coordinate_system_zero_based"]
        pandas_meta["Pandas DataFrame&lt;br/&gt;df.attrs"]
    end

    subgraph RangeOps["Range Operations"]
        overlap["overlap()"]
        nearest["nearest()"]
        count["count_overlaps()"]
        coverage["coverage()"]
        merge["merge()"]
        cluster["cluster()"]
        complement["complement()"]
        subtract["subtract()"]
    end

    subgraph Validation["Metadata Validation"]
        validate["validate_coordinate_systems()"]
        error1["MissingCoordinateSystemError"]
        error2["CoordinateSystemMismatchError"]
        fallback["Fallback to global config&lt;br/&gt;+ emit warning"]
    end

    scan --&gt; |"sets metadata"| polars_meta
    read --&gt; |"sets metadata"| polars_meta
    zero_based --&gt; |"use_zero_based param&lt;br/&gt;or default"| scan
    zero_based --&gt; |"use_zero_based param&lt;br/&gt;or default"| read

    polars_meta --&gt; overlap
    polars_meta --&gt; nearest
    polars_meta --&gt; count
    polars_meta --&gt; coverage
    polars_meta --&gt; merge
    polars_meta --&gt; cluster
    polars_meta --&gt; complement
    polars_meta --&gt; subtract
    pandas_meta --&gt; overlap

    overlap --&gt; validate
    nearest --&gt; validate
    count --&gt; validate
    coverage --&gt; validate
    merge --&gt; validate
    cluster --&gt; validate
    complement --&gt; validate
    subtract --&gt; validate

    validate --&gt; |"metadata missing"| check
    validate --&gt; |"metadata mismatch"| error2
    check --&gt; |"true (strict)"| error1
    check --&gt; |"false (lenient)"| fallback
    fallback --&gt; zero_based</code></pre>
<h3 id="session-parameters">Session parameters<a class="headerlink" href="#session-parameters" title="Permanent link"></a></h3>
<p>polars-bio provides two session parameters to control coordinate system behavior:</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>datafusion.bio.coordinate_system_zero_based</code></td>
<td><code>"false"</code> (1-based)</td>
<td>Default coordinate system for I/O operations when <code>use_zero_based</code> is not specified</td>
</tr>
<tr>
<td><code>datafusion.bio.coordinate_system_check</code></td>
<td><code>"false"</code> (lenient)</td>
<td>Whether to raise an error when DataFrame metadata is missing</td>
</tr>
</tbody>
</table>
<div class="language-python highlight"><pre><span></span><code><span id="__span-1-1"><a href="#__codelineno-1-1" id="__codelineno-1-1" name="__codelineno-1-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-1-2"><a href="#__codelineno-1-2" id="__codelineno-1-2" name="__codelineno-1-2"></a>
</span><span id="__span-1-3"><a href="#__codelineno-1-3" id="__codelineno-1-3" name="__codelineno-1-3"></a><span class="c1"># Check current settings</span>
</span><span id="__span-1-4"><a href="#__codelineno-1-4" id="__codelineno-1-4" name="__codelineno-1-4"></a><span class="nb">print</span><span class="p">(</span><span class="n">pb</span><span class="o">.</span><span class="n">get_option</span><span class="p">(</span><span class="n">pb</span><span class="o">.</span><span class="n">POLARS_BIO_COORDINATE_SYSTEM_ZERO_BASED</span><span class="p">))</span>  <span class="c1"># "false"</span>
</span><span id="__span-1-5"><a href="#__codelineno-1-5" id="__codelineno-1-5" name="__codelineno-1-5"></a><span class="nb">print</span><span class="p">(</span><span class="n">pb</span><span class="o">.</span><span class="n">get_option</span><span class="p">(</span><span class="n">pb</span><span class="o">.</span><span class="n">POLARS_BIO_COORDINATE_SYSTEM_CHECK</span><span class="p">))</span>       <span class="c1"># "false"</span>
</span><span id="__span-1-6"><a href="#__codelineno-1-6" id="__codelineno-1-6" name="__codelineno-1-6"></a>
</span><span id="__span-1-7"><a href="#__codelineno-1-7" id="__codelineno-1-7" name="__codelineno-1-7"></a><span class="c1"># Change to 0-based coordinates globally</span>
</span><span id="__span-1-8"><a href="#__codelineno-1-8" id="__codelineno-1-8" name="__codelineno-1-8"></a><span class="n">pb</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="n">pb</span><span class="o">.</span><span class="n">POLARS_BIO_COORDINATE_SYSTEM_ZERO_BASED</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="reading-files-with-coordinate-system-metadata">Reading files with coordinate system metadata<a class="headerlink" href="#reading-files-with-coordinate-system-metadata" title="Permanent link"></a></h3>
<p>When you read genomic files using polars-bio I/O functions, the coordinate system metadata is automatically set on the returned DataFrame:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-2-1"><a href="#__codelineno-2-1" id="__codelineno-2-1" name="__codelineno-2-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-2-2"><a href="#__codelineno-2-2" id="__codelineno-2-2" name="__codelineno-2-2"></a>
</span><span id="__span-2-3"><a href="#__codelineno-2-3" id="__codelineno-2-3" name="__codelineno-2-3"></a><span class="c1"># Default: 1-based coordinates (use_zero_based=False)</span>
</span><span id="__span-2-4"><a href="#__codelineno-2-4" id="__codelineno-2-4" name="__codelineno-2-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-2-5"><a href="#__codelineno-2-5" id="__codelineno-2-5" name="__codelineno-2-5"></a><span class="c1"># Metadata is automatically set: coordinate_system_zero_based=False</span>
</span><span id="__span-2-6"><a href="#__codelineno-2-6" id="__codelineno-2-6" name="__codelineno-2-6"></a>
</span><span id="__span-2-7"><a href="#__codelineno-2-7" id="__codelineno-2-7" name="__codelineno-2-7"></a><span class="c1"># Explicit 0-based coordinates</span>
</span><span id="__span-2-8"><a href="#__codelineno-2-8" id="__codelineno-2-8" name="__codelineno-2-8"></a><span class="n">df_zero</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_bed</span><span class="p">(</span><span class="s2">"regions.bed"</span><span class="p">,</span> <span class="n">use_zero_based</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-2-9"><a href="#__codelineno-2-9" id="__codelineno-2-9" name="__codelineno-2-9"></a><span class="c1"># Metadata is automatically set: coordinate_system_zero_based=True</span>
</span><span id="__span-2-10"><a href="#__codelineno-2-10" id="__codelineno-2-10" name="__codelineno-2-10"></a>
</span><span id="__span-2-11"><a href="#__codelineno-2-11" id="__codelineno-2-11" name="__codelineno-2-11"></a><span class="c1"># Range operations read coordinate system from metadata</span>
</span><span id="__span-2-12"><a href="#__codelineno-2-12" id="__codelineno-2-12" name="__codelineno-2-12"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">df_zero</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>  <span class="c1"># Raises CoordinateSystemMismatchError!</span>
</span></code></pre></div>
<h3 id="setting-metadata-on-dataframes">Setting metadata on DataFrames<a class="headerlink" href="#setting-metadata-on-dataframes" title="Permanent link"></a></h3>
<p>For DataFrames not created via polars-bio I/O functions, you must set the coordinate system metadata manually:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio"/><input id="__tabbed_1_2" name="__tabbed_1" type="radio"/><div class="tabbed-labels"><label for="__tabbed_1_1">Polars DataFrame/LazyFrame</label><label for="__tabbed_1_2">Pandas DataFrame</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-python highlight"><pre><span></span><code><span id="__span-3-1"><a href="#__codelineno-3-1" id="__codelineno-3-1" name="__codelineno-3-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
</span><span id="__span-3-2"><a href="#__codelineno-3-2" id="__codelineno-3-2" name="__codelineno-3-2"></a>
</span><span id="__span-3-3"><a href="#__codelineno-3-3" id="__codelineno-3-3" name="__codelineno-3-3"></a><span class="c1"># Create a DataFrame</span>
</span><span id="__span-3-4"><a href="#__codelineno-3-4" id="__codelineno-3-4" name="__codelineno-3-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
</span><span id="__span-3-5"><a href="#__codelineno-3-5" id="__codelineno-3-5" name="__codelineno-3-5"></a>    <span class="s2">"chrom"</span><span class="p">:</span> <span class="p">[</span><span class="s2">"chr1"</span><span class="p">,</span> <span class="s2">"chr1"</span><span class="p">],</span>
</span><span id="__span-3-6"><a href="#__codelineno-3-6" id="__codelineno-3-6" name="__codelineno-3-6"></a>    <span class="s2">"start"</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span>
</span><span id="__span-3-7"><a href="#__codelineno-3-7" id="__codelineno-3-7" name="__codelineno-3-7"></a>    <span class="s2">"end"</span><span class="p">:</span> <span class="p">[</span><span class="mi">150</span><span class="p">,</span> <span class="mi">250</span><span class="p">]</span>
</span><span id="__span-3-8"><a href="#__codelineno-3-8" id="__codelineno-3-8" name="__codelineno-3-8"></a><span class="p">})</span><span class="o">.</span><span class="n">lazy</span><span class="p">()</span>
</span><span id="__span-3-9"><a href="#__codelineno-3-9" id="__codelineno-3-9" name="__codelineno-3-9"></a>
</span><span id="__span-3-10"><a href="#__codelineno-3-10" id="__codelineno-3-10" name="__codelineno-3-10"></a><span class="c1"># Set coordinate system metadata (requires polars-config-meta)</span>
</span><span id="__span-3-11"><a href="#__codelineno-3-11" id="__codelineno-3-11" name="__codelineno-3-11"></a><span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">config_meta</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">coordinate_system_zero_based</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  <span class="c1"># 1-based</span>
</span><span id="__span-3-12"><a href="#__codelineno-3-12" id="__codelineno-3-12" name="__codelineno-3-12"></a>
</span><span id="__span-3-13"><a href="#__codelineno-3-13" id="__codelineno-3-13" name="__codelineno-3-13"></a><span class="c1"># Now it can be used with range operations</span>
</span><span id="__span-3-14"><a href="#__codelineno-3-14" id="__codelineno-3-14" name="__codelineno-3-14"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">other_df</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
</span></code></pre></div>
</div>
<div class="tabbed-block">
<div class="language-python highlight"><pre><span></span><code><span id="__span-4-1"><a href="#__codelineno-4-1" id="__codelineno-4-1" name="__codelineno-4-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
</span><span id="__span-4-2"><a href="#__codelineno-4-2" id="__codelineno-4-2" name="__codelineno-4-2"></a>
</span><span id="__span-4-3"><a href="#__codelineno-4-3" id="__codelineno-4-3" name="__codelineno-4-3"></a><span class="c1"># Create a DataFrame</span>
</span><span id="__span-4-4"><a href="#__codelineno-4-4" id="__codelineno-4-4" name="__codelineno-4-4"></a><span class="n">pdf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
</span><span id="__span-4-5"><a href="#__codelineno-4-5" id="__codelineno-4-5" name="__codelineno-4-5"></a>    <span class="s2">"chrom"</span><span class="p">:</span> <span class="p">[</span><span class="s2">"chr1"</span><span class="p">,</span> <span class="s2">"chr1"</span><span class="p">],</span>
</span><span id="__span-4-6"><a href="#__codelineno-4-6" id="__codelineno-4-6" name="__codelineno-4-6"></a>    <span class="s2">"start"</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span>
</span><span id="__span-4-7"><a href="#__codelineno-4-7" id="__codelineno-4-7" name="__codelineno-4-7"></a>    <span class="s2">"end"</span><span class="p">:</span> <span class="p">[</span><span class="mi">150</span><span class="p">,</span> <span class="mi">250</span><span class="p">]</span>
</span><span id="__span-4-8"><a href="#__codelineno-4-8" id="__codelineno-4-8" name="__codelineno-4-8"></a><span class="p">})</span>
</span><span id="__span-4-9"><a href="#__codelineno-4-9" id="__codelineno-4-9" name="__codelineno-4-9"></a>
</span><span id="__span-4-10"><a href="#__codelineno-4-10" id="__codelineno-4-10" name="__codelineno-4-10"></a><span class="c1"># Set coordinate system metadata via df.attrs</span>
</span><span id="__span-4-11"><a href="#__codelineno-4-11" id="__codelineno-4-11" name="__codelineno-4-11"></a><span class="n">pdf</span><span class="o">.</span><span class="n">attrs</span><span class="p">[</span><span class="s2">"coordinate_system_zero_based"</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># 1-based</span>
</span><span id="__span-4-12"><a href="#__codelineno-4-12" id="__codelineno-4-12" name="__codelineno-4-12"></a>
</span><span id="__span-4-13"><a href="#__codelineno-4-13" id="__codelineno-4-13" name="__codelineno-4-13"></a><span class="c1"># Now it can be used with range operations</span>
</span><span id="__span-4-14"><a href="#__codelineno-4-14" id="__codelineno-4-14" name="__codelineno-4-14"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">pdf</span><span class="p">,</span> <span class="n">other_df</span><span class="p">,</span> <span class="n">output_type</span><span class="o">=</span><span class="s2">"pandas.DataFrame"</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
</span></code></pre></div>
</div>
</div>
</div>
<h3 id="error-handling">Error handling<a class="headerlink" href="#error-handling" title="Permanent link"></a></h3>
<p>polars-bio raises specific errors to prevent coordinate system mismatches:</p>
<h4 id="missingcoordinatesystemerror">MissingCoordinateSystemError<a class="headerlink" href="#missingcoordinatesystemerror" title="Permanent link"></a></h4>
<p>Raised when a DataFrame lacks coordinate system metadata:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-5-1"><a href="#__codelineno-5-1" id="__codelineno-5-1" name="__codelineno-5-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
</span><span id="__span-5-2"><a href="#__codelineno-5-2" id="__codelineno-5-2" name="__codelineno-5-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-5-3"><a href="#__codelineno-5-3" id="__codelineno-5-3" name="__codelineno-5-3"></a>
</span><span id="__span-5-4"><a href="#__codelineno-5-4" id="__codelineno-5-4" name="__codelineno-5-4"></a><span class="c1"># DataFrame without metadata</span>
</span><span id="__span-5-5"><a href="#__codelineno-5-5" id="__codelineno-5-5" name="__codelineno-5-5"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">"chrom"</span><span class="p">:</span> <span class="p">[</span><span class="s2">"chr1"</span><span class="p">],</span> <span class="s2">"start"</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">],</span> <span class="s2">"end"</span><span class="p">:</span> <span class="p">[</span><span class="mi">200</span><span class="p">]})</span><span class="o">.</span><span class="n">lazy</span><span class="p">()</span>
</span><span id="__span-5-6"><a href="#__codelineno-5-6" id="__codelineno-5-6" name="__codelineno-5-6"></a>
</span><span id="__span-5-7"><a href="#__codelineno-5-7" id="__codelineno-5-7" name="__codelineno-5-7"></a><span class="c1"># This raises MissingCoordinateSystemError</span>
</span><span id="__span-5-8"><a href="#__codelineno-5-8" id="__codelineno-5-8" name="__codelineno-5-8"></a><span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">other_df</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
</span></code></pre></div>
<p><strong>How to fix:</strong> Set metadata on your DataFrame before passing it to range operations (see examples above).</p>
<h4 id="coordinatesystemmismatcherror">CoordinateSystemMismatchError<a class="headerlink" href="#coordinatesystemmismatcherror" title="Permanent link"></a></h4>
<p>Raised when two DataFrames have different coordinate systems:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-6-1"><a href="#__codelineno-6-1" id="__codelineno-6-1" name="__codelineno-6-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-6-2"><a href="#__codelineno-6-2" id="__codelineno-6-2" name="__codelineno-6-2"></a>
</span><span id="__span-6-3"><a href="#__codelineno-6-3" id="__codelineno-6-3" name="__codelineno-6-3"></a><span class="c1"># One DataFrame is 1-based, another is 0-based</span>
</span><span id="__span-6-4"><a href="#__codelineno-6-4" id="__codelineno-6-4" name="__codelineno-6-4"></a><span class="n">df1</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"file.vcf"</span><span class="p">)</span>                    <span class="c1"># 1-based (default)</span>
</span><span id="__span-6-5"><a href="#__codelineno-6-5" id="__codelineno-6-5" name="__codelineno-6-5"></a><span class="n">df2</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_bed</span><span class="p">(</span><span class="s2">"file.bed"</span><span class="p">,</span> <span class="n">use_zero_based</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># 0-based</span>
</span><span id="__span-6-6"><a href="#__codelineno-6-6" id="__codelineno-6-6" name="__codelineno-6-6"></a>
</span><span id="__span-6-7"><a href="#__codelineno-6-7" id="__codelineno-6-7" name="__codelineno-6-7"></a><span class="c1"># This raises CoordinateSystemMismatchError</span>
</span><span id="__span-6-8"><a href="#__codelineno-6-8" id="__codelineno-6-8" name="__codelineno-6-8"></a><span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
</span></code></pre></div>
<p><strong>How to fix:</strong> Ensure both DataFrames use the same coordinate system.</p>
<h3 id="default-behavior-lenient-validation">Default behavior (lenient validation)<a class="headerlink" href="#default-behavior-lenient-validation" title="Permanent link"></a></h3>
<p>By default, polars-bio uses lenient validation (<code>coordinate_system_check=false</code>). When a DataFrame lacks coordinate system metadata, it falls back to the global configuration and emits a warning:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-7-1"><a href="#__codelineno-7-1" id="__codelineno-7-1" name="__codelineno-7-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
</span><span id="__span-7-2"><a href="#__codelineno-7-2" id="__codelineno-7-2" name="__codelineno-7-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-7-3"><a href="#__codelineno-7-3" id="__codelineno-7-3" name="__codelineno-7-3"></a>
</span><span id="__span-7-4"><a href="#__codelineno-7-4" id="__codelineno-7-4" name="__codelineno-7-4"></a><span class="c1"># DataFrames without metadata will use the global config with a warning</span>
</span><span id="__span-7-5"><a href="#__codelineno-7-5" id="__codelineno-7-5" name="__codelineno-7-5"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">"chrom"</span><span class="p">:</span> <span class="p">[</span><span class="s2">"chr1"</span><span class="p">],</span> <span class="s2">"start"</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">],</span> <span class="s2">"end"</span><span class="p">:</span> <span class="p">[</span><span class="mi">200</span><span class="p">]})</span><span class="o">.</span><span class="n">lazy</span><span class="p">()</span>
</span><span id="__span-7-6"><a href="#__codelineno-7-6" id="__codelineno-7-6" name="__codelineno-7-6"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">other_df</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>  <span class="c1"># Uses global coordinate system setting</span>
</span><span id="__span-7-7"><a href="#__codelineno-7-7" id="__codelineno-7-7" name="__codelineno-7-7"></a><span class="c1"># Warning: Coordinate system metadata is missing. Using global config...</span>
</span></code></pre></div>
<h3 id="strict-mode">Strict mode<a class="headerlink" href="#strict-mode" title="Permanent link"></a></h3>
<p>For production pipelines where coordinate system consistency is critical, you can enable strict validation:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-8-1"><a href="#__codelineno-8-1" id="__codelineno-8-1" name="__codelineno-8-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-8-2"><a href="#__codelineno-8-2" id="__codelineno-8-2" name="__codelineno-8-2"></a>
</span><span id="__span-8-3"><a href="#__codelineno-8-3" id="__codelineno-8-3" name="__codelineno-8-3"></a><span class="c1"># Enable strict coordinate system check</span>
</span><span id="__span-8-4"><a href="#__codelineno-8-4" id="__codelineno-8-4" name="__codelineno-8-4"></a><span class="n">pb</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="n">pb</span><span class="o">.</span><span class="n">POLARS_BIO_COORDINATE_SYSTEM_CHECK</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</span><span id="__span-8-5"><a href="#__codelineno-8-5" id="__codelineno-8-5" name="__codelineno-8-5"></a>
</span><span id="__span-8-6"><a href="#__codelineno-8-6" id="__codelineno-8-6" name="__codelineno-8-6"></a><span class="c1"># Now DataFrames without metadata will raise MissingCoordinateSystemError</span>
</span></code></pre></div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Enable strict mode in production pipelines to catch coordinate system mismatches early and prevent incorrect results.</p>
</div>
<h3 id="migration-from-previous-versions">Migration from previous versions<a class="headerlink" href="#migration-from-previous-versions" title="Permanent link"></a></h3>
<p>If you're upgrading from a previous version of polars-bio:</p>
<ol>
<li><strong>Range operations no longer accept <code>use_zero_based</code> parameter</strong> - coordinate system is read from DataFrame metadata</li>
<li><strong>I/O functions use <code>use_zero_based</code> parameter</strong> (renamed from <code>one_based</code> with inverted logic)</li>
<li><strong>Pandas DataFrames require explicit metadata</strong> - set <code>df.attrs["coordinate_system_zero_based"]</code> before range operations</li>
</ol>
<div class="language-python highlight"><pre><span></span><code><span id="__span-9-1"><a href="#__codelineno-9-1" id="__codelineno-9-1" name="__codelineno-9-1"></a><span class="c1"># Before (old API)</span>
</span><span id="__span-9-2"><a href="#__codelineno-9-2" id="__codelineno-9-2" name="__codelineno-9-2"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="n">use_zero_based</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>
</span><span id="__span-9-3"><a href="#__codelineno-9-3" id="__codelineno-9-3" name="__codelineno-9-3"></a>
</span><span id="__span-9-4"><a href="#__codelineno-9-4" id="__codelineno-9-4" name="__codelineno-9-4"></a><span class="c1"># After (new API) - set metadata at I/O time or on DataFrames</span>
</span><span id="__span-9-5"><a href="#__codelineno-9-5" id="__codelineno-9-5" name="__codelineno-9-5"></a><span class="n">df1</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"file.vcf"</span><span class="p">,</span> <span class="n">use_zero_based</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-9-6"><a href="#__codelineno-9-6" id="__codelineno-9-6" name="__codelineno-9-6"></a><span class="n">df2</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_bed</span><span class="p">(</span><span class="s2">"file.bed"</span><span class="p">,</span> <span class="n">use_zero_based</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-9-7"><a href="#__codelineno-9-7" id="__codelineno-9-7" name="__codelineno-9-7"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span> <span class="n">df2</span><span class="p">,</span> <span class="o">...</span><span class="p">)</span>  <span class="c1"># Reads from metadata</span>
</span></code></pre></div>
<h2 id="file-metadata">File Metadata<a class="headerlink" href="#file-metadata" title="Permanent link"></a></h2>
<p>polars-bio automatically attaches comprehensive metadata to DataFrames when reading genomic files. This metadata includes format information, coordinate systems, and format-specific details like VCF header fields.</p>
<h3 id="vcf-changes-since-0240">VCF changes since 0.24.0<a class="headerlink" href="#vcf-changes-since-0240" title="Permanent link"></a></h3>
<p>Compared to <code>v0.24.0</code>, VCF handling has the following behavior changes:</p>
<ol>
<li>
<p><strong>Multisample FORMAT schema changed (breaking)</strong>
   Multisample VCF FORMAT data is now exposed as a nested <code>genotypes</code> column (<code>list&lt;struct&lt;sample_id, values&gt;&gt;</code>) instead of flattened columns like <code>NA12878_GT</code>.</p>
</li>
<li>
<p><strong>Single-sample FORMAT schema unchanged</strong>
   Single-sample VCFs still expose FORMAT fields as top-level columns (<code>GT</code>, <code>DP</code>, <code>GQ</code>, ...).</p>
</li>
<li>
<p><strong><code>info_fields=None</code> now includes all INFO fields by default</strong>
   When <code>info_fields</code> is not provided, all header INFO fields are available in the schema.
   Use <code>info_fields=[]</code> to exclude INFO columns.</p>
</li>
<li>
<p><strong>FORMAT metadata fidelity improved</strong>
<code>meta["header"]["format_fields"]</code> now preserves FORMAT <code>number</code>/<code>type</code>/<code>description</code> via schema-level metadata.</p>
</li>
</ol>
<h4 id="migration-examples">Migration examples<a class="headerlink" href="#migration-examples" title="Permanent link"></a></h4>
<div class="language-python highlight"><pre><span></span><code><span id="__span-10-1"><a href="#__codelineno-10-1" id="__codelineno-10-1" name="__codelineno-10-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-10-2"><a href="#__codelineno-10-2" id="__codelineno-10-2" name="__codelineno-10-2"></a>
</span><span id="__span-10-3"><a href="#__codelineno-10-3" id="__codelineno-10-3" name="__codelineno-10-3"></a><span class="c1"># Before (v0.24.0 style multisample access):</span>
</span><span id="__span-10-4"><a href="#__codelineno-10-4" id="__codelineno-10-4" name="__codelineno-10-4"></a><span class="c1"># df = pb.read_vcf("multisample.vcf", format_fields=["GT", "DP"])</span>
</span><span id="__span-10-5"><a href="#__codelineno-10-5" id="__codelineno-10-5" name="__codelineno-10-5"></a><span class="c1"># df.select(["chrom", "start", "NA12878_GT", "NA12878_DP"])</span>
</span><span id="__span-10-6"><a href="#__codelineno-10-6" id="__codelineno-10-6" name="__codelineno-10-6"></a>
</span><span id="__span-10-7"><a href="#__codelineno-10-7" id="__codelineno-10-7" name="__codelineno-10-7"></a><span class="c1"># Now (current):</span>
</span><span id="__span-10-8"><a href="#__codelineno-10-8" id="__codelineno-10-8" name="__codelineno-10-8"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="s2">"multisample.vcf"</span><span class="p">,</span> <span class="n">format_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"GT"</span><span class="p">,</span> <span class="s2">"DP"</span><span class="p">])</span>
</span><span id="__span-10-9"><a href="#__codelineno-10-9" id="__codelineno-10-9" name="__codelineno-10-9"></a><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="s2">"chrom"</span><span class="p">,</span> <span class="s2">"start"</span><span class="p">,</span> <span class="s2">"genotypes"</span><span class="p">])</span>
</span><span id="__span-10-10"><a href="#__codelineno-10-10" id="__codelineno-10-10" name="__codelineno-10-10"></a>
</span><span id="__span-10-11"><a href="#__codelineno-10-11" id="__codelineno-10-11" name="__codelineno-10-11"></a><span class="c1"># INFO behavior:</span>
</span><span id="__span-10-12"><a href="#__codelineno-10-12" id="__codelineno-10-12" name="__codelineno-10-12"></a><span class="n">df_all_info</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>              <span class="c1"># all INFO fields</span>
</span><span id="__span-10-13"><a href="#__codelineno-10-13" id="__codelineno-10-13" name="__codelineno-10-13"></a><span class="n">df_no_info</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">,</span> <span class="n">info_fields</span><span class="o">=</span><span class="p">[])</span>  <span class="c1"># no INFO fields</span>
</span></code></pre></div>
<h3 id="metadata-structure">Metadata Structure<a class="headerlink" href="#metadata-structure" title="Permanent link"></a></h3>
<p>The metadata is stored in a clean, user-friendly structure:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-11-1"><a href="#__codelineno-11-1" id="__codelineno-11-1" name="__codelineno-11-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-11-2"><a href="#__codelineno-11-2" id="__codelineno-11-2" name="__codelineno-11-2"></a>
</span><span id="__span-11-3"><a href="#__codelineno-11-3" id="__codelineno-11-3" name="__codelineno-11-3"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-11-4"><a href="#__codelineno-11-4" id="__codelineno-11-4" name="__codelineno-11-4"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-11-5"><a href="#__codelineno-11-5" id="__codelineno-11-5" name="__codelineno-11-5"></a>
</span><span id="__span-11-6"><a href="#__codelineno-11-6" id="__codelineno-11-6" name="__codelineno-11-6"></a><span class="c1"># Returns:</span>
</span><span id="__span-11-7"><a href="#__codelineno-11-7" id="__codelineno-11-7" name="__codelineno-11-7"></a><span class="p">{</span>
</span><span id="__span-11-8"><a href="#__codelineno-11-8" id="__codelineno-11-8" name="__codelineno-11-8"></a>  <span class="s2">"format"</span><span class="p">:</span> <span class="s2">"vcf"</span><span class="p">,</span>                           <span class="c1"># File format</span>
</span><span id="__span-11-9"><a href="#__codelineno-11-9" id="__codelineno-11-9" name="__codelineno-11-9"></a>  <span class="s2">"path"</span><span class="p">:</span> <span class="s2">"variants.vcf"</span><span class="p">,</span>                    <span class="c1"># Source file path</span>
</span><span id="__span-11-10"><a href="#__codelineno-11-10" id="__codelineno-11-10" name="__codelineno-11-10"></a>  <span class="s2">"coordinate_system_zero_based"</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>     <span class="c1"># Coordinate system (VCF is 1-based)</span>
</span><span id="__span-11-11"><a href="#__codelineno-11-11" id="__codelineno-11-11" name="__codelineno-11-11"></a>  <span class="s2">"header"</span><span class="p">:</span> <span class="p">{</span>
</span><span id="__span-11-12"><a href="#__codelineno-11-12" id="__codelineno-11-12" name="__codelineno-11-12"></a>    <span class="s2">"version"</span><span class="p">:</span> <span class="s2">"VCFv4.2"</span><span class="p">,</span>                    <span class="c1"># VCF version</span>
</span><span id="__span-11-13"><a href="#__codelineno-11-13" id="__codelineno-11-13" name="__codelineno-11-13"></a>    <span class="s2">"sample_names"</span><span class="p">:</span> <span class="p">[</span><span class="s2">"Sample1"</span><span class="p">,</span> <span class="s2">"Sample2"</span><span class="p">],</span>  <span class="c1"># Sample names</span>
</span><span id="__span-11-14"><a href="#__codelineno-11-14" id="__codelineno-11-14" name="__codelineno-11-14"></a>    <span class="s2">"info_fields"</span><span class="p">:</span> <span class="p">{</span>                         <span class="c1"># INFO field definitions</span>
</span><span id="__span-11-15"><a href="#__codelineno-11-15" id="__codelineno-11-15" name="__codelineno-11-15"></a>      <span class="s2">"AF"</span><span class="p">:</span> <span class="p">{</span>
</span><span id="__span-11-16"><a href="#__codelineno-11-16" id="__codelineno-11-16" name="__codelineno-11-16"></a>        <span class="s2">"number"</span><span class="p">:</span> <span class="s2">"A"</span><span class="p">,</span>
</span><span id="__span-11-17"><a href="#__codelineno-11-17" id="__codelineno-11-17" name="__codelineno-11-17"></a>        <span class="s2">"type"</span><span class="p">:</span> <span class="s2">"Float"</span><span class="p">,</span>
</span><span id="__span-11-18"><a href="#__codelineno-11-18" id="__codelineno-11-18" name="__codelineno-11-18"></a>        <span class="s2">"description"</span><span class="p">:</span> <span class="s2">"Allele Frequency"</span><span class="p">,</span>
</span><span id="__span-11-19"><a href="#__codelineno-11-19" id="__codelineno-11-19" name="__codelineno-11-19"></a>        <span class="s2">"id"</span><span class="p">:</span> <span class="s2">"AF"</span>
</span><span id="__span-11-20"><a href="#__codelineno-11-20" id="__codelineno-11-20" name="__codelineno-11-20"></a>      <span class="p">}</span>
</span><span id="__span-11-21"><a href="#__codelineno-11-21" id="__codelineno-11-21" name="__codelineno-11-21"></a>    <span class="p">},</span>
</span><span id="__span-11-22"><a href="#__codelineno-11-22" id="__codelineno-11-22" name="__codelineno-11-22"></a>    <span class="s2">"format_fields"</span><span class="p">:</span> <span class="p">{</span>                       <span class="c1"># FORMAT field definitions</span>
</span><span id="__span-11-23"><a href="#__codelineno-11-23" id="__codelineno-11-23" name="__codelineno-11-23"></a>      <span class="s2">"GT"</span><span class="p">:</span> <span class="p">{</span>
</span><span id="__span-11-24"><a href="#__codelineno-11-24" id="__codelineno-11-24" name="__codelineno-11-24"></a>        <span class="s2">"number"</span><span class="p">:</span> <span class="s2">"1"</span><span class="p">,</span>
</span><span id="__span-11-25"><a href="#__codelineno-11-25" id="__codelineno-11-25" name="__codelineno-11-25"></a>        <span class="s2">"type"</span><span class="p">:</span> <span class="s2">"String"</span><span class="p">,</span>
</span><span id="__span-11-26"><a href="#__codelineno-11-26" id="__codelineno-11-26" name="__codelineno-11-26"></a>        <span class="s2">"description"</span><span class="p">:</span> <span class="s2">"Genotype"</span>
</span><span id="__span-11-27"><a href="#__codelineno-11-27" id="__codelineno-11-27" name="__codelineno-11-27"></a>      <span class="p">}</span>
</span><span id="__span-11-28"><a href="#__codelineno-11-28" id="__codelineno-11-28" name="__codelineno-11-28"></a>    <span class="p">},</span>
</span><span id="__span-11-29"><a href="#__codelineno-11-29" id="__codelineno-11-29" name="__codelineno-11-29"></a>    <span class="s2">"contigs"</span><span class="p">:</span> <span class="p">[</span><span class="o">...</span><span class="p">],</span>                        <span class="c1"># Contig definitions</span>
</span><span id="__span-11-30"><a href="#__codelineno-11-30" id="__codelineno-11-30" name="__codelineno-11-30"></a>    <span class="s2">"filters"</span><span class="p">:</span> <span class="p">[</span><span class="o">...</span><span class="p">],</span>                        <span class="c1"># Filter definitions</span>
</span><span id="__span-11-31"><a href="#__codelineno-11-31" id="__codelineno-11-31" name="__codelineno-11-31"></a>    <span class="s2">"_datafusion_table_name"</span><span class="p">:</span> <span class="s2">"variants"</span>     <span class="c1"># Internal table name (for debugging)</span>
</span><span id="__span-11-32"><a href="#__codelineno-11-32" id="__codelineno-11-32" name="__codelineno-11-32"></a>  <span class="p">}</span>
</span><span id="__span-11-33"><a href="#__codelineno-11-33" id="__codelineno-11-33" name="__codelineno-11-33"></a><span class="p">}</span>
</span></code></pre></div>
<h3 id="accessing-metadata">Accessing Metadata<a class="headerlink" href="#accessing-metadata" title="Permanent link"></a></h3>
<p>polars-bio provides three main functions for working with metadata:</p>
<h4 id="1-get-all-metadata-as-a-dictionary">1. Get all metadata as a dictionary<a class="headerlink" href="#1-get-all-metadata-as-a-dictionary" title="Permanent link"></a></h4>
<div class="language-python highlight"><pre><span></span><code><span id="__span-12-1"><a href="#__codelineno-12-1" id="__codelineno-12-1" name="__codelineno-12-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-12-2"><a href="#__codelineno-12-2" id="__codelineno-12-2" name="__codelineno-12-2"></a>
</span><span id="__span-12-3"><a href="#__codelineno-12-3" id="__codelineno-12-3" name="__codelineno-12-3"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"file.vcf"</span><span class="p">)</span>
</span><span id="__span-12-4"><a href="#__codelineno-12-4" id="__codelineno-12-4" name="__codelineno-12-4"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-12-5"><a href="#__codelineno-12-5" id="__codelineno-12-5" name="__codelineno-12-5"></a>
</span><span id="__span-12-6"><a href="#__codelineno-12-6" id="__codelineno-12-6" name="__codelineno-12-6"></a><span class="c1"># Access different parts</span>
</span><span id="__span-12-7"><a href="#__codelineno-12-7" id="__codelineno-12-7" name="__codelineno-12-7"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"format"</span><span class="p">])</span>                       <span class="c1"># "vcf"</span>
</span><span id="__span-12-8"><a href="#__codelineno-12-8" id="__codelineno-12-8" name="__codelineno-12-8"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"path"</span><span class="p">])</span>                         <span class="c1"># "file.vcf"</span>
</span><span id="__span-12-9"><a href="#__codelineno-12-9" id="__codelineno-12-9" name="__codelineno-12-9"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"coordinate_system_zero_based"</span><span class="p">])</span> <span class="c1"># False (1-based)</span>
</span><span id="__span-12-10"><a href="#__codelineno-12-10" id="__codelineno-12-10" name="__codelineno-12-10"></a>
</span><span id="__span-12-11"><a href="#__codelineno-12-11" id="__codelineno-12-11" name="__codelineno-12-11"></a><span class="c1"># Access VCF-specific fields</span>
</span><span id="__span-12-12"><a href="#__codelineno-12-12" id="__codelineno-12-12" name="__codelineno-12-12"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"version"</span><span class="p">])</span>            <span class="c1"># "VCFv4.2"</span>
</span><span id="__span-12-13"><a href="#__codelineno-12-13" id="__codelineno-12-13" name="__codelineno-12-13"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"sample_names"</span><span class="p">])</span>       <span class="c1"># ["Sample1", "Sample2"]</span>
</span><span id="__span-12-14"><a href="#__codelineno-12-14" id="__codelineno-12-14" name="__codelineno-12-14"></a>
</span><span id="__span-12-15"><a href="#__codelineno-12-15" id="__codelineno-12-15" name="__codelineno-12-15"></a><span class="c1"># Access INFO field definitions</span>
</span><span id="__span-12-16"><a href="#__codelineno-12-16" id="__codelineno-12-16" name="__codelineno-12-16"></a><span class="n">af_field</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"info_fields"</span><span class="p">][</span><span class="s2">"AF"</span><span class="p">]</span>
</span><span id="__span-12-17"><a href="#__codelineno-12-17" id="__codelineno-12-17" name="__codelineno-12-17"></a><span class="nb">print</span><span class="p">(</span><span class="n">af_field</span><span class="p">[</span><span class="s2">"type"</span><span class="p">])</span>                     <span class="c1"># "Float"</span>
</span><span id="__span-12-18"><a href="#__codelineno-12-18" id="__codelineno-12-18" name="__codelineno-12-18"></a><span class="nb">print</span><span class="p">(</span><span class="n">af_field</span><span class="p">[</span><span class="s2">"description"</span><span class="p">])</span>              <span class="c1"># "Allele Frequency"</span>
</span><span id="__span-12-19"><a href="#__codelineno-12-19" id="__codelineno-12-19" name="__codelineno-12-19"></a>
</span><span id="__span-12-20"><a href="#__codelineno-12-20" id="__codelineno-12-20" name="__codelineno-12-20"></a><span class="c1"># Access FORMAT field definitions</span>
</span><span id="__span-12-21"><a href="#__codelineno-12-21" id="__codelineno-12-21" name="__codelineno-12-21"></a><span class="n">gt_field</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"format_fields"</span><span class="p">][</span><span class="s2">"GT"</span><span class="p">]</span>
</span><span id="__span-12-22"><a href="#__codelineno-12-22" id="__codelineno-12-22" name="__codelineno-12-22"></a><span class="nb">print</span><span class="p">(</span><span class="n">gt_field</span><span class="p">[</span><span class="s2">"type"</span><span class="p">])</span>                     <span class="c1"># "String"</span>
</span></code></pre></div>
<h4 id="2-print-metadata-as-formatted-json">2. Print metadata as formatted JSON<a class="headerlink" href="#2-print-metadata-as-formatted-json" title="Permanent link"></a></h4>
<div class="language-python highlight"><pre><span></span><code><span id="__span-13-1"><a href="#__codelineno-13-1" id="__codelineno-13-1" name="__codelineno-13-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-13-2"><a href="#__codelineno-13-2" id="__codelineno-13-2" name="__codelineno-13-2"></a>
</span><span id="__span-13-3"><a href="#__codelineno-13-3" id="__codelineno-13-3" name="__codelineno-13-3"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"file.vcf"</span><span class="p">)</span>
</span><span id="__span-13-4"><a href="#__codelineno-13-4" id="__codelineno-13-4" name="__codelineno-13-4"></a>
</span><span id="__span-13-5"><a href="#__codelineno-13-5" id="__codelineno-13-5" name="__codelineno-13-5"></a><span class="c1"># Print as pretty JSON</span>
</span><span id="__span-13-6"><a href="#__codelineno-13-6" id="__codelineno-13-6" name="__codelineno-13-6"></a><span class="n">pb</span><span class="o">.</span><span class="n">print_metadata_json</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-13-7"><a href="#__codelineno-13-7" id="__codelineno-13-7" name="__codelineno-13-7"></a>
</span><span id="__span-13-8"><a href="#__codelineno-13-8" id="__codelineno-13-8" name="__codelineno-13-8"></a><span class="c1"># Customize indentation</span>
</span><span id="__span-13-9"><a href="#__codelineno-13-9" id="__codelineno-13-9" name="__codelineno-13-9"></a><span class="n">pb</span><span class="o">.</span><span class="n">print_metadata_json</span><span class="p">(</span><span class="n">lf</span><span class="p">,</span> <span class="n">indent</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</span></code></pre></div>
<h4 id="3-print-human-readable-summary">3. Print human-readable summary<a class="headerlink" href="#3-print-human-readable-summary" title="Permanent link"></a></h4>
<div class="language-python highlight"><pre><span></span><code><span id="__span-14-1"><a href="#__codelineno-14-1" id="__codelineno-14-1" name="__codelineno-14-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-14-2"><a href="#__codelineno-14-2" id="__codelineno-14-2" name="__codelineno-14-2"></a>
</span><span id="__span-14-3"><a href="#__codelineno-14-3" id="__codelineno-14-3" name="__codelineno-14-3"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"file.vcf"</span><span class="p">)</span>
</span><span id="__span-14-4"><a href="#__codelineno-14-4" id="__codelineno-14-4" name="__codelineno-14-4"></a><span class="n">pb</span><span class="o">.</span><span class="n">print_metadata_summary</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span></code></pre></div>
<p>Output:
</p><div class="language-text highlight"><pre><span></span><code><span id="__span-15-1"><a href="#__codelineno-15-1" id="__codelineno-15-1" name="__codelineno-15-1"></a>======================================================================
</span><span id="__span-15-2"><a href="#__codelineno-15-2" id="__codelineno-15-2" name="__codelineno-15-2"></a>Metadata Summary
</span><span id="__span-15-3"><a href="#__codelineno-15-3" id="__codelineno-15-3" name="__codelineno-15-3"></a>======================================================================
</span><span id="__span-15-4"><a href="#__codelineno-15-4" id="__codelineno-15-4" name="__codelineno-15-4"></a>
</span><span id="__span-15-5"><a href="#__codelineno-15-5" id="__codelineno-15-5" name="__codelineno-15-5"></a>Format: vcf
</span><span id="__span-15-6"><a href="#__codelineno-15-6" id="__codelineno-15-6" name="__codelineno-15-6"></a>Path: file.vcf
</span><span id="__span-15-7"><a href="#__codelineno-15-7" id="__codelineno-15-7" name="__codelineno-15-7"></a>Coordinate System: 1-based
</span><span id="__span-15-8"><a href="#__codelineno-15-8" id="__codelineno-15-8" name="__codelineno-15-8"></a>
</span><span id="__span-15-9"><a href="#__codelineno-15-9" id="__codelineno-15-9" name="__codelineno-15-9"></a>Format-specific metadata:
</span><span id="__span-15-10"><a href="#__codelineno-15-10" id="__codelineno-15-10" name="__codelineno-15-10"></a>----------------------------------------------------------------------
</span><span id="__span-15-11"><a href="#__codelineno-15-11" id="__codelineno-15-11" name="__codelineno-15-11"></a>  VCF Version: VCFv4.2
</span><span id="__span-15-12"><a href="#__codelineno-15-12" id="__codelineno-15-12" name="__codelineno-15-12"></a>  Samples (3): Sample1, Sample2, Sample3
</span><span id="__span-15-13"><a href="#__codelineno-15-13" id="__codelineno-15-13" name="__codelineno-15-13"></a>  INFO fields: 5
</span><span id="__span-15-14"><a href="#__codelineno-15-14" id="__codelineno-15-14" name="__codelineno-15-14"></a>    - AF: Float (Allele Frequency)
</span><span id="__span-15-15"><a href="#__codelineno-15-15" id="__codelineno-15-15" name="__codelineno-15-15"></a>    - DP: Integer (Total Depth)
</span><span id="__span-15-16"><a href="#__codelineno-15-16" id="__codelineno-15-16" name="__codelineno-15-16"></a>    - AC: Integer (Allele Count)
</span><span id="__span-15-17"><a href="#__codelineno-15-17" id="__codelineno-15-17" name="__codelineno-15-17"></a>  FORMAT fields: 3
</span><span id="__span-15-18"><a href="#__codelineno-15-18" id="__codelineno-15-18" name="__codelineno-15-18"></a>    - GT: String (Genotype)
</span><span id="__span-15-19"><a href="#__codelineno-15-19" id="__codelineno-15-19" name="__codelineno-15-19"></a>    - DP: Integer (Read Depth)
</span><span id="__span-15-20"><a href="#__codelineno-15-20" id="__codelineno-15-20" name="__codelineno-15-20"></a>    - GQ: Integer (Genotype Quality)
</span><span id="__span-15-21"><a href="#__codelineno-15-21" id="__codelineno-15-21" name="__codelineno-15-21"></a>
</span><span id="__span-15-22"><a href="#__codelineno-15-22" id="__codelineno-15-22" name="__codelineno-15-22"></a>======================================================================
</span></code></pre></div><p></p>
<h3 id="format-specific-metadata">Format-Specific Metadata<a class="headerlink" href="#format-specific-metadata" title="Permanent link"></a></h3>
<p>Different file formats include different metadata:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:3"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio"/><input id="__tabbed_2_2" name="__tabbed_2" type="radio"/><input id="__tabbed_2_3" name="__tabbed_2" type="radio"/><div class="tabbed-labels"><label for="__tabbed_2_1">VCF</label><label for="__tabbed_2_2">FASTQ</label><label for="__tabbed_2_3">BED/BAM/GFF</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-python highlight"><pre><span></span><code><span id="__span-16-1"><a href="#__codelineno-16-1" id="__codelineno-16-1" name="__codelineno-16-1"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-16-2"><a href="#__codelineno-16-2" id="__codelineno-16-2" name="__codelineno-16-2"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-16-3"><a href="#__codelineno-16-3" id="__codelineno-16-3" name="__codelineno-16-3"></a>
</span><span id="__span-16-4"><a href="#__codelineno-16-4" id="__codelineno-16-4" name="__codelineno-16-4"></a><span class="c1"># VCF header metadata</span>
</span><span id="__span-16-5"><a href="#__codelineno-16-5" id="__codelineno-16-5" name="__codelineno-16-5"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"version"</span><span class="p">]</span>          <span class="c1"># VCF version</span>
</span><span id="__span-16-6"><a href="#__codelineno-16-6" id="__codelineno-16-6" name="__codelineno-16-6"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"sample_names"</span><span class="p">]</span>     <span class="c1"># Sample names</span>
</span><span id="__span-16-7"><a href="#__codelineno-16-7" id="__codelineno-16-7" name="__codelineno-16-7"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"info_fields"</span><span class="p">]</span>      <span class="c1"># INFO field definitions</span>
</span><span id="__span-16-8"><a href="#__codelineno-16-8" id="__codelineno-16-8" name="__codelineno-16-8"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"format_fields"</span><span class="p">]</span>    <span class="c1"># FORMAT field definitions</span>
</span><span id="__span-16-9"><a href="#__codelineno-16-9" id="__codelineno-16-9" name="__codelineno-16-9"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"contigs"</span><span class="p">]</span>          <span class="c1"># Contig definitions</span>
</span><span id="__span-16-10"><a href="#__codelineno-16-10" id="__codelineno-16-10" name="__codelineno-16-10"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"filters"</span><span class="p">]</span>          <span class="c1"># Filter definitions</span>
</span></code></pre></div>
</div>
<div class="tabbed-block">
<div class="language-python highlight"><pre><span></span><code><span id="__span-17-1"><a href="#__codelineno-17-1" id="__codelineno-17-1" name="__codelineno-17-1"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_fastq</span><span class="p">(</span><span class="s2">"reads.fastq.gz"</span><span class="p">)</span>
</span><span id="__span-17-2"><a href="#__codelineno-17-2" id="__codelineno-17-2" name="__codelineno-17-2"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-17-3"><a href="#__codelineno-17-3" id="__codelineno-17-3" name="__codelineno-17-3"></a>
</span><span id="__span-17-4"><a href="#__codelineno-17-4" id="__codelineno-17-4" name="__codelineno-17-4"></a><span class="c1"># FASTQ-specific metadata</span>
</span><span id="__span-17-5"><a href="#__codelineno-17-5" id="__codelineno-17-5" name="__codelineno-17-5"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"format"</span><span class="p">]</span>                     <span class="c1"># "fastq"</span>
</span><span id="__span-17-6"><a href="#__codelineno-17-6" id="__codelineno-17-6" name="__codelineno-17-6"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"path"</span><span class="p">]</span>                       <span class="c1"># "reads.fastq.gz"</span>
</span><span id="__span-17-7"><a href="#__codelineno-17-7" id="__codelineno-17-7" name="__codelineno-17-7"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"coordinate_system_zero_based"</span><span class="p">]</span> <span class="c1"># None (N/A for FASTQ)</span>
</span></code></pre></div>
</div>
<div class="tabbed-block">
<div class="language-python highlight"><pre><span></span><code><span id="__span-18-1"><a href="#__codelineno-18-1" id="__codelineno-18-1" name="__codelineno-18-1"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_bed</span><span class="p">(</span><span class="s2">"regions.bed"</span><span class="p">)</span>
</span><span id="__span-18-2"><a href="#__codelineno-18-2" id="__codelineno-18-2" name="__codelineno-18-2"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-18-3"><a href="#__codelineno-18-3" id="__codelineno-18-3" name="__codelineno-18-3"></a>
</span><span id="__span-18-4"><a href="#__codelineno-18-4" id="__codelineno-18-4" name="__codelineno-18-4"></a><span class="c1"># Basic metadata</span>
</span><span id="__span-18-5"><a href="#__codelineno-18-5" id="__codelineno-18-5" name="__codelineno-18-5"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"format"</span><span class="p">]</span>                     <span class="c1"># "bed"</span>
</span><span id="__span-18-6"><a href="#__codelineno-18-6" id="__codelineno-18-6" name="__codelineno-18-6"></a><span class="n">meta</span><span class="p">[</span><span class="s2">"coordinate_system_zero_based"</span><span class="p">]</span> <span class="c1"># True (0-based)</span>
</span></code></pre></div>
</div>
</div>
</div>
<h3 id="setting-custom-metadata">Setting Custom Metadata<a class="headerlink" href="#setting-custom-metadata" title="Permanent link"></a></h3>
<p>You can set metadata on DataFrames created from other sources:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-19-1"><a href="#__codelineno-19-1" id="__codelineno-19-1" name="__codelineno-19-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
</span><span id="__span-19-2"><a href="#__codelineno-19-2" id="__codelineno-19-2" name="__codelineno-19-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-19-3"><a href="#__codelineno-19-3" id="__codelineno-19-3" name="__codelineno-19-3"></a>
</span><span id="__span-19-4"><a href="#__codelineno-19-4" id="__codelineno-19-4" name="__codelineno-19-4"></a><span class="c1"># Create a DataFrame</span>
</span><span id="__span-19-5"><a href="#__codelineno-19-5" id="__codelineno-19-5" name="__codelineno-19-5"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pl</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span>
</span><span id="__span-19-6"><a href="#__codelineno-19-6" id="__codelineno-19-6" name="__codelineno-19-6"></a>    <span class="s2">"chrom"</span><span class="p">:</span> <span class="p">[</span><span class="s2">"chr1"</span><span class="p">,</span> <span class="s2">"chr1"</span><span class="p">],</span>
</span><span id="__span-19-7"><a href="#__codelineno-19-7" id="__codelineno-19-7" name="__codelineno-19-7"></a>    <span class="s2">"start"</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span>
</span><span id="__span-19-8"><a href="#__codelineno-19-8" id="__codelineno-19-8" name="__codelineno-19-8"></a>    <span class="s2">"end"</span><span class="p">:</span> <span class="p">[</span><span class="mi">150</span><span class="p">,</span> <span class="mi">250</span><span class="p">]</span>
</span><span id="__span-19-9"><a href="#__codelineno-19-9" id="__codelineno-19-9" name="__codelineno-19-9"></a><span class="p">})</span><span class="o">.</span><span class="n">lazy</span><span class="p">()</span>
</span><span id="__span-19-10"><a href="#__codelineno-19-10" id="__codelineno-19-10" name="__codelineno-19-10"></a>
</span><span id="__span-19-11"><a href="#__codelineno-19-11" id="__codelineno-19-11" name="__codelineno-19-11"></a><span class="c1"># Set metadata</span>
</span><span id="__span-19-12"><a href="#__codelineno-19-12" id="__codelineno-19-12" name="__codelineno-19-12"></a><span class="n">pb</span><span class="o">.</span><span class="n">set_source_metadata</span><span class="p">(</span>
</span><span id="__span-19-13"><a href="#__codelineno-19-13" id="__codelineno-19-13" name="__codelineno-19-13"></a>    <span class="n">df</span><span class="p">,</span>
</span><span id="__span-19-14"><a href="#__codelineno-19-14" id="__codelineno-19-14" name="__codelineno-19-14"></a>    <span class="nb">format</span><span class="o">=</span><span class="s2">"bed"</span><span class="p">,</span>
</span><span id="__span-19-15"><a href="#__codelineno-19-15" id="__codelineno-19-15" name="__codelineno-19-15"></a>    <span class="n">path</span><span class="o">=</span><span class="s2">"custom.bed"</span><span class="p">,</span>
</span><span id="__span-19-16"><a href="#__codelineno-19-16" id="__codelineno-19-16" name="__codelineno-19-16"></a>    <span class="n">header</span><span class="o">=</span><span class="p">{</span><span class="s2">"description"</span><span class="p">:</span> <span class="s2">"Custom intervals"</span><span class="p">}</span>
</span><span id="__span-19-17"><a href="#__codelineno-19-17" id="__codelineno-19-17" name="__codelineno-19-17"></a><span class="p">)</span>
</span><span id="__span-19-18"><a href="#__codelineno-19-18" id="__codelineno-19-18" name="__codelineno-19-18"></a>
</span><span id="__span-19-19"><a href="#__codelineno-19-19" id="__codelineno-19-19" name="__codelineno-19-19"></a><span class="c1"># Now metadata is available</span>
</span><span id="__span-19-20"><a href="#__codelineno-19-20" id="__codelineno-19-20" name="__codelineno-19-20"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
</span><span id="__span-19-21"><a href="#__codelineno-19-21" id="__codelineno-19-21" name="__codelineno-19-21"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"format"</span><span class="p">])</span>  <span class="c1"># "bed"</span>
</span><span id="__span-19-22"><a href="#__codelineno-19-22" id="__codelineno-19-22" name="__codelineno-19-22"></a><span class="nb">print</span><span class="p">(</span><span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"description"</span><span class="p">])</span>  <span class="c1"># "Custom intervals"</span>
</span></code></pre></div>
<h3 id="metadata-preservation">Metadata Preservation<a class="headerlink" href="#metadata-preservation" title="Permanent link"></a></h3>
<p>Metadata is preserved through Polars operations:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-20-1"><a href="#__codelineno-20-1" id="__codelineno-20-1" name="__codelineno-20-1"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-20-2"><a href="#__codelineno-20-2" id="__codelineno-20-2" name="__codelineno-20-2"></a>
</span><span id="__span-20-3"><a href="#__codelineno-20-3" id="__codelineno-20-3" name="__codelineno-20-3"></a><span class="c1"># Metadata persists after operations</span>
</span><span id="__span-20-4"><a href="#__codelineno-20-4" id="__codelineno-20-4" name="__codelineno-20-4"></a><span class="n">filtered</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"qual"</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">30</span><span class="p">)</span>
</span><span id="__span-20-5"><a href="#__codelineno-20-5" id="__codelineno-20-5" name="__codelineno-20-5"></a><span class="n">selected</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="s2">"chrom"</span><span class="p">,</span> <span class="s2">"start"</span><span class="p">,</span> <span class="s2">"end"</span><span class="p">])</span>
</span><span id="__span-20-6"><a href="#__codelineno-20-6" id="__codelineno-20-6" name="__codelineno-20-6"></a><span class="n">limited</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
</span><span id="__span-20-7"><a href="#__codelineno-20-7" id="__codelineno-20-7" name="__codelineno-20-7"></a>
</span><span id="__span-20-8"><a href="#__codelineno-20-8" id="__codelineno-20-8" name="__codelineno-20-8"></a><span class="c1"># All have the same metadata</span>
</span><span id="__span-20-9"><a href="#__codelineno-20-9" id="__codelineno-20-9" name="__codelineno-20-9"></a><span class="n">meta1</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-20-10"><a href="#__codelineno-20-10" id="__codelineno-20-10" name="__codelineno-20-10"></a><span class="n">meta2</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">filtered</span><span class="p">)</span>
</span><span id="__span-20-11"><a href="#__codelineno-20-11" id="__codelineno-20-11" name="__codelineno-20-11"></a><span class="n">meta3</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">selected</span><span class="p">)</span>
</span><span id="__span-20-12"><a href="#__codelineno-20-12" id="__codelineno-20-12" name="__codelineno-20-12"></a>
</span><span id="__span-20-13"><a href="#__codelineno-20-13" id="__codelineno-20-13" name="__codelineno-20-13"></a><span class="k">assert</span> <span class="n">meta1</span><span class="p">[</span><span class="s2">"format"</span><span class="p">]</span> <span class="o">==</span> <span class="n">meta2</span><span class="p">[</span><span class="s2">"format"</span><span class="p">]</span> <span class="o">==</span> <span class="n">meta3</span><span class="p">[</span><span class="s2">"format"</span><span class="p">]</span>  <span class="c1"># All "vcf"</span>
</span></code></pre></div>
<h3 id="using-metadata-for-debugging">Using Metadata for Debugging<a class="headerlink" href="#using-metadata-for-debugging" title="Permanent link"></a></h3>
<p>The <code>_datafusion_table_name</code> field is useful for debugging DataFusion SQL queries:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-21-1"><a href="#__codelineno-21-1" id="__codelineno-21-1" name="__codelineno-21-1"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-21-2"><a href="#__codelineno-21-2" id="__codelineno-21-2" name="__codelineno-21-2"></a><span class="n">meta</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">get_metadata</span><span class="p">(</span><span class="n">lf</span><span class="p">)</span>
</span><span id="__span-21-3"><a href="#__codelineno-21-3" id="__codelineno-21-3" name="__codelineno-21-3"></a>
</span><span id="__span-21-4"><a href="#__codelineno-21-4" id="__codelineno-21-4" name="__codelineno-21-4"></a><span class="c1"># Get internal table name</span>
</span><span id="__span-21-5"><a href="#__codelineno-21-5" id="__codelineno-21-5" name="__codelineno-21-5"></a><span class="n">table_name</span> <span class="o">=</span> <span class="n">meta</span><span class="p">[</span><span class="s2">"header"</span><span class="p">][</span><span class="s2">"_datafusion_table_name"</span><span class="p">]</span>
</span><span id="__span-21-6"><a href="#__codelineno-21-6" id="__codelineno-21-6" name="__codelineno-21-6"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Table name: </span><span class="si">{</span><span class="n">table_name</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>  <span class="c1"># "variants"</span>
</span><span id="__span-21-7"><a href="#__codelineno-21-7" id="__codelineno-21-7" name="__codelineno-21-7"></a>
</span><span id="__span-21-8"><a href="#__codelineno-21-8" id="__codelineno-21-8" name="__codelineno-21-8"></a><span class="c1"># Use it in SQL queries for debugging</span>
</span><span id="__span-21-9"><a href="#__codelineno-21-9" id="__codelineno-21-9" name="__codelineno-21-9"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="sa">f</span><span class="s2">"SELECT COUNT(*) FROM </span><span class="si">{</span><span class="n">table_name</span><span class="si">}</span><span class="s2">"</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="api-reference">API Reference<a class="headerlink" href="#api-reference" title="Permanent link"></a></h3>
<table>
<thead>
<tr>
<th>Function</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../api/#polars_bio.get_metadata"><code>get_metadata(df)</code></a></td>
<td>Get all metadata as a dictionary</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.print_metadata_json"><code>print_metadata_json(df, indent=2)</code></a></td>
<td>Print metadata as formatted JSON</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.print_metadata_summary"><code>print_metadata_summary(df)</code></a></td>
<td>Print human-readable metadata summary</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.set_source_metadata"><code>set_source_metadata(df, format, path, header)</code></a></td>
<td>Set metadata on a DataFrame</td>
</tr>
</tbody>
</table>
<h2 id="file-formats-support">File formats support<a class="headerlink" href="#file-formats-support" title="Permanent link"></a></h2>
<p>For bioinformatic format there are always three methods available: <code>read_*</code> (eager), <code>scan_*</code> (lazy) and <code>register_*</code> that can be used to either read file into Polars DataFrame/LazyFrame or register it as a DataFusion table for further processing using SQL or builtin interval methods. In either case, local and or cloud storage files can be used as an input. Please refer to <a href="#cloud-storage">cloud storage</a> section for more details.</p>
<table>
<thead>
<tr>
<th>Format</th>
<th>Single-threaded</th>
<th>Parallel (indexed)</th>
<th>Limit pushdown</th>
<th>Predicate pushdown</th>
<th>Projection pushdown</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../api/#polars_bio.data_input.read_bed">BED</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>‚ùå</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>‚ùå</td>
<td>‚ùå</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_vcf">VCF</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/> (TBI/CSI)</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_bam">BAM</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/> (BAI/CSI)</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_cram">CRAM</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/> (CRAI)</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_fastq">FASTQ</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/> (GZI)</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>‚ùå</td>
<td>‚ùå</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_fasta">FASTA</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>‚ùå</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>‚ùå</td>
<td>‚ùå</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_gff">GFF3</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/> (TBI/CSI)</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_input.read_pairs">Pairs</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/> (TBI/CSI)</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
</tbody>
</table>
<h3 id="indexed-reads-predicate-pushdown">Indexed reads &amp; predicate pushdown<a class="headerlink" href="#indexed-reads-predicate-pushdown" title="Permanent link"></a></h3>
<p>When an index file is present alongside the data file (BAI/CSI for BAM, CRAI for CRAM, TBI/CSI for VCF, GFF, and Pairs), polars-bio can push genomic region filters down to the DataFusion execution layer. This enables <strong>index-based random access</strong> ‚Äî only the relevant genomic regions are read from disk, dramatically improving performance for selective queries on large files.</p>
<p>Index files are <strong>auto-discovered</strong> by convention. Predicate pushdown is <strong>enabled by default</strong> for BAM, CRAM, VCF, GFF, and Pairs formats ‚Äî no extra configuration is needed.</p>
<h4 id="supported-index-formats">Supported index formats<a class="headerlink" href="#supported-index-formats" title="Permanent link"></a></h4>
<table>
<thead>
<tr>
<th>Data Format</th>
<th>Index Formats</th>
<th>Naming Convention</th>
</tr>
</thead>
<tbody>
<tr>
<td>BAM</td>
<td>BAI, CSI</td>
<td><code>sample.bam.bai</code> or <code>sample.bai</code>, <code>sample.bam.csi</code></td>
</tr>
<tr>
<td>CRAM</td>
<td>CRAI</td>
<td><code>sample.cram.crai</code></td>
</tr>
<tr>
<td>VCF (bgzf)</td>
<td>TBI, CSI</td>
<td><code>sample.vcf.gz.tbi</code>, <code>sample.vcf.gz.csi</code></td>
</tr>
<tr>
<td>GFF (bgzf)</td>
<td>TBI, CSI</td>
<td><code>sample.gff.gz.tbi</code>, <code>sample.gff.gz.csi</code></td>
</tr>
<tr>
<td>Pairs (bgzf)</td>
<td>TBI, CSI</td>
<td><code>contacts.pairs.gz.tbi</code>, <code>contacts.pairs.gz.csi</code></td>
</tr>
<tr>
<td>FASTQ (bgzf)</td>
<td>GZI</td>
<td><code>sample.fastq.bgz.gzi</code></td>
</tr>
</tbody>
</table>
<h4 id="usage-with-the-scanread-api">Usage with the scan/read API<a class="headerlink" href="#usage-with-the-scanread-api" title="Permanent link"></a></h4>
<p>Simply use <code>.filter()</code> ‚Äî predicate pushdown is enabled by default for BAM, CRAM, VCF, GFF, and Pairs:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-22-1"><a href="#__codelineno-22-1" id="__codelineno-22-1" name="__codelineno-22-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
</span><span id="__span-22-2"><a href="#__codelineno-22-2" id="__codelineno-22-2" name="__codelineno-22-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-22-3"><a href="#__codelineno-22-3" id="__codelineno-22-3" name="__codelineno-22-3"></a>
</span><span id="__span-22-4"><a href="#__codelineno-22-4" id="__codelineno-22-4" name="__codelineno-22-4"></a><span class="c1"># Single chromosome filter ‚Äî only chr1 data is read from disk</span>
</span><span id="__span-22-5"><a href="#__codelineno-22-5" id="__codelineno-22-5" name="__codelineno-22-5"></a><span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-22-6"><a href="#__codelineno-22-6" id="__codelineno-22-6" name="__codelineno-22-6"></a>    <span class="n">pb</span><span class="o">.</span><span class="n">scan_bam</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">)</span>
</span><span id="__span-22-7"><a href="#__codelineno-22-7" id="__codelineno-22-7" name="__codelineno-22-7"></a>    <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"chrom"</span><span class="p">)</span> <span class="o">==</span> <span class="s2">"chr1"</span><span class="p">)</span>
</span><span id="__span-22-8"><a href="#__codelineno-22-8" id="__codelineno-22-8" name="__codelineno-22-8"></a>    <span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-22-9"><a href="#__codelineno-22-9" id="__codelineno-22-9" name="__codelineno-22-9"></a><span class="p">)</span>
</span><span id="__span-22-10"><a href="#__codelineno-22-10" id="__codelineno-22-10" name="__codelineno-22-10"></a>
</span><span id="__span-22-11"><a href="#__codelineno-22-11" id="__codelineno-22-11" name="__codelineno-22-11"></a><span class="c1"># Multi-chromosome filter</span>
</span><span id="__span-22-12"><a href="#__codelineno-22-12" id="__codelineno-22-12" name="__codelineno-22-12"></a><span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-22-13"><a href="#__codelineno-22-13" id="__codelineno-22-13" name="__codelineno-22-13"></a>    <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf.gz"</span><span class="p">)</span>
</span><span id="__span-22-14"><a href="#__codelineno-22-14" id="__codelineno-22-14" name="__codelineno-22-14"></a>    <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"chrom"</span><span class="p">)</span><span class="o">.</span><span class="n">is_in</span><span class="p">([</span><span class="s2">"chr21"</span><span class="p">,</span> <span class="s2">"chr22"</span><span class="p">]))</span>
</span><span id="__span-22-15"><a href="#__codelineno-22-15" id="__codelineno-22-15" name="__codelineno-22-15"></a>    <span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-22-16"><a href="#__codelineno-22-16" id="__codelineno-22-16" name="__codelineno-22-16"></a><span class="p">)</span>
</span><span id="__span-22-17"><a href="#__codelineno-22-17" id="__codelineno-22-17" name="__codelineno-22-17"></a>
</span><span id="__span-22-18"><a href="#__codelineno-22-18" id="__codelineno-22-18" name="__codelineno-22-18"></a><span class="c1"># Region query ‚Äî combines chromosome and coordinate filters</span>
</span><span id="__span-22-19"><a href="#__codelineno-22-19" id="__codelineno-22-19" name="__codelineno-22-19"></a><span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-22-20"><a href="#__codelineno-22-20" id="__codelineno-22-20" name="__codelineno-22-20"></a>    <span class="n">pb</span><span class="o">.</span><span class="n">scan_bam</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">)</span>
</span><span id="__span-22-21"><a href="#__codelineno-22-21" id="__codelineno-22-21" name="__codelineno-22-21"></a>    <span class="o">.</span><span class="n">filter</span><span class="p">(</span>
</span><span id="__span-22-22"><a href="#__codelineno-22-22" id="__codelineno-22-22" name="__codelineno-22-22"></a>        <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"chrom"</span><span class="p">)</span> <span class="o">==</span> <span class="s2">"chr1"</span><span class="p">)</span>
</span><span id="__span-22-23"><a href="#__codelineno-22-23" id="__codelineno-22-23" name="__codelineno-22-23"></a>        <span class="o">&amp;</span> <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"start"</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">10000</span><span class="p">)</span>
</span><span id="__span-22-24"><a href="#__codelineno-22-24" id="__codelineno-22-24" name="__codelineno-22-24"></a>        <span class="o">&amp;</span> <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"end"</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">50000</span><span class="p">)</span>
</span><span id="__span-22-25"><a href="#__codelineno-22-25" id="__codelineno-22-25" name="__codelineno-22-25"></a>    <span class="p">)</span>
</span><span id="__span-22-26"><a href="#__codelineno-22-26" id="__codelineno-22-26" name="__codelineno-22-26"></a>    <span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-22-27"><a href="#__codelineno-22-27" id="__codelineno-22-27" name="__codelineno-22-27"></a><span class="p">)</span>
</span><span id="__span-22-28"><a href="#__codelineno-22-28" id="__codelineno-22-28" name="__codelineno-22-28"></a>
</span><span id="__span-22-29"><a href="#__codelineno-22-29" id="__codelineno-22-29" name="__codelineno-22-29"></a><span class="c1"># CRAM with predicate pushdown</span>
</span><span id="__span-22-30"><a href="#__codelineno-22-30" id="__codelineno-22-30" name="__codelineno-22-30"></a><span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-22-31"><a href="#__codelineno-22-31" id="__codelineno-22-31" name="__codelineno-22-31"></a>    <span class="n">pb</span><span class="o">.</span><span class="n">scan_cram</span><span class="p">(</span><span class="s2">"alignments.cram"</span><span class="p">)</span>
</span><span id="__span-22-32"><a href="#__codelineno-22-32" id="__codelineno-22-32" name="__codelineno-22-32"></a>    <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"chrom"</span><span class="p">)</span> <span class="o">==</span> <span class="s2">"chr1"</span><span class="p">)</span>
</span><span id="__span-22-33"><a href="#__codelineno-22-33" id="__codelineno-22-33" name="__codelineno-22-33"></a>    <span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-22-34"><a href="#__codelineno-22-34" id="__codelineno-22-34" name="__codelineno-22-34"></a><span class="p">)</span>
</span></code></pre></div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>Predicate pushdown supports: equality (<code>==</code>), comparisons (<code>&gt;=</code>, <code>&lt;=</code>, <code>&gt;</code>, <code>&lt;</code>), <code>is_in()</code>, <code>is_null()</code>, <code>is_not_null()</code>, and combinations with <code>&amp;</code> (AND). Complex predicates like <code>.str.contains()</code> or OR logic are automatically filtered client-side. To disable pushdown, pass <code>predicate_pushdown=False</code>.</p>
</div>
<h4 id="usage-with-the-sql-api">Usage with the SQL API<a class="headerlink" href="#usage-with-the-sql-api" title="Permanent link"></a></h4>
<p>The SQL path works automatically ‚Äî DataFusion parses the WHERE clause and uses the index without any extra flags:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-23-1"><a href="#__codelineno-23-1" id="__codelineno-23-1" name="__codelineno-23-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-23-2"><a href="#__codelineno-23-2" id="__codelineno-23-2" name="__codelineno-23-2"></a>
</span><span id="__span-23-3"><a href="#__codelineno-23-3" id="__codelineno-23-3" name="__codelineno-23-3"></a><span class="n">pb</span><span class="o">.</span><span class="n">register_bam</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">,</span> <span class="s2">"reads"</span><span class="p">)</span>
</span><span id="__span-23-4"><a href="#__codelineno-23-4" id="__codelineno-23-4" name="__codelineno-23-4"></a>
</span><span id="__span-23-5"><a href="#__codelineno-23-5" id="__codelineno-23-5" name="__codelineno-23-5"></a><span class="c1"># Single chromosome</span>
</span><span id="__span-23-6"><a href="#__codelineno-23-6" id="__codelineno-23-6" name="__codelineno-23-6"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="s2">"SELECT * FROM reads WHERE chrom = 'chr1'"</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-23-7"><a href="#__codelineno-23-7" id="__codelineno-23-7" name="__codelineno-23-7"></a>
</span><span id="__span-23-8"><a href="#__codelineno-23-8" id="__codelineno-23-8" name="__codelineno-23-8"></a><span class="c1"># Region query</span>
</span><span id="__span-23-9"><a href="#__codelineno-23-9" id="__codelineno-23-9" name="__codelineno-23-9"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span>
</span><span id="__span-23-10"><a href="#__codelineno-23-10" id="__codelineno-23-10" name="__codelineno-23-10"></a>    <span class="s2">"SELECT * FROM reads WHERE chrom = 'chr1' AND start &gt;= 10000 AND </span><span class="se">\"</span><span class="s2">end</span><span class="se">\"</span><span class="s2"> &lt;= 50000"</span>
</span><span id="__span-23-11"><a href="#__codelineno-23-11" id="__codelineno-23-11" name="__codelineno-23-11"></a><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-23-12"><a href="#__codelineno-23-12" id="__codelineno-23-12" name="__codelineno-23-12"></a>
</span><span id="__span-23-13"><a href="#__codelineno-23-13" id="__codelineno-23-13" name="__codelineno-23-13"></a><span class="c1"># Combined genomic and record filters</span>
</span><span id="__span-23-14"><a href="#__codelineno-23-14" id="__codelineno-23-14" name="__codelineno-23-14"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span>
</span><span id="__span-23-15"><a href="#__codelineno-23-15" id="__codelineno-23-15" name="__codelineno-23-15"></a>    <span class="s2">"SELECT * FROM reads WHERE chrom = 'chr1' AND mapping_quality &gt;= 30"</span>
</span><span id="__span-23-16"><a href="#__codelineno-23-16" id="__codelineno-23-16" name="__codelineno-23-16"></a><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span></code></pre></div>
<h4 id="automatic-parallel-partitioning">Automatic parallel partitioning<a class="headerlink" href="#automatic-parallel-partitioning" title="Permanent link"></a></h4>
<p>When an index file is present, DataFusion distributes genomic regions across balanced partitions using index-derived size estimates, enabling parallel
execution. Formats with known contig lengths (BAM, CRAM) can split large regions into sub-regions for full parallelism even on single-chromosome queries. For FASTQ files, a <strong>GZI index</strong> alongside a BGZF-compressed file enables parallel decoding of compressed blocks. This is controlled by the global <code>target_partitions</code> setting:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-24-1"><a href="#__codelineno-24-1" id="__codelineno-24-1" name="__codelineno-24-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-24-2"><a href="#__codelineno-24-2" id="__codelineno-24-2" name="__codelineno-24-2"></a>
</span><span id="__span-24-3"><a href="#__codelineno-24-3" id="__codelineno-24-3" name="__codelineno-24-3"></a><span class="n">pb</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">"datafusion.execution.target_partitions"</span><span class="p">,</span> <span class="s2">"8"</span><span class="p">)</span>
</span><span id="__span-24-4"><a href="#__codelineno-24-4" id="__codelineno-24-4" name="__codelineno-24-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="s2">"large_file.bam"</span><span class="p">)</span>  <span class="c1"># 8 partitions will be used for parallel execution</span>
</span><span id="__span-24-5"><a href="#__codelineno-24-5" id="__codelineno-24-5" name="__codelineno-24-5"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_fastq</span><span class="p">(</span><span class="s2">"reads.fastq.bgz"</span><span class="p">)</span>  <span class="c1"># parallel BGZF decoding when .gzi index is present</span>
</span></code></pre></div>
<p><strong>Partitioning behavior (BAM, CRAM, VCF, GFF):</strong></p>
<table>
<thead>
<tr>
<th>Index Available?</th>
<th>SQL Filters</th>
<th>Partitions</th>
</tr>
</thead>
<tbody>
<tr>
<td>Yes</td>
<td><code>chrom = 'chr1' AND start &gt;= 1000</code></td>
<td>up to target_partitions (region split into sub-regions)</td>
</tr>
<tr>
<td>Yes</td>
<td><code>chrom IN ('chr1', 'chr2')</code></td>
<td>up to target_partitions (both regions split to fill bins)</td>
</tr>
<tr>
<td>Yes</td>
<td><code>mapping_quality &gt;= 30</code> (no genomic filter)</td>
<td>up to target_partitions (all chroms balanced + split)</td>
</tr>
<tr>
<td>Yes</td>
<td>None (full scan)</td>
<td>up to target_partitions (all chroms balanced + split)</td>
</tr>
<tr>
<td>No</td>
<td>Any</td>
<td>1 (sequential full scan)</td>
</tr>
</tbody>
</table>
<p><strong>Partitioning behavior (FASTQ):</strong></p>
<table>
<thead>
<tr>
<th>File type</th>
<th>GZI Index?</th>
<th>Partitions</th>
</tr>
</thead>
<tbody>
<tr>
<td>BGZF (<code>.fastq.bgz</code>)</td>
<td>Yes (<code>.fastq.bgz.gzi</code>)</td>
<td>up to target_partitions (parallel block decoding)</td>
</tr>
<tr>
<td>BGZF (<code>.fastq.bgz</code>)</td>
<td>No</td>
<td>1 (sequential read)</td>
</tr>
<tr>
<td>GZIP (<code>.fastq.gz</code>)</td>
<td>N/A</td>
<td>1 (sequential ‚Äî GZIP cannot be parallelized)</td>
</tr>
<tr>
<td>Uncompressed (<code>.fastq</code>)</td>
<td>N/A</td>
<td>up to target_partitions (byte-range parallel)</td>
</tr>
</tbody>
</table>
<h4 id="record-level-filter-pushdown">Record-level filter pushdown<a class="headerlink" href="#record-level-filter-pushdown" title="Permanent link"></a></h4>
<p>All formats support record-level predicate evaluation ‚Äî filters on columns like <code>mapping_quality</code>, <code>flag</code>, or <code>strand</code> are evaluated per-record during scan, with or without an index file. See the <a href="../developers/#predicate-pushdown">Developers Guide</a> for the translation pipeline internals and <a href="../developers/#predicate-projection-pushdown-examples">examples</a>.</p>
<h4 id="projection-pushdown">Projection pushdown<a class="headerlink" href="#projection-pushdown" title="Permanent link"></a></h4>
<p>BAM, CRAM, VCF, and Pairs formats support parsing-level projection pushdown ‚Äî unprojected fields are skipped entirely during record parsing. Enabled by default (<code>projection_pushdown=True</code>). See the <a href="../developers/#projection-pushdown">Developers Guide</a> for internals and <a href="../developers/#inspecting-the-execution-plan">execution plan inspection</a>.</p>
<h4 id="index-file-generation">Index file generation<a class="headerlink" href="#index-file-generation" title="Permanent link"></a></h4>
<div class="admonition tip">
<p class="admonition-title">Creating index files</p>
<p>Create index files using standard bioinformatics tools:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-25-1"><a href="#__codelineno-25-1" id="__codelineno-25-1" name="__codelineno-25-1"></a><span class="c1"># BAM: sort and index</span>
</span><span id="__span-25-2"><a href="#__codelineno-25-2" id="__codelineno-25-2" name="__codelineno-25-2"></a>samtools<span class="w"> </span>sort<span class="w"> </span>input.bam<span class="w"> </span>-o<span class="w"> </span>sorted.bam
</span><span id="__span-25-3"><a href="#__codelineno-25-3" id="__codelineno-25-3" name="__codelineno-25-3"></a>samtools<span class="w"> </span>index<span class="w"> </span>sorted.bam<span class="w">                </span><span class="c1"># creates sorted.bam.bai</span>
</span><span id="__span-25-4"><a href="#__codelineno-25-4" id="__codelineno-25-4" name="__codelineno-25-4"></a>
</span><span id="__span-25-5"><a href="#__codelineno-25-5" id="__codelineno-25-5" name="__codelineno-25-5"></a><span class="c1"># CRAM: sort and index</span>
</span><span id="__span-25-6"><a href="#__codelineno-25-6" id="__codelineno-25-6" name="__codelineno-25-6"></a>samtools<span class="w"> </span>sort<span class="w"> </span>input.cram<span class="w"> </span>-o<span class="w"> </span>sorted.cram<span class="w"> </span>--reference<span class="w"> </span>ref.fa
</span><span id="__span-25-7"><a href="#__codelineno-25-7" id="__codelineno-25-7" name="__codelineno-25-7"></a>samtools<span class="w"> </span>index<span class="w"> </span>sorted.cram<span class="w">               </span><span class="c1"># creates sorted.cram.crai</span>
</span><span id="__span-25-8"><a href="#__codelineno-25-8" id="__codelineno-25-8" name="__codelineno-25-8"></a>
</span><span id="__span-25-9"><a href="#__codelineno-25-9" id="__codelineno-25-9" name="__codelineno-25-9"></a><span class="c1"># VCF: sort, compress, and index</span>
</span><span id="__span-25-10"><a href="#__codelineno-25-10" id="__codelineno-25-10" name="__codelineno-25-10"></a>bcftools<span class="w"> </span>sort<span class="w"> </span>input.vcf<span class="w"> </span>-Oz<span class="w"> </span>-o<span class="w"> </span>sorted.vcf.gz
</span><span id="__span-25-11"><a href="#__codelineno-25-11" id="__codelineno-25-11" name="__codelineno-25-11"></a>bcftools<span class="w"> </span>index<span class="w"> </span>-t<span class="w"> </span>sorted.vcf.gz<span class="w">          </span><span class="c1"># creates sorted.vcf.gz.tbi</span>
</span><span id="__span-25-12"><a href="#__codelineno-25-12" id="__codelineno-25-12" name="__codelineno-25-12"></a>
</span><span id="__span-25-13"><a href="#__codelineno-25-13" id="__codelineno-25-13" name="__codelineno-25-13"></a><span class="c1"># GFF: sort, compress, and index</span>
</span><span id="__span-25-14"><a href="#__codelineno-25-14" id="__codelineno-25-14" name="__codelineno-25-14"></a><span class="o">(</span>grep<span class="w"> </span><span class="s2">"^#"</span><span class="w"> </span>input.gff<span class="p">;</span><span class="w"> </span>grep<span class="w"> </span>-v<span class="w"> </span><span class="s2">"^#"</span><span class="w"> </span>input.gff<span class="w"> </span><span class="p">|</span><span class="w"> </span>sort<span class="w"> </span>-k1,1<span class="w"> </span>-k4,4n<span class="o">)</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>bgzip<span class="w"> </span>&gt;<span class="w"> </span>sorted.gff.gz
</span><span id="__span-25-15"><a href="#__codelineno-25-15" id="__codelineno-25-15" name="__codelineno-25-15"></a>tabix<span class="w"> </span>-p<span class="w"> </span>gff<span class="w"> </span>sorted.gff.gz<span class="w">               </span><span class="c1"># creates sorted.gff.gz.tbi</span>
</span><span id="__span-25-16"><a href="#__codelineno-25-16" id="__codelineno-25-16" name="__codelineno-25-16"></a>
</span><span id="__span-25-17"><a href="#__codelineno-25-17" id="__codelineno-25-17" name="__codelineno-25-17"></a><span class="c1"># Pairs: sort, compress, and index (col 2=chr1, col 3=pos1)</span>
</span><span id="__span-25-18"><a href="#__codelineno-25-18" id="__codelineno-25-18" name="__codelineno-25-18"></a>sort<span class="w"> </span>-k2,2<span class="w"> </span>-k3,3n<span class="w"> </span>contacts.pairs<span class="w"> </span><span class="p">|</span><span class="w"> </span>bgzip<span class="w"> </span>&gt;<span class="w"> </span>contacts.pairs.gz
</span><span id="__span-25-19"><a href="#__codelineno-25-19" id="__codelineno-25-19" name="__codelineno-25-19"></a>tabix<span class="w"> </span>-s<span class="w"> </span><span class="m">2</span><span class="w"> </span>-b<span class="w"> </span><span class="m">3</span><span class="w"> </span>-e<span class="w"> </span><span class="m">3</span><span class="w"> </span>contacts.pairs.gz<span class="w">   </span><span class="c1"># creates contacts.pairs.gz.tbi</span>
</span><span id="__span-25-20"><a href="#__codelineno-25-20" id="__codelineno-25-20" name="__codelineno-25-20"></a>
</span><span id="__span-25-21"><a href="#__codelineno-25-21" id="__codelineno-25-21" name="__codelineno-25-21"></a><span class="c1"># FASTQ: BGZF compress and create GZI index for parallel reads</span>
</span><span id="__span-25-22"><a href="#__codelineno-25-22" id="__codelineno-25-22" name="__codelineno-25-22"></a>bgzip<span class="w"> </span>reads.fastq<span class="w">                         </span><span class="c1"># creates reads.fastq.bgz</span>
</span><span id="__span-25-23"><a href="#__codelineno-25-23" id="__codelineno-25-23" name="__codelineno-25-23"></a>bgzip<span class="w"> </span>-r<span class="w"> </span>reads.fastq.bgz<span class="w">                 </span><span class="c1"># creates reads.fastq.bgz.gzi</span>
</span></code></pre></div>
</div>
<h2 id="file-output">File Output<a class="headerlink" href="#file-output" title="Permanent link"></a></h2>
<p>polars-bio supports writing DataFrames back to bioinformatic file formats. Two methods are available for each supported format:</p>
<ul>
<li><code>write_*</code> - Eager write that collects the DataFrame and writes it to disk, returns row count</li>
<li><code>sink_*</code> - Streaming write for LazyFrames that processes data in batches without full materialization</li>
</ul>
<h3 id="output-format-support">Output Format Support<a class="headerlink" href="#output-format-support" title="Permanent link"></a></h3>
<table>
<thead>
<tr>
<th>Format</th>
<th>write_*</th>
<th>sink_*</th>
<th>Compression</th>
<th>Notes</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="../api/#polars_bio.data_output.write_vcf">VCF</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><code>.vcf.gz</code>, <code>.vcf.bgz</code></td>
<td>Auto-detected from extension</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_output.write_bam">BAM</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>BGZF (built-in)</td>
<td>Binary alignment format</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_output.write_sam">SAM</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>None</td>
<td>Plain text alignment format</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_output.write_cram">CRAM</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td>Built-in</td>
<td>Requires reference FASTA</td>
</tr>
<tr>
<td><a href="../api/#polars_bio.data_output.write_fastq">FASTQ</a></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><code>.fastq.gz</code>, <code>.fastq.bgz</code></td>
<td>Auto-detected from extension</td>
</tr>
</tbody>
</table>
<h3 id="basic-usage">Basic Usage<a class="headerlink" href="#basic-usage" title="Permanent link"></a></h3>
<div class="language-python highlight"><pre><span></span><code><span id="__span-26-1"><a href="#__codelineno-26-1" id="__codelineno-26-1" name="__codelineno-26-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-26-2"><a href="#__codelineno-26-2" id="__codelineno-26-2" name="__codelineno-26-2"></a>
</span><span id="__span-26-3"><a href="#__codelineno-26-3" id="__codelineno-26-3" name="__codelineno-26-3"></a><span class="c1"># Read, transform, and write back</span>
</span><span id="__span-26-4"><a href="#__codelineno-26-4" id="__codelineno-26-4" name="__codelineno-26-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="s2">"input.bam"</span><span class="p">,</span> <span class="n">tag_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"NM"</span><span class="p">,</span> <span class="s2">"AS"</span><span class="p">])</span>
</span><span id="__span-26-5"><a href="#__codelineno-26-5" id="__codelineno-26-5" name="__codelineno-26-5"></a><span class="n">filtered</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"mapping_quality"</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span><span class="p">)</span>
</span><span id="__span-26-6"><a href="#__codelineno-26-6" id="__codelineno-26-6" name="__codelineno-26-6"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_bam</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span> <span class="s2">"output.bam"</span><span class="p">)</span>
</span><span id="__span-26-7"><a href="#__codelineno-26-7" id="__codelineno-26-7" name="__codelineno-26-7"></a>
</span><span id="__span-26-8"><a href="#__codelineno-26-8" id="__codelineno-26-8" name="__codelineno-26-8"></a><span class="c1"># Streaming write with LazyFrame</span>
</span><span id="__span-26-9"><a href="#__codelineno-26-9" id="__codelineno-26-9" name="__codelineno-26-9"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-26-10"><a href="#__codelineno-26-10" id="__codelineno-26-10" name="__codelineno-26-10"></a><span class="n">pb</span><span class="o">.</span><span class="n">sink_vcf</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"qual"</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">30</span><span class="p">),</span> <span class="s2">"filtered.vcf.bgz"</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="sorted-output-with-sort_on_write">Sorted Output with <code>sort_on_write</code><a class="headerlink" href="#sorted-output-with-sort_on_write" title="Permanent link"></a></h3>
<p>BAM, SAM, and CRAM write functions support the <code>sort_on_write</code> parameter to produce coordinate-sorted output:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-27-1"><a href="#__codelineno-27-1" id="__codelineno-27-1" name="__codelineno-27-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-27-2"><a href="#__codelineno-27-2" id="__codelineno-27-2" name="__codelineno-27-2"></a>
</span><span id="__span-27-3"><a href="#__codelineno-27-3" id="__codelineno-27-3" name="__codelineno-27-3"></a><span class="c1"># Write coordinate-sorted BAM</span>
</span><span id="__span-27-4"><a href="#__codelineno-27-4" id="__codelineno-27-4" name="__codelineno-27-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="s2">"unsorted.bam"</span><span class="p">)</span>
</span><span id="__span-27-5"><a href="#__codelineno-27-5" id="__codelineno-27-5" name="__codelineno-27-5"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_bam</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"sorted.bam"</span><span class="p">,</span> <span class="n">sort_on_write</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-27-6"><a href="#__codelineno-27-6" id="__codelineno-27-6" name="__codelineno-27-6"></a>
</span><span id="__span-27-7"><a href="#__codelineno-27-7" id="__codelineno-27-7" name="__codelineno-27-7"></a><span class="c1"># Streaming sorted write</span>
</span><span id="__span-27-8"><a href="#__codelineno-27-8" id="__codelineno-27-8" name="__codelineno-27-8"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_sam</span><span class="p">(</span><span class="s2">"input.sam"</span><span class="p">)</span>
</span><span id="__span-27-9"><a href="#__codelineno-27-9" id="__codelineno-27-9" name="__codelineno-27-9"></a><span class="n">pb</span><span class="o">.</span><span class="n">sink_bam</span><span class="p">(</span><span class="n">lf</span><span class="p">,</span> <span class="s2">"sorted.bam"</span><span class="p">,</span> <span class="n">sort_on_write</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span></code></pre></div>
<p>When <code>sort_on_write=True</code>:</p>
<ul>
<li>Records are sorted by <code>(chrom ASC, start ASC)</code> during write</li>
<li>Output header contains <code>@HD ... SO:coordinate</code></li>
</ul>
<p>When <code>sort_on_write=False</code> (default):</p>
<ul>
<li>Records are written in input order</li>
<li>Output header contains <code>@HD ... SO:unsorted</code></li>
</ul>
<h3 id="cram-output">CRAM Output<a class="headerlink" href="#cram-output" title="Permanent link"></a></h3>
<p>CRAM format requires a reference FASTA file for writing:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-28-1"><a href="#__codelineno-28-1" id="__codelineno-28-1" name="__codelineno-28-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-28-2"><a href="#__codelineno-28-2" id="__codelineno-28-2" name="__codelineno-28-2"></a>
</span><span id="__span-28-3"><a href="#__codelineno-28-3" id="__codelineno-28-3" name="__codelineno-28-3"></a><span class="c1"># CRAM write requires reference_path</span>
</span><span id="__span-28-4"><a href="#__codelineno-28-4" id="__codelineno-28-4" name="__codelineno-28-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_cram</span><span class="p">(</span><span class="s2">"input.cram"</span><span class="p">,</span> <span class="n">reference_path</span><span class="o">=</span><span class="s2">"reference.fa"</span><span class="p">)</span>
</span><span id="__span-28-5"><a href="#__codelineno-28-5" id="__codelineno-28-5" name="__codelineno-28-5"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_cram</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.cram"</span><span class="p">,</span> <span class="n">reference_path</span><span class="o">=</span><span class="s2">"reference.fa"</span><span class="p">)</span>
</span><span id="__span-28-6"><a href="#__codelineno-28-6" id="__codelineno-28-6" name="__codelineno-28-6"></a>
</span><span id="__span-28-7"><a href="#__codelineno-28-7" id="__codelineno-28-7" name="__codelineno-28-7"></a><span class="c1"># Streaming CRAM write</span>
</span><span id="__span-28-8"><a href="#__codelineno-28-8" id="__codelineno-28-8" name="__codelineno-28-8"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_cram</span><span class="p">(</span><span class="s2">"input.cram"</span><span class="p">,</span> <span class="n">reference_path</span><span class="o">=</span><span class="s2">"reference.fa"</span><span class="p">)</span>
</span><span id="__span-28-9"><a href="#__codelineno-28-9" id="__codelineno-28-9" name="__codelineno-28-9"></a><span class="n">pb</span><span class="o">.</span><span class="n">sink_cram</span><span class="p">(</span><span class="n">lf</span><span class="p">,</span> <span class="s2">"output.cram"</span><span class="p">,</span> <span class="n">reference_path</span><span class="o">=</span><span class="s2">"reference.fa"</span><span class="p">,</span> <span class="n">sort_on_write</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span></code></pre></div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <code>reference_path</code> parameter is <strong>required</strong> for <code>write_cram()</code> and <code>sink_cram()</code>. Attempting to write CRAM without a reference will raise an error.</p>
</div>
<h3 id="output-compression">Output Compression<a class="headerlink" href="#output-compression" title="Permanent link"></a></h3>
<p>Output compression is auto-detected from the file extension for VCF and FASTQ formats:</p>
<table>
<thead>
<tr>
<th>Extension</th>
<th>Compression</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>.vcf</code> / <code>.fastq</code></td>
<td>None (plain text)</td>
</tr>
<tr>
<td><code>.vcf.gz</code> / <code>.fastq.gz</code></td>
<td>GZIP</td>
</tr>
<tr>
<td><code>.vcf.bgz</code> / <code>.fastq.bgz</code></td>
<td>BGZF (block gzip)</td>
</tr>
</tbody>
</table>
<div class="language-python highlight"><pre><span></span><code><span id="__span-29-1"><a href="#__codelineno-29-1" id="__codelineno-29-1" name="__codelineno-29-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-29-2"><a href="#__codelineno-29-2" id="__codelineno-29-2" name="__codelineno-29-2"></a>
</span><span id="__span-29-3"><a href="#__codelineno-29-3" id="__codelineno-29-3" name="__codelineno-29-3"></a><span class="c1"># VCF</span>
</span><span id="__span-29-4"><a href="#__codelineno-29-4" id="__codelineno-29-4" name="__codelineno-29-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">)</span>
</span><span id="__span-29-5"><a href="#__codelineno-29-5" id="__codelineno-29-5" name="__codelineno-29-5"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.vcf"</span><span class="p">)</span>        <span class="c1"># plain text</span>
</span><span id="__span-29-6"><a href="#__codelineno-29-6" id="__codelineno-29-6" name="__codelineno-29-6"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.vcf.gz"</span><span class="p">)</span>     <span class="c1"># GZIP</span>
</span><span id="__span-29-7"><a href="#__codelineno-29-7" id="__codelineno-29-7" name="__codelineno-29-7"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.vcf.bgz"</span><span class="p">)</span>    <span class="c1"># BGZF (recommended for indexing)</span>
</span><span id="__span-29-8"><a href="#__codelineno-29-8" id="__codelineno-29-8" name="__codelineno-29-8"></a>
</span><span id="__span-29-9"><a href="#__codelineno-29-9" id="__codelineno-29-9" name="__codelineno-29-9"></a><span class="c1"># FASTQ</span>
</span><span id="__span-29-10"><a href="#__codelineno-29-10" id="__codelineno-29-10" name="__codelineno-29-10"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_fastq</span><span class="p">(</span><span class="s2">"reads.fastq"</span><span class="p">)</span>
</span><span id="__span-29-11"><a href="#__codelineno-29-11" id="__codelineno-29-11" name="__codelineno-29-11"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_fastq</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.fastq"</span><span class="p">)</span>       <span class="c1"># plain text</span>
</span><span id="__span-29-12"><a href="#__codelineno-29-12" id="__codelineno-29-12" name="__codelineno-29-12"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_fastq</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.fastq.gz"</span><span class="p">)</span>    <span class="c1"># GZIP</span>
</span><span id="__span-29-13"><a href="#__codelineno-29-13" id="__codelineno-29-13" name="__codelineno-29-13"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_fastq</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">"output.fastq.bgz"</span><span class="p">)</span>   <span class="c1"># BGZF (recommended for parallel reads with GZI index)</span>
</span><span id="__span-29-14"><a href="#__codelineno-29-14" id="__codelineno-29-14" name="__codelineno-29-14"></a>
</span><span id="__span-29-15"><a href="#__codelineno-29-15" id="__codelineno-29-15" name="__codelineno-29-15"></a><span class="c1"># Streaming write</span>
</span><span id="__span-29-16"><a href="#__codelineno-29-16" id="__codelineno-29-16" name="__codelineno-29-16"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_fastq</span><span class="p">(</span><span class="s2">"large_reads.fastq.gz"</span><span class="p">)</span>
</span><span id="__span-29-17"><a href="#__codelineno-29-17" id="__codelineno-29-17" name="__codelineno-29-17"></a><span class="n">pb</span><span class="o">.</span><span class="n">sink_fastq</span><span class="p">(</span><span class="n">lf</span><span class="o">.</span><span class="n">limit</span><span class="p">(</span><span class="mi">1000</span><span class="p">),</span> <span class="s2">"sample.fastq"</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="header-preservation">Header Preservation<a class="headerlink" href="#header-preservation" title="Permanent link"></a></h3>
<p>When reading and writing alignment files (BAM/SAM/CRAM), polars-bio preserves header metadata including:</p>
<ul>
<li><code>@SQ</code> (sequence dictionary)</li>
<li><code>@RG</code> (read groups)</li>
<li><code>@PG</code> (program records)</li>
</ul>
<p>This enables lossless roundtrip workflows:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-30-1"><a href="#__codelineno-30-1" id="__codelineno-30-1" name="__codelineno-30-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-30-2"><a href="#__codelineno-30-2" id="__codelineno-30-2" name="__codelineno-30-2"></a>
</span><span id="__span-30-3"><a href="#__codelineno-30-3" id="__codelineno-30-3" name="__codelineno-30-3"></a><span class="c1"># Read with full header preservation</span>
</span><span id="__span-30-4"><a href="#__codelineno-30-4" id="__codelineno-30-4" name="__codelineno-30-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="s2">"input.bam"</span><span class="p">)</span>
</span><span id="__span-30-5"><a href="#__codelineno-30-5" id="__codelineno-30-5" name="__codelineno-30-5"></a>
</span><span id="__span-30-6"><a href="#__codelineno-30-6" id="__codelineno-30-6" name="__codelineno-30-6"></a><span class="c1"># Filter records</span>
</span><span id="__span-30-7"><a href="#__codelineno-30-7" id="__codelineno-30-7" name="__codelineno-30-7"></a><span class="n">filtered</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"mapping_quality"</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span><span class="p">)</span>
</span><span id="__span-30-8"><a href="#__codelineno-30-8" id="__codelineno-30-8" name="__codelineno-30-8"></a>
</span><span id="__span-30-9"><a href="#__codelineno-30-9" id="__codelineno-30-9" name="__codelineno-30-9"></a><span class="c1"># Write back - header metadata is preserved</span>
</span><span id="__span-30-10"><a href="#__codelineno-30-10" id="__codelineno-30-10" name="__codelineno-30-10"></a><span class="n">pb</span><span class="o">.</span><span class="n">write_bam</span><span class="p">(</span><span class="n">filtered</span><span class="p">,</span> <span class="s2">"filtered.bam"</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="polars-extension-methods">Polars Extension Methods<a class="headerlink" href="#polars-extension-methods" title="Permanent link"></a></h3>
<p>Write functions are also available as Polars namespace extensions:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-31-1"><a href="#__codelineno-31-1" id="__codelineno-31-1" name="__codelineno-31-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-31-2"><a href="#__codelineno-31-2" id="__codelineno-31-2" name="__codelineno-31-2"></a>
</span><span id="__span-31-3"><a href="#__codelineno-31-3" id="__codelineno-31-3" name="__codelineno-31-3"></a><span class="c1"># DataFrame extensions</span>
</span><span id="__span-31-4"><a href="#__codelineno-31-4" id="__codelineno-31-4" name="__codelineno-31-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span><span class="s2">"input.bam"</span><span class="p">)</span>
</span><span id="__span-31-5"><a href="#__codelineno-31-5" id="__codelineno-31-5" name="__codelineno-31-5"></a><span class="n">df</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">write_bam</span><span class="p">(</span><span class="s2">"output.bam"</span><span class="p">,</span> <span class="n">sort_on_write</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-31-6"><a href="#__codelineno-31-6" id="__codelineno-31-6" name="__codelineno-31-6"></a><span class="n">df</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">write_sam</span><span class="p">(</span><span class="s2">"output.sam"</span><span class="p">)</span>
</span><span id="__span-31-7"><a href="#__codelineno-31-7" id="__codelineno-31-7" name="__codelineno-31-7"></a><span class="n">df</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">write_cram</span><span class="p">(</span><span class="s2">"output.cram"</span><span class="p">,</span> <span class="n">reference_path</span><span class="o">=</span><span class="s2">"ref.fa"</span><span class="p">)</span>
</span><span id="__span-31-8"><a href="#__codelineno-31-8" id="__codelineno-31-8" name="__codelineno-31-8"></a><span class="n">df</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">write_vcf</span><span class="p">(</span><span class="s2">"output.vcf.bgz"</span><span class="p">)</span>
</span><span id="__span-31-9"><a href="#__codelineno-31-9" id="__codelineno-31-9" name="__codelineno-31-9"></a><span class="n">df</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">write_fastq</span><span class="p">(</span><span class="s2">"output.fastq.gz"</span><span class="p">)</span>
</span><span id="__span-31-10"><a href="#__codelineno-31-10" id="__codelineno-31-10" name="__codelineno-31-10"></a>
</span><span id="__span-31-11"><a href="#__codelineno-31-11" id="__codelineno-31-11" name="__codelineno-31-11"></a><span class="c1"># LazyFrame extensions</span>
</span><span id="__span-31-12"><a href="#__codelineno-31-12" id="__codelineno-31-12" name="__codelineno-31-12"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_bam</span><span class="p">(</span><span class="s2">"input.bam"</span><span class="p">)</span>
</span><span id="__span-31-13"><a href="#__codelineno-31-13" id="__codelineno-31-13" name="__codelineno-31-13"></a><span class="n">lf</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">sink_bam</span><span class="p">(</span><span class="s2">"output.bam"</span><span class="p">,</span> <span class="n">sort_on_write</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-31-14"><a href="#__codelineno-31-14" id="__codelineno-31-14" name="__codelineno-31-14"></a><span class="n">lf</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">sink_sam</span><span class="p">(</span><span class="s2">"output.sam"</span><span class="p">)</span>
</span><span id="__span-31-15"><a href="#__codelineno-31-15" id="__codelineno-31-15" name="__codelineno-31-15"></a><span class="n">lf</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">sink_cram</span><span class="p">(</span><span class="s2">"output.cram"</span><span class="p">,</span> <span class="n">reference_path</span><span class="o">=</span><span class="s2">"ref.fa"</span><span class="p">)</span>
</span><span id="__span-31-16"><a href="#__codelineno-31-16" id="__codelineno-31-16" name="__codelineno-31-16"></a><span class="n">lf</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">sink_vcf</span><span class="p">(</span><span class="s2">"output.vcf.bgz"</span><span class="p">)</span>
</span><span id="__span-31-17"><a href="#__codelineno-31-17" id="__codelineno-31-17" name="__codelineno-31-17"></a><span class="n">lf</span><span class="o">.</span><span class="n">pb</span><span class="o">.</span><span class="n">sink_fastq</span><span class="p">(</span><span class="s2">"output.fastq.bgz"</span><span class="p">)</span>
</span></code></pre></div>
<h2 id="sql-powered-data-processing">SQL-powered data processing<a class="headerlink" href="#sql-powered-data-processing" title="Permanent link"></a></h2>
<p>polars-bio provides a SQL-like API for bioinformatic data querying or manipulation.
Check <a href="https://datafusion.apache.org/user-guide/sql/index.html">SQL reference</a> for more details.</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-32-1"><a href="#__codelineno-32-1" id="__codelineno-32-1" name="__codelineno-32-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-32-2"><a href="#__codelineno-32-2" id="__codelineno-32-2" name="__codelineno-32-2"></a><span class="n">pb</span><span class="o">.</span><span class="n">register_vcf</span><span class="p">(</span><span class="s2">"gs://gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.vcf.gz"</span><span class="p">,</span> <span class="s2">"gnomad_sv"</span><span class="p">,</span> <span class="n">info_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"SVTYPE"</span><span class="p">,</span> <span class="s2">"SVLEN"</span><span class="p">])</span>
</span><span id="__span-32-3"><a href="#__codelineno-32-3" id="__codelineno-32-3" name="__codelineno-32-3"></a><span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="s2">"SELECT * FROM gnomad_sv WHERE SVTYPE = 'DEL' AND SVLEN &gt; 1000"</span><span class="p">)</span><span class="o">.</span><span class="n">limit</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span></code></pre></div>
<h3 id="accessing-registered-tables">Accessing registered tables<a class="headerlink" href="#accessing-registered-tables" title="Permanent link"></a></h3>
<p>You can access registered tables programmatically using the <code>ctx.table()</code> method, which returns a DataFusion DataFrame:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-33-1"><a href="#__codelineno-33-1" id="__codelineno-33-1" name="__codelineno-33-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-33-2"><a href="#__codelineno-33-2" id="__codelineno-33-2" name="__codelineno-33-2"></a><span class="kn">from</span><span class="w"> </span><span class="nn">polars_bio.context</span><span class="w"> </span><span class="kn">import</span> <span class="n">ctx</span>
</span><span id="__span-33-3"><a href="#__codelineno-33-3" id="__codelineno-33-3" name="__codelineno-33-3"></a>
</span><span id="__span-33-4"><a href="#__codelineno-33-4" id="__codelineno-33-4" name="__codelineno-33-4"></a><span class="c1"># Register a file as a table</span>
</span><span id="__span-33-5"><a href="#__codelineno-33-5" id="__codelineno-33-5" name="__codelineno-33-5"></a><span class="n">pb</span><span class="o">.</span><span class="n">register_vcf</span><span class="p">(</span><span class="s2">"variants.vcf"</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">"my_variants"</span><span class="p">)</span>
</span><span id="__span-33-6"><a href="#__codelineno-33-6" id="__codelineno-33-6" name="__codelineno-33-6"></a>
</span><span id="__span-33-7"><a href="#__codelineno-33-7" id="__codelineno-33-7" name="__codelineno-33-7"></a><span class="c1"># Get the table as a DataFusion DataFrame</span>
</span><span id="__span-33-8"><a href="#__codelineno-33-8" id="__codelineno-33-8" name="__codelineno-33-8"></a><span class="n">df</span> <span class="o">=</span> <span class="n">ctx</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">"my_variants"</span><span class="p">)</span>
</span><span id="__span-33-9"><a href="#__codelineno-33-9" id="__codelineno-33-9" name="__codelineno-33-9"></a>
</span><span id="__span-33-10"><a href="#__codelineno-33-10" id="__codelineno-33-10" name="__codelineno-33-10"></a><span class="c1"># Access the Arrow schema (includes coordinate system metadata)</span>
</span><span id="__span-33-11"><a href="#__codelineno-33-11" id="__codelineno-33-11" name="__codelineno-33-11"></a><span class="n">schema</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">schema</span><span class="p">()</span>
</span><span id="__span-33-12"><a href="#__codelineno-33-12" id="__codelineno-33-12" name="__codelineno-33-12"></a><span class="nb">print</span><span class="p">(</span><span class="n">schema</span><span class="o">.</span><span class="n">metadata</span><span class="p">)</span>  <span class="c1"># {b'bio.coordinate_system_zero_based': b'false'}</span>
</span><span id="__span-33-13"><a href="#__codelineno-33-13" id="__codelineno-33-13" name="__codelineno-33-13"></a>
</span><span id="__span-33-14"><a href="#__codelineno-33-14" id="__codelineno-33-14" name="__codelineno-33-14"></a><span class="c1"># Execute queries on the DataFrame</span>
</span><span id="__span-33-15"><a href="#__codelineno-33-15" id="__codelineno-33-15" name="__codelineno-33-15"></a><span class="n">result</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">"chrom"</span><span class="p">]</span> <span class="o">==</span> <span class="s2">"chr1"</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span></code></pre></div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>The <code>ctx.table()</code> method is useful for:</p>
<ol>
<li>Accessing Arrow schema metadata (including coordinate system information)</li>
<li>Using the DataFusion DataFrame API directly</li>
<li>Integrating with other DataFusion-based tools</li>
</ol>
</div>
<h3 id="schema-inspection">Schema Inspection<a class="headerlink" href="#schema-inspection" title="Permanent link"></a></h3>
<p>Quickly inspect BAM/CRAM file schemas without reading the entire file:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-34-1"><a href="#__codelineno-34-1" id="__codelineno-34-1" name="__codelineno-34-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-34-2"><a href="#__codelineno-34-2" id="__codelineno-34-2" name="__codelineno-34-2"></a>
</span><span id="__span-34-3"><a href="#__codelineno-34-3" id="__codelineno-34-3" name="__codelineno-34-3"></a><span class="c1"># Get schema information for BAM file</span>
</span><span id="__span-34-4"><a href="#__codelineno-34-4" id="__codelineno-34-4" name="__codelineno-34-4"></a><span class="n">schema</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">describe_bam</span><span class="p">(</span><span class="s2">"file.bam"</span><span class="p">)</span>
</span><span id="__span-34-5"><a href="#__codelineno-34-5" id="__codelineno-34-5" name="__codelineno-34-5"></a><span class="nb">print</span><span class="p">(</span><span class="n">schema</span><span class="p">)</span>
</span><span id="__span-34-6"><a href="#__codelineno-34-6" id="__codelineno-34-6" name="__codelineno-34-6"></a><span class="c1"># shape: (11, 2)</span>
</span><span id="__span-34-7"><a href="#__codelineno-34-7" id="__codelineno-34-7" name="__codelineno-34-7"></a><span class="c1"># ‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê</span>
</span><span id="__span-34-8"><a href="#__codelineno-34-8" id="__codelineno-34-8" name="__codelineno-34-8"></a><span class="c1"># ‚îÇ column          ‚îÜ datatype ‚îÇ</span>
</span><span id="__span-34-9"><a href="#__codelineno-34-9" id="__codelineno-34-9" name="__codelineno-34-9"></a><span class="c1"># ‚îÇ ---             ‚îÜ ---      ‚îÇ</span>
</span><span id="__span-34-10"><a href="#__codelineno-34-10" id="__codelineno-34-10" name="__codelineno-34-10"></a><span class="c1"># ‚îÇ str             ‚îÜ str      ‚îÇ</span>
</span><span id="__span-34-11"><a href="#__codelineno-34-11" id="__codelineno-34-11" name="__codelineno-34-11"></a><span class="c1"># ‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°</span>
</span><span id="__span-34-12"><a href="#__codelineno-34-12" id="__codelineno-34-12" name="__codelineno-34-12"></a><span class="c1"># ‚îÇ name            ‚îÜ String   ‚îÇ</span>
</span><span id="__span-34-13"><a href="#__codelineno-34-13" id="__codelineno-34-13" name="__codelineno-34-13"></a><span class="c1"># ‚îÇ chrom           ‚îÜ String   ‚îÇ</span>
</span><span id="__span-34-14"><a href="#__codelineno-34-14" id="__codelineno-34-14" name="__codelineno-34-14"></a><span class="c1"># ‚îÇ start           ‚îÜ UInt32   ‚îÇ</span>
</span><span id="__span-34-15"><a href="#__codelineno-34-15" id="__codelineno-34-15" name="__codelineno-34-15"></a><span class="c1"># ...</span>
</span><span id="__span-34-16"><a href="#__codelineno-34-16" id="__codelineno-34-16" name="__codelineno-34-16"></a>
</span><span id="__span-34-17"><a href="#__codelineno-34-17" id="__codelineno-34-17" name="__codelineno-34-17"></a><span class="c1"># Include tag columns in schema</span>
</span><span id="__span-34-18"><a href="#__codelineno-34-18" id="__codelineno-34-18" name="__codelineno-34-18"></a><span class="n">schema</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">describe_bam</span><span class="p">(</span><span class="s2">"file.bam"</span><span class="p">,</span> <span class="n">tag_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"NM"</span><span class="p">,</span> <span class="s2">"AS"</span><span class="p">,</span> <span class="s2">"MD"</span><span class="p">])</span>
</span><span id="__span-34-19"><a href="#__codelineno-34-19" id="__codelineno-34-19" name="__codelineno-34-19"></a><span class="nb">print</span><span class="p">(</span><span class="n">schema</span><span class="p">)</span>  <span class="c1"># Shows 14 columns including tags</span>
</span><span id="__span-34-20"><a href="#__codelineno-34-20" id="__codelineno-34-20" name="__codelineno-34-20"></a>
</span><span id="__span-34-21"><a href="#__codelineno-34-21" id="__codelineno-34-21" name="__codelineno-34-21"></a><span class="c1"># CRAM schema</span>
</span><span id="__span-34-22"><a href="#__codelineno-34-22" id="__codelineno-34-22" name="__codelineno-34-22"></a><span class="n">schema</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">describe_cram</span><span class="p">(</span><span class="s2">"file.cram"</span><span class="p">)</span>
</span></code></pre></div>
<h3 id="bam-optional-tags">BAM Optional Tags<a class="headerlink" href="#bam-optional-tags" title="Permanent link"></a></h3>
<p>polars-bio supports reading BAM optional alignment tags as individual columns. Tags are only parsed when explicitly requested, ensuring zero overhead for standard reads.</p>
<blockquote>
<p><strong>Note</strong>: CRAM tag support is planned for a future release. The <code>tag_fields</code> parameter is accepted for CRAM functions but currently ignored with a warning.</p>
</blockquote>
<h4 id="usage">Usage<a class="headerlink" href="#usage" title="Permanent link"></a></h4>
<div class="language-python highlight"><pre><span></span><code><span id="__span-35-1"><a href="#__codelineno-35-1" id="__codelineno-35-1" name="__codelineno-35-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-35-2"><a href="#__codelineno-35-2" id="__codelineno-35-2" name="__codelineno-35-2"></a>
</span><span id="__span-35-3"><a href="#__codelineno-35-3" id="__codelineno-35-3" name="__codelineno-35-3"></a><span class="c1"># Read BAM with specific tags</span>
</span><span id="__span-35-4"><a href="#__codelineno-35-4" id="__codelineno-35-4" name="__codelineno-35-4"></a><span class="n">df</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">read_bam</span><span class="p">(</span>
</span><span id="__span-35-5"><a href="#__codelineno-35-5" id="__codelineno-35-5" name="__codelineno-35-5"></a>    <span class="s2">"alignments.bam"</span><span class="p">,</span>
</span><span id="__span-35-6"><a href="#__codelineno-35-6" id="__codelineno-35-6" name="__codelineno-35-6"></a>    <span class="n">tag_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"NM"</span><span class="p">,</span> <span class="s2">"AS"</span><span class="p">,</span> <span class="s2">"MD"</span><span class="p">]</span>  <span class="c1"># Edit distance, alignment score, mismatch string</span>
</span><span id="__span-35-7"><a href="#__codelineno-35-7" id="__codelineno-35-7" name="__codelineno-35-7"></a><span class="p">)</span>
</span><span id="__span-35-8"><a href="#__codelineno-35-8" id="__codelineno-35-8" name="__codelineno-35-8"></a>
</span><span id="__span-35-9"><a href="#__codelineno-35-9" id="__codelineno-35-9" name="__codelineno-35-9"></a><span class="c1"># Tags appear as regular columns</span>
</span><span id="__span-35-10"><a href="#__codelineno-35-10" id="__codelineno-35-10" name="__codelineno-35-10"></a><span class="nb">print</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="s2">"name"</span><span class="p">,</span> <span class="s2">"chrom"</span><span class="p">,</span> <span class="s2">"NM"</span><span class="p">,</span> <span class="s2">"AS"</span><span class="p">]))</span>
</span><span id="__span-35-11"><a href="#__codelineno-35-11" id="__codelineno-35-11" name="__codelineno-35-11"></a>
</span><span id="__span-35-12"><a href="#__codelineno-35-12" id="__codelineno-35-12" name="__codelineno-35-12"></a><span class="c1"># Lazy scan with tag filtering</span>
</span><span id="__span-35-13"><a href="#__codelineno-35-13" id="__codelineno-35-13" name="__codelineno-35-13"></a><span class="n">lf</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">scan_bam</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">,</span> <span class="n">tag_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"NM"</span><span class="p">,</span> <span class="s2">"AS"</span><span class="p">])</span>
</span><span id="__span-35-14"><a href="#__codelineno-35-14" id="__codelineno-35-14" name="__codelineno-35-14"></a><span class="n">high_quality</span> <span class="o">=</span> <span class="n">lf</span><span class="o">.</span><span class="n">filter</span><span class="p">((</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"NM"</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">2</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"AS"</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">100</span><span class="p">))</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-35-15"><a href="#__codelineno-35-15" id="__codelineno-35-15" name="__codelineno-35-15"></a>
</span><span id="__span-35-16"><a href="#__codelineno-35-16" id="__codelineno-35-16" name="__codelineno-35-16"></a><span class="c1"># SQL queries (tags must be quoted)</span>
</span><span id="__span-35-17"><a href="#__codelineno-35-17" id="__codelineno-35-17" name="__codelineno-35-17"></a><span class="n">pb</span><span class="o">.</span><span class="n">register_bam</span><span class="p">(</span><span class="s2">"alignments.bam"</span><span class="p">,</span> <span class="s2">"reads"</span><span class="p">,</span> <span class="n">tag_fields</span><span class="o">=</span><span class="p">[</span><span class="s2">"NM"</span><span class="p">,</span> <span class="s2">"RG"</span><span class="p">])</span>
</span><span id="__span-35-18"><a href="#__codelineno-35-18" id="__codelineno-35-18" name="__codelineno-35-18"></a><span class="n">result</span> <span class="o">=</span> <span class="n">pb</span><span class="o">.</span><span class="n">sql</span><span class="p">(</span><span class="s1">'SELECT name, "NM" FROM reads WHERE "NM" &lt;= 2'</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span></code></pre></div>
<h4 id="common-tags">Common Tags<a class="headerlink" href="#common-tags" title="Permanent link"></a></h4>
<ul>
<li><strong>NM</strong> (Int32): Edit distance to reference</li>
<li><strong>MD</strong> (Utf8): Mismatch positions string</li>
<li><strong>AS</strong> (Int32): Alignment score</li>
<li><strong>XS</strong> (Int32): Secondary alignment score</li>
<li><strong>RG</strong> (Utf8): Read group identifier</li>
<li><strong>CB</strong> (Utf8): Cell barcode (single-cell)</li>
<li><strong>UB</strong> (Utf8): UMI barcode (single-cell)</li>
</ul>
<p>Full registry includes ~40 common SAM tags.</p>
<h4 id="performance">Performance<a class="headerlink" href="#performance" title="Permanent link"></a></h4>
<ul>
<li>Zero overhead when <code>tag_fields=None</code> (default)</li>
<li>Projection pushdown: only selected tags are parsed</li>
<li>Tags parsed once per batch, not per record</li>
</ul>
<div class="language-shell highlight"><pre><span></span><code><span id="__span-36-1"><a href="#__codelineno-36-1" id="__codelineno-36-1" name="__codelineno-36-1"></a>shape:<span class="w"> </span><span class="o">(</span><span class="m">3</span>,<span class="w"> </span><span class="m">10</span><span class="o">)</span>
</span><span id="__span-36-2"><a href="#__codelineno-36-2" id="__codelineno-36-2" name="__codelineno-36-2"></a>‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
</span><span id="__span-36-3"><a href="#__codelineno-36-3" id="__codelineno-36-3" name="__codelineno-36-3"></a>‚îÇ<span class="w"> </span>chrom<span class="w"> </span>‚îÜ<span class="w"> </span>start<span class="w"> </span>‚îÜ<span class="w"> </span>end<span class="w">   </span>‚îÜ<span class="w"> </span>id<span class="w">                             </span>‚îÜ<span class="w"> </span>‚Ä¶<span class="w"> </span>‚îÜ<span class="w"> </span>qual<span class="w">  </span>‚îÜ<span class="w"> </span>filter<span class="w">     </span>‚îÜ<span class="w"> </span>svtype<span class="w"> </span>‚îÜ<span class="w"> </span>svlen<span class="w"> </span>‚îÇ
</span><span id="__span-36-4"><a href="#__codelineno-36-4" id="__codelineno-36-4" name="__codelineno-36-4"></a>‚îÇ<span class="w"> </span>---<span class="w">   </span>‚îÜ<span class="w"> </span>---<span class="w">   </span>‚îÜ<span class="w"> </span>---<span class="w">   </span>‚îÜ<span class="w"> </span>---<span class="w">                            </span>‚îÜ<span class="w">   </span>‚îÜ<span class="w"> </span>---<span class="w">   </span>‚îÜ<span class="w"> </span>---<span class="w">        </span>‚îÜ<span class="w"> </span>---<span class="w">    </span>‚îÜ<span class="w"> </span>---<span class="w">   </span>‚îÇ
</span><span id="__span-36-5"><a href="#__codelineno-36-5" id="__codelineno-36-5" name="__codelineno-36-5"></a>‚îÇ<span class="w"> </span>str<span class="w">   </span>‚îÜ<span class="w"> </span>u32<span class="w">   </span>‚îÜ<span class="w"> </span>u32<span class="w">   </span>‚îÜ<span class="w"> </span>str<span class="w">                            </span>‚îÜ<span class="w">   </span>‚îÜ<span class="w"> </span>f64<span class="w">   </span>‚îÜ<span class="w"> </span>str<span class="w">        </span>‚îÜ<span class="w"> </span>str<span class="w">    </span>‚îÜ<span class="w"> </span>i32<span class="w">   </span>‚îÇ
</span><span id="__span-36-6"><a href="#__codelineno-36-6" id="__codelineno-36-6" name="__codelineno-36-6"></a>‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°
</span><span id="__span-36-7"><a href="#__codelineno-36-7" id="__codelineno-36-7" name="__codelineno-36-7"></a>‚îÇ<span class="w"> </span>chr1<span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">22000</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">30000</span><span class="w"> </span>‚îÜ<span class="w"> </span>gnomAD-SV_v3_DEL_chr1_fa103016<span class="w"> </span>‚îÜ<span class="w"> </span>‚Ä¶<span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">999</span>.0<span class="w"> </span>‚îÜ<span class="w"> </span>HIGH_NCR<span class="w">   </span>‚îÜ<span class="w"> </span>DEL<span class="w">    </span>‚îÜ<span class="w"> </span><span class="m">8000</span><span class="w">  </span>‚îÇ
</span><span id="__span-36-8"><a href="#__codelineno-36-8" id="__codelineno-36-8" name="__codelineno-36-8"></a>‚îÇ<span class="w"> </span>chr1<span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">40000</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">47000</span><span class="w"> </span>‚îÜ<span class="w"> </span>gnomAD-SV_v3_DEL_chr1_b26f63f7<span class="w"> </span>‚îÜ<span class="w"> </span>‚Ä¶<span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">145</span>.0<span class="w"> </span>‚îÜ<span class="w"> </span>PASS<span class="w">       </span>‚îÜ<span class="w"> </span>DEL<span class="w">    </span>‚îÜ<span class="w"> </span><span class="m">7000</span><span class="w">  </span>‚îÇ
</span><span id="__span-36-9"><a href="#__codelineno-36-9" id="__codelineno-36-9" name="__codelineno-36-9"></a>‚îÇ<span class="w"> </span>chr1<span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">79086</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">88118</span><span class="w"> </span>‚îÜ<span class="w"> </span>gnomAD-SV_v3_DEL_chr1_733c4ef0<span class="w"> </span>‚îÜ<span class="w"> </span>‚Ä¶<span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">344</span>.0<span class="w"> </span>‚îÜ<span class="w"> </span>UNRESOLVED<span class="w"> </span>‚îÜ<span class="w"> </span>DEL<span class="w">    </span>‚îÜ<span class="w"> </span><span class="m">9032</span><span class="w">  </span>‚îÇ
</span><span id="__span-36-10"><a href="#__codelineno-36-10" id="__codelineno-36-10" name="__codelineno-36-10"></a>‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
</span></code></pre></div>
<p>You can use <a href="../api/#polars_bio.register_view">view</a> mechanism to create a virtual table from a DataFrame that contain preprocessing steps and reuse it in multiple steps.
To avoid materializing the intermediate results in memory, you can run your processing in  <a href="./#streaming">streaming</a> mode.</p>
<h2 id="parallel-engine">Parallel engine üèéÔ∏è<a class="headerlink" href="#parallel-engine" title="Permanent link"></a></h2>
<p>It is straightforward to parallelize operations in polars-bio. The library is built on top of <a href="https://datafusion.apache.org/">Apache DataFusion</a>  you can set
the degree of parallelism using the <code>datafusion.execution.target_partitions</code> option, e.g.:
</p><div class="language-python highlight"><pre><span></span><code><span id="__span-37-1"><a href="#__codelineno-37-1" id="__codelineno-37-1" name="__codelineno-37-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-37-2"><a href="#__codelineno-37-2" id="__codelineno-37-2" name="__codelineno-37-2"></a><span class="n">pb</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">"datafusion.execution.target_partitions"</span><span class="p">,</span> <span class="s2">"8"</span><span class="p">)</span>
</span></code></pre></div><p></p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<ol>
<li>The default value is <strong>1</strong> (parallel execution disabled).</li>
<li>The <code>datafusion.execution.target_partitions</code> option is a global setting and affects all operations in the current session.</li>
<li>Check <a href="../performance/#parallel-execution-and-scalability">available strategies</a> for optimal performance.</li>
<li>See  the other configuration settings in the Apache DataFusion <a href="https://datafusion.apache.org/user-guide/configs.html">documentation</a>.</li>
</ol>
</div>
<h2 id="cloud-storage">Cloud storage ‚òÅÔ∏è<a class="headerlink" href="#cloud-storage" title="Permanent link"></a></h2>
<p>polars-bio supports direct streamed reading from cloud storages (e.g. S3, GCS) enabling processing large-scale genomics data without materializing in memory.
It is built upon the <a href="https://opendal.apache.org/">OpenDAL</a> project, a unified data access layer for cloud storage, which allows to read  bioinformatic file formats from various cloud storage providers. For Apache DataFusion <strong>native</strong> file formats, such as Parquet or CSV please
refer to <a href="https://datafusion.apache.org/user-guide/cli/datasources.html#locations">DataFusion user guide</a>.</p>
<h3 id="example">Example<a class="headerlink" href="#example" title="Permanent link"></a></h3>
<p></p><div class="language-python highlight"><pre><span></span><code><span id="__span-38-1"><a href="#__codelineno-38-1" id="__codelineno-38-1" name="__codelineno-38-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-38-2"><a href="#__codelineno-38-2" id="__codelineno-38-2" name="__codelineno-38-2"></a><span class="c1">## Register VCF files from Google Cloud Storage that will be streamed - no need to download them to the local disk, size ~0.8TB</span>
</span><span id="__span-38-3"><a href="#__codelineno-38-3" id="__codelineno-38-3" name="__codelineno-38-3"></a><span class="n">pb</span><span class="o">.</span><span class="n">register_vcf</span><span class="p">(</span><span class="s2">"gs://gcp-public-data--gnomad/release/2.1.1/liftover_grch38/vcf/genomes/gnomad.genomes.r2.1.1.sites.liftover_grch38.vcf.bgz"</span><span class="p">,</span> <span class="s2">"gnomad_big"</span><span class="p">,</span> <span class="n">allow_anonymous</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-38-4"><a href="#__codelineno-38-4" id="__codelineno-38-4" name="__codelineno-38-4"></a><span class="n">pb</span><span class="o">.</span><span class="n">register_vcf</span><span class="p">(</span><span class="s2">"gs://gcp-public-data--gnomad/release/4.1/genome_sv/gnomad.v4.1.sv.sites.vcf.gz"</span><span class="p">,</span> <span class="s2">"gnomad_sv"</span><span class="p">,</span> <span class="n">allow_anonymous</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span><span id="__span-38-5"><a href="#__codelineno-38-5" id="__codelineno-38-5" name="__codelineno-38-5"></a><span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="s2">"gnomad_sv"</span><span class="p">,</span> <span class="s2">"gnomad_big"</span><span class="p">,</span> <span class="n">streaming</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">sink_parquet</span><span class="p">(</span><span class="s2">"/tmp/overlap.parquet"</span><span class="p">)</span>
</span></code></pre></div>
It is  especially useful when combined with <a href="./#sql-powered-data-processing">SQL</a> support for preprocessing and <a href="./#streaming">streaming</a> processing capabilities.<p></p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you access cloud storage with authentication provided, please make sure the <code>allow_anonymous</code> parameter is set to <code>False</code> in the read/describe/register_table functions.</p>
</div>
<h3 id="supported-features">Supported features<a class="headerlink" href="#supported-features" title="Permanent link"></a></h3>
<table>
<thead>
<tr>
<th>Feature</th>
<th>AWS S3</th>
<th>Google Cloud Storage</th>
<th>Azure Blob Storage</th>
</tr>
</thead>
<tbody>
<tr>
<td>Anonymous access</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
</tr>
<tr>
<td>Authenticated access</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td>Requester Pays</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Concurrent requests<sup>1</sup></td>
<td></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
</tr>
<tr>
<td>Streaming reads</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><sup>1</sup>For more information on concurrent requests and block size tuning please refer to <a href="https://github.com/biodatageeks/polars-bio/issues/132#issuecomment-2967687947">issue</a>.</p>
</div>
<h3 id="aws-s3-configuration">AWS S3 configuration<a class="headerlink" href="#aws-s3-configuration" title="Permanent link"></a></h3>
<p>Supported environment variables:</p>
<table>
<thead>
<tr>
<th>Variable</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>AWS_ACCESS_KEY_ID</td>
<td>AWS access key ID for authenticated access to S3.</td>
</tr>
<tr>
<td>AWS_SECRET_ACCESS_KEY</td>
<td>AWS secret access key for authenticated access to S3.</td>
</tr>
<tr>
<td>AWS_ENDPOINT_URL</td>
<td>Custom S3 endpoint URL for accessing S3-compatible storage.</td>
</tr>
<tr>
<td>AWS_REGION  or AWS_DEFAULT_REGION</td>
<td>AWS region for accessing S3.</td>
</tr>
</tbody>
</table>
<h3 id="google-cloud-storage-configuration">Google Cloud Storage configuration<a class="headerlink" href="#google-cloud-storage-configuration" title="Permanent link"></a></h3>
<p>Supported environment variables:</p>
<table>
<thead>
<tr>
<th>Variable</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>GOOGLE_APPLICATION_CREDENTIALS</td>
<td>Path to the Google Cloud service account key file for authenticated access to GCS.</td>
</tr>
</tbody>
</table>
<h3 id="azure-blob-storage-configuration">Azure Blob Storage configuration<a class="headerlink" href="#azure-blob-storage-configuration" title="Permanent link"></a></h3>
<p>Supported environment variables:</p>
<table>
<thead>
<tr>
<th>Variable</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>AZURE_STORAGE_ACCOUNT</td>
<td>Azure Storage account name for authenticated access to Azure Blob Storage.</td>
</tr>
<tr>
<td>AZURE_STORAGE_KEY</td>
<td>Azure Storage account key for authenticated access to Azure Blob Storage.</td>
</tr>
<tr>
<td>AZURE_ENDPOINT_URL</td>
<td>Azure Blob Storage endpoint URL for accessing Azure Blob Storage.</td>
</tr>
</tbody>
</table>
<h2 id="streaming">Streaming üöÇ<a class="headerlink" href="#streaming" title="Permanent link"></a></h2>
<p>polars-bio supports out-of-core processing with Apache DataFusion async <a href="https://docs.rs/datafusion/46.0.0/datafusion/physical_plan/trait.ExecutionPlan.html#tymethod.execute">streams</a> and Polars LazyFrame <a href="https://docs.pola.rs/user-guide/concepts/_streaming/">streaming</a> option.
It can bring  significant speedup as well reduction in memory usage allowing to process large datasets that do not fit in memory.
See our benchmark <a href="../performance/#calculate-overlaps-and-export-to-a-csv-file-7-8">results</a>.
There are 2 ways of using streaming mode:</p>
<ol>
<li>
<p>By setting the <code>output_type</code> to <code>datafusion.DataFrame</code> and using the Python DataFrame <a href="https://datafusion.apache.org/python/autoapi/datafusion/dataframe/index.html">API</a>, including methods such as <a href="https://datafusion.apache.org/python/autoapi/datafusion/dataframe/index.html#datafusion.dataframe.DataFrame.count">count</a>, <a href="https://datafusion.apache.org/python/autoapi/datafusion/dataframe/index.html#datafusion.dataframe.DataFrame.write_parquet">write_parquet</a> or <a href="https://datafusion.apache.org/python/autoapi/datafusion/dataframe/index.html#datafusion.dataframe.DataFrame.write_csv">write_csv</a> or <a href="https://datafusion.apache.org/python/autoapi/datafusion/dataframe/index.html#datafusion.dataframe.DataFrame.write_json">write_json</a>. In this option you completely bypass the polars streaming engine.</p>
<p><code>python
import polars_bio as pb
import polars as pl
pb.overlap("/tmp/gnomad.v4.1.sv.sites.parquet", "/tmp/gnomad.exomes.v4.1.sites.chr1.parquet", output_type="datafusion.DataFrame").write_parquet("/tmp/overlap.parquet")
pl.scan_parquet("/tmp/overlap.parquet").collect().count()</code>
</p><div class="language-shell highlight"><pre><span></span><code><span id="__span-39-1"><a href="#__codelineno-39-1" id="__codelineno-39-1" name="__codelineno-39-1"></a><span class="w"> </span>shape:<span class="w"> </span><span class="o">(</span><span class="m">1</span>,<span class="w"> </span><span class="m">6</span><span class="o">)</span>
</span><span id="__span-39-2"><a href="#__codelineno-39-2" id="__codelineno-39-2" name="__codelineno-39-2"></a><span class="w"> </span>‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
</span><span id="__span-39-3"><a href="#__codelineno-39-3" id="__codelineno-39-3" name="__codelineno-39-3"></a><span class="w"> </span>‚îÇ<span class="w"> </span>chrom_1<span class="w">    </span>‚îÜ<span class="w"> </span>start_1<span class="w">    </span>‚îÜ<span class="w"> </span>end_1<span class="w">      </span>‚îÜ<span class="w"> </span>chrom_2<span class="w">    </span>‚îÜ<span class="w"> </span>start_2<span class="w">    </span>‚îÜ<span class="w"> </span>end_2<span class="w">      </span>‚îÇ
</span><span id="__span-39-4"><a href="#__codelineno-39-4" id="__codelineno-39-4" name="__codelineno-39-4"></a><span class="w"> </span>‚îÇ<span class="w"> </span>---<span class="w">        </span>‚îÜ<span class="w"> </span>---<span class="w">        </span>‚îÜ<span class="w"> </span>---<span class="w">        </span>‚îÜ<span class="w"> </span>---<span class="w">        </span>‚îÜ<span class="w"> </span>---<span class="w">        </span>‚îÜ<span class="w"> </span>---<span class="w">        </span>‚îÇ
</span><span id="__span-39-5"><a href="#__codelineno-39-5" id="__codelineno-39-5" name="__codelineno-39-5"></a><span class="w"> </span>‚îÇ<span class="w"> </span>u32<span class="w">        </span>‚îÜ<span class="w"> </span>u32<span class="w">        </span>‚îÜ<span class="w"> </span>u32<span class="w">        </span>‚îÜ<span class="w"> </span>u32<span class="w">        </span>‚îÜ<span class="w"> </span>u32<span class="w">        </span>‚îÜ<span class="w"> </span>u32<span class="w">        </span>‚îÇ
</span><span id="__span-39-6"><a href="#__codelineno-39-6" id="__codelineno-39-6" name="__codelineno-39-6"></a><span class="w"> </span>‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°
</span><span id="__span-39-7"><a href="#__codelineno-39-7" id="__codelineno-39-7" name="__codelineno-39-7"></a><span class="w"> </span>‚îÇ<span class="w"> </span><span class="m">2629727337</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">2629727337</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">2629727337</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">2629727337</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">2629727337</span><span class="w"> </span>‚îÜ<span class="w"> </span><span class="m">2629727337</span><span class="w"> </span>‚îÇ
</span><span id="__span-39-8"><a href="#__codelineno-39-8" id="__codelineno-39-8" name="__codelineno-39-8"></a><span class="w"> </span>‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
</span></code></pre></div><p></p>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>If you only need to write the results as fast as possible into one of the above file formats or quickly get the row count, then it is in the most cases the <strong>best</strong> option.</p>
</div>
</li>
<li>
<p>Using polars new streaming engine:</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-40-1"><a href="#__codelineno-40-1" id="__codelineno-40-1" name="__codelineno-40-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
</span><span id="__span-40-2"><a href="#__codelineno-40-2" id="__codelineno-40-2" name="__codelineno-40-2"></a><span class="kn">import</span><span class="w"> </span><span class="nn">polars_bio</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pb</span>
</span><span id="__span-40-3"><a href="#__codelineno-40-3" id="__codelineno-40-3" name="__codelineno-40-3"></a><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">'BENCH_DATA_ROOT'</span><span class="p">]</span> <span class="o">=</span> <span class="s2">"/Users/mwiewior/research/data/databio"</span>
</span><span id="__span-40-4"><a href="#__codelineno-40-4" id="__codelineno-40-4" name="__codelineno-40-4"></a><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">'POLARS_VERBOSE'</span><span class="p">]</span> <span class="o">=</span> <span class="s2">"1"</span>
</span><span id="__span-40-5"><a href="#__codelineno-40-5" id="__codelineno-40-5" name="__codelineno-40-5"></a>
</span><span id="__span-40-6"><a href="#__codelineno-40-6" id="__codelineno-40-6" name="__codelineno-40-6"></a><span class="n">cols</span><span class="o">=</span><span class="p">[</span><span class="s2">"contig"</span><span class="p">,</span> <span class="s2">"pos_start"</span><span class="p">,</span> <span class="s2">"pos_end"</span><span class="p">]</span>
</span><span id="__span-40-7"><a href="#__codelineno-40-7" id="__codelineno-40-7" name="__codelineno-40-7"></a><span class="n">BENCH_DATA_ROOT</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getenv</span><span class="p">(</span><span class="s1">'BENCH_DATA_ROOT'</span><span class="p">,</span> <span class="s1">'/data/bench_data/databio'</span><span class="p">)</span>
</span><span id="__span-40-8"><a href="#__codelineno-40-8" id="__codelineno-40-8" name="__codelineno-40-8"></a><span class="n">df_1</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">BENCH_DATA_ROOT</span><span class="si">}</span><span class="s2">/exons/*.parquet"</span>
</span><span id="__span-40-9"><a href="#__codelineno-40-9" id="__codelineno-40-9" name="__codelineno-40-9"></a><span class="n">df_2</span> <span class="o">=</span>  <span class="sa">f</span><span class="s2">"</span><span class="si">{</span><span class="n">BENCH_DATA_ROOT</span><span class="si">}</span><span class="s2">/exons/*.parquet"</span>
</span><span id="__span-40-10"><a href="#__codelineno-40-10" id="__codelineno-40-10" name="__codelineno-40-10"></a><span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df_1</span><span class="p">,</span> <span class="n">df_2</span><span class="p">,</span> <span class="n">cols1</span><span class="o">=</span><span class="n">cols</span><span class="p">,</span> <span class="n">cols2</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="s2">"streaming"</span><span class="p">)</span><span class="o">.</span><span class="n">limit</span><span class="p">()</span>
</span></code></pre></div>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-41-1"><a href="#__codelineno-41-1" id="__codelineno-41-1" name="__codelineno-41-1"></a>1652814rows<span class="w"> </span><span class="o">[</span><span class="m">00</span>:00,<span class="w"> </span><span class="m">20208793</span>.67rows/s<span class="o">]</span>
</span><span id="__span-41-2"><a href="#__codelineno-41-2" id="__codelineno-41-2" name="__codelineno-41-2"></a><span class="o">[</span>MultiScanState<span class="o">]</span>:<span class="w"> </span>Readers<span class="w"> </span>disconnected
</span><span id="__span-41-3"><a href="#__codelineno-41-3" id="__codelineno-41-3" name="__codelineno-41-3"></a>polars-stream:<span class="w"> </span><span class="k">done</span><span class="w"> </span>running<span class="w"> </span>graph<span class="w"> </span>phase
</span><span id="__span-41-4"><a href="#__codelineno-41-4" id="__codelineno-41-4" name="__codelineno-41-4"></a>polars-stream:<span class="w"> </span>updating<span class="w"> </span>graph<span class="w"> </span>state
</span><span id="__span-41-5"><a href="#__codelineno-41-5" id="__codelineno-41-5" name="__codelineno-41-5"></a>shape:<span class="w"> </span><span class="o">(</span><span class="m">5</span>,<span class="w"> </span><span class="m">6</span><span class="o">)</span>
</span><span id="__span-41-6"><a href="#__codelineno-41-6" id="__codelineno-41-6" name="__codelineno-41-6"></a>‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
</span><span id="__span-41-7"><a href="#__codelineno-41-7" id="__codelineno-41-7" name="__codelineno-41-7"></a>‚îÇ<span class="w"> </span>contig_1<span class="w"> </span>‚îÜ<span class="w"> </span>pos_start_1<span class="w"> </span>‚îÜ<span class="w"> </span>pos_end_1<span class="w"> </span>‚îÜ<span class="w"> </span>contig_2<span class="w"> </span>‚îÜ<span class="w"> </span>pos_start_2<span class="w"> </span>‚îÜ<span class="w"> </span>pos_end_2<span class="w"> </span>‚îÇ
</span><span id="__span-41-8"><a href="#__codelineno-41-8" id="__codelineno-41-8" name="__codelineno-41-8"></a>‚îÇ<span class="w"> </span>---<span class="w">      </span>‚îÜ<span class="w"> </span>---<span class="w">         </span>‚îÜ<span class="w"> </span>---<span class="w">       </span>‚îÜ<span class="w"> </span>---<span class="w">      </span>‚îÜ<span class="w"> </span>---<span class="w">         </span>‚îÜ<span class="w"> </span>---<span class="w">       </span>‚îÇ
</span><span id="__span-41-9"><a href="#__codelineno-41-9" id="__codelineno-41-9" name="__codelineno-41-9"></a>‚îÇ<span class="w"> </span>str<span class="w">      </span>‚îÜ<span class="w"> </span>i32<span class="w">         </span>‚îÜ<span class="w"> </span>i32<span class="w">       </span>‚îÜ<span class="w"> </span>str<span class="w">      </span>‚îÜ<span class="w"> </span>i32<span class="w">         </span>‚îÜ<span class="w"> </span>i32<span class="w">       </span>‚îÇ
</span><span id="__span-41-10"><a href="#__codelineno-41-10" id="__codelineno-41-10" name="__codelineno-41-10"></a>‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°
</span><span id="__span-41-11"><a href="#__codelineno-41-11" id="__codelineno-41-11" name="__codelineno-41-11"></a>‚îÇ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">11873</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">12227</span><span class="w">     </span>‚îÜ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">11873</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">12227</span><span class="w">     </span>‚îÇ
</span><span id="__span-41-12"><a href="#__codelineno-41-12" id="__codelineno-41-12" name="__codelineno-41-12"></a>‚îÇ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">12612</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">12721</span><span class="w">     </span>‚îÜ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">12612</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">12721</span><span class="w">     </span>‚îÇ
</span><span id="__span-41-13"><a href="#__codelineno-41-13" id="__codelineno-41-13" name="__codelineno-41-13"></a>‚îÇ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">13220</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">14409</span><span class="w">     </span>‚îÜ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">13220</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">14409</span><span class="w">     </span>‚îÇ
</span><span id="__span-41-14"><a href="#__codelineno-41-14" id="__codelineno-41-14" name="__codelineno-41-14"></a>‚îÇ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">13220</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">14409</span><span class="w">     </span>‚îÜ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">14361</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">14829</span><span class="w">     </span>‚îÇ
</span><span id="__span-41-15"><a href="#__codelineno-41-15" id="__codelineno-41-15" name="__codelineno-41-15"></a>‚îÇ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">14361</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">14829</span><span class="w">     </span>‚îÜ<span class="w"> </span>chr1<span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">13220</span><span class="w">       </span>‚îÜ<span class="w"> </span><span class="m">14409</span><span class="w">     </span>‚îÇ
</span><span id="__span-41-16"><a href="#__codelineno-41-16" id="__codelineno-41-16" name="__codelineno-41-16"></a>‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
</span></code></pre></div>
</li>
</ol>
<p>Parallellism can be controlled using the <code>datafusion.execution.target_partitions</code>option as described in the <a href="./#parallel-engine-üèéÔ∏è">parallel engine</a> section (compare the row/s metric in the following examples).</p>
<p></p><div class="language-python highlight"><pre><span></span><code><span id="__span-42-1"><a href="#__codelineno-42-1" id="__codelineno-42-1" name="__codelineno-42-1"></a> <span class="n">pb</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">"datafusion.execution.target_partitions"</span><span class="p">,</span> <span class="s2">"1"</span><span class="p">)</span>
</span><span id="__span-42-2"><a href="#__codelineno-42-2" id="__codelineno-42-2" name="__codelineno-42-2"></a> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df_1</span><span class="p">,</span> <span class="n">df_2</span><span class="p">,</span> <span class="n">cols1</span><span class="o">=</span><span class="n">cols</span><span class="p">,</span> <span class="n">cols2</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="s2">"streaming"</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</span></code></pre></div>
<div class="language-shell highlight"><pre><span></span><code><span id="__span-43-1"><a href="#__codelineno-43-1" id="__codelineno-43-1" name="__codelineno-43-1"></a><span class="w"> </span>1652814rows<span class="w"> </span><span class="o">[</span><span class="m">00</span>:00,<span class="w"> </span><span class="m">19664163</span>.99rows/s<span class="o">]</span>
</span><span id="__span-43-2"><a href="#__codelineno-43-2" id="__codelineno-43-2" name="__codelineno-43-2"></a><span class="w"> </span>shape:<span class="w"> </span><span class="o">(</span><span class="m">1</span>,<span class="w"> </span><span class="m">6</span><span class="o">)</span>
</span><span id="__span-43-3"><a href="#__codelineno-43-3" id="__codelineno-43-3" name="__codelineno-43-3"></a><span class="w"> </span>‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
</span><span id="__span-43-4"><a href="#__codelineno-43-4" id="__codelineno-43-4" name="__codelineno-43-4"></a><span class="w"> </span>‚îÇ<span class="w"> </span>contig_1<span class="w"> </span>‚îÜ<span class="w"> </span>pos_start_1<span class="w"> </span>‚îÜ<span class="w"> </span>pos_end_1<span class="w"> </span>‚îÜ<span class="w"> </span>contig_2<span class="w"> </span>‚îÜ<span class="w"> </span>pos_start_2<span class="w"> </span>‚îÜ<span class="w"> </span>pos_end_2<span class="w"> </span>‚îÇ
</span><span id="__span-43-5"><a href="#__codelineno-43-5" id="__codelineno-43-5" name="__codelineno-43-5"></a><span class="w"> </span>‚îÇ<span class="w"> </span>---<span class="w">      </span>‚îÜ<span class="w"> </span>---<span class="w">         </span>‚îÜ<span class="w"> </span>---<span class="w">       </span>‚îÜ<span class="w"> </span>---<span class="w">      </span>‚îÜ<span class="w"> </span>---<span class="w">         </span>‚îÜ<span class="w"> </span>---<span class="w">       </span>‚îÇ
</span><span id="__span-43-6"><a href="#__codelineno-43-6" id="__codelineno-43-6" name="__codelineno-43-6"></a><span class="w"> </span>‚îÇ<span class="w"> </span>u32<span class="w">      </span>‚îÜ<span class="w"> </span>u32<span class="w">         </span>‚îÜ<span class="w"> </span>u32<span class="w">       </span>‚îÜ<span class="w"> </span>u32<span class="w">      </span>‚îÜ<span class="w"> </span>u32<span class="w">         </span>‚îÜ<span class="w"> </span>u32<span class="w">       </span>‚îÇ
</span><span id="__span-43-7"><a href="#__codelineno-43-7" id="__codelineno-43-7" name="__codelineno-43-7"></a><span class="w"> </span>‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°
</span><span id="__span-43-8"><a href="#__codelineno-43-8" id="__codelineno-43-8" name="__codelineno-43-8"></a><span class="w"> </span>‚îÇ<span class="w"> </span><span class="m">1652814</span><span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">   </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">   </span>‚îÇ
</span><span id="__span-43-9"><a href="#__codelineno-43-9" id="__codelineno-43-9" name="__codelineno-43-9"></a><span class="w"> </span>‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
</span></code></pre></div>
<div class="language-python highlight"><pre><span></span><code><span id="__span-44-1"><a href="#__codelineno-44-1" id="__codelineno-44-1" name="__codelineno-44-1"></a> <span class="n">pb</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">"datafusion.execution.target_partitions"</span><span class="p">,</span> <span class="s2">"2"</span><span class="p">)</span>
</span><span id="__span-44-2"><a href="#__codelineno-44-2" id="__codelineno-44-2" name="__codelineno-44-2"></a> <span class="n">pb</span><span class="o">.</span><span class="n">overlap</span><span class="p">(</span><span class="n">df_1</span><span class="p">,</span> <span class="n">df_2</span><span class="p">,</span> <span class="n">cols1</span><span class="o">=</span><span class="n">cols</span><span class="p">,</span> <span class="n">cols2</span><span class="o">=</span><span class="n">cols</span><span class="p">)</span><span class="o">.</span><span class="n">collect</span><span class="p">(</span><span class="n">engine</span><span class="o">=</span><span class="s2">"streaming"</span><span class="p">)</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
</span></code></pre></div><p></p>
<div class="language-shell highlight"><pre><span></span><code><span id="__span-45-1"><a href="#__codelineno-45-1" id="__codelineno-45-1" name="__codelineno-45-1"></a><span class="w"> </span>1652814rows<span class="w"> </span><span class="o">[</span><span class="m">00</span>:00,<span class="w"> </span><span class="m">27841987</span>.75rows/s<span class="o">]</span>
</span><span id="__span-45-2"><a href="#__codelineno-45-2" id="__codelineno-45-2" name="__codelineno-45-2"></a><span class="w"> </span>shape:<span class="w"> </span><span class="o">(</span><span class="m">1</span>,<span class="w"> </span><span class="m">6</span><span class="o">)</span>
</span><span id="__span-45-3"><a href="#__codelineno-45-3" id="__codelineno-45-3" name="__codelineno-45-3"></a><span class="w"> </span>‚îå‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¨‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îê
</span><span id="__span-45-4"><a href="#__codelineno-45-4" id="__codelineno-45-4" name="__codelineno-45-4"></a><span class="w"> </span>‚îÇ<span class="w"> </span>contig_1<span class="w"> </span>‚îÜ<span class="w"> </span>pos_start_1<span class="w"> </span>‚îÜ<span class="w"> </span>pos_end_1<span class="w"> </span>‚îÜ<span class="w"> </span>contig_2<span class="w"> </span>‚îÜ<span class="w"> </span>pos_start_2<span class="w"> </span>‚îÜ<span class="w"> </span>pos_end_2<span class="w"> </span>‚îÇ
</span><span id="__span-45-5"><a href="#__codelineno-45-5" id="__codelineno-45-5" name="__codelineno-45-5"></a><span class="w"> </span>‚îÇ<span class="w"> </span>---<span class="w">      </span>‚îÜ<span class="w"> </span>---<span class="w">         </span>‚îÜ<span class="w"> </span>---<span class="w">       </span>‚îÜ<span class="w"> </span>---<span class="w">      </span>‚îÜ<span class="w"> </span>---<span class="w">         </span>‚îÜ<span class="w"> </span>---<span class="w">       </span>‚îÇ
</span><span id="__span-45-6"><a href="#__codelineno-45-6" id="__codelineno-45-6" name="__codelineno-45-6"></a><span class="w"> </span>‚îÇ<span class="w"> </span>u32<span class="w">      </span>‚îÜ<span class="w"> </span>u32<span class="w">         </span>‚îÜ<span class="w"> </span>u32<span class="w">       </span>‚îÜ<span class="w"> </span>u32<span class="w">      </span>‚îÜ<span class="w"> </span>u32<span class="w">         </span>‚îÜ<span class="w"> </span>u32<span class="w">       </span>‚îÇ
</span><span id="__span-45-7"><a href="#__codelineno-45-7" id="__codelineno-45-7" name="__codelineno-45-7"></a><span class="w"> </span>‚ïû‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï™‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ïê‚ï°
</span><span id="__span-45-8"><a href="#__codelineno-45-8" id="__codelineno-45-8" name="__codelineno-45-8"></a><span class="w"> </span>‚îÇ<span class="w"> </span><span class="m">1652814</span><span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">   </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">  </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">     </span>‚îÜ<span class="w"> </span><span class="m">1652814</span><span class="w">   </span>‚îÇ
</span><span id="__span-45-9"><a href="#__codelineno-45-9" id="__codelineno-45-9" name="__codelineno-45-9"></a><span class="w"> </span>‚îî‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚î¥‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îÄ‚îò
</span></code></pre></div>
<h2 id="compression">Compression<a class="headerlink" href="#compression" title="Permanent link"></a></h2>
<p><em>polars-bio</em> supports <strong>GZIP</strong> (default file extension <code>*.gz</code>) and <strong>Block GZIP</strong> (BGZIP, default file extension <code>*.bgz</code>) when reading files from local and cloud storages.
For BGZIP-compressed FASTQ files, parallel decoding of compressed blocks is <strong>automatic</strong> ‚Äî see <a href="#automatic-parallel-partitioning">Automatic parallel partitioning</a> and <a href="#index-file-generation">Index file generation</a> for details. Please take a look at the following <a href="https://github.com/biodatageeks/polars-bio/issues/132">GitHub discussion</a>.</p>
<h2 id="dataframes-support">DataFrames support<a class="headerlink" href="#dataframes-support" title="Permanent link"></a></h2>
<table>
<thead>
<tr>
<th>I/O</th>
<th>Bioframe</th>
<th>polars-bio</th>
<th>PyRanges</th>
<th>Pybedtools</th>
<th>GenomicRanges</th>
</tr>
</thead>
<tbody>
<tr>
<td>Pandas DataFrame</td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td>Polars DataFrame</td>
<td></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
<td></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
</tr>
<tr>
<td>Polars LazyFrame</td>
<td></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>Native readers</td>
<td></td>
<td><img alt="‚úÖ" class="twemoji" src="https://cdn.jsdelivr.net/gh/jdecked/twemoji@16.0.1/assets/svg/2705.svg" title=":white_check_mark:"/></td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<h2 id="polars-integration">Polars Integration<a class="headerlink" href="#polars-integration" title="Permanent link"></a></h2>
<p>polars-bio leverages deep integration with Polars through the <a href="https://arrow.apache.org/docs/format/CDataInterface.html">Arrow C Data Interface</a>, enabling high-performance zero-copy data exchange between Polars LazyFrames and the Rust-based genomic range operations engine. Requires <strong>Polars &gt;= 1.37.0</strong>.</p>
<p>See the <a href="../developers/#polars-integration">Developers Guide</a> for architecture diagrams, performance details, and batch size configuration.</p>
</article>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
<button class="md-top md-icon" data-md-component="top" hidden="" type="button">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
</main>
<footer class="md-footer">
<nav aria-label="Footer" class="md-footer__inner md-grid">
<a aria-label="Previous: üèÉüèº‚Äç‚ôÇÔ∏è Quick start" class="md-footer__link md-footer__link--prev" href="../quickstart/">
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</div>
<div class="md-footer__title">
<span class="md-footer__direction">
                Previous
              </span>
<div class="md-ellipsis">
                üèÉüèº‚Äç‚ôÇÔ∏è Quick start
              </div>
</div>
</a>
<a aria-label="Next: üìöTutorial" class="md-footer__link md-footer__link--next" href="../notebooks/tutorial/">
<div class="md-footer__title">
<span class="md-footer__direction">
                Next
              </span>
<div class="md-ellipsis">
                üìöTutorial
              </div>
</div>
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
</div>
</a>
</nav>
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
<div class="md-social">
<a class="md-social__link" href="https://discord.gg/bpxQ4Yxhk5" rel="noopener" target="_blank" title="Discord">
<svg viewbox="0 0 576 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M492.5 69.8c-.2-.3-.4-.6-.8-.7-38.1-17.5-78.4-30-119.7-37.1-.4-.1-.8 0-1.1.1s-.6.4-.8.8c-5.5 9.9-10.5 20.2-14.9 30.6-44.6-6.8-89.9-6.8-134.4 0-4.5-10.5-9.5-20.7-15.1-30.6-.2-.3-.5-.6-.8-.8s-.7-.2-1.1-.2C162.5 39 122.2 51.5 84.1 69c-.3.1-.6.4-.8.7C7.1 183.5-13.8 294.6-3.6 404.2c0 .3.1.5.2.8s.3.4.5.6c44.4 32.9 94 58 146.8 74.2.4.1.8.1 1.1 0s.7-.4.9-.7c11.3-15.4 21.4-31.8 30-48.8.1-.2.2-.5.2-.8s0-.5-.1-.8-.2-.5-.4-.6-.4-.3-.7-.4c-15.8-6.1-31.2-13.4-45.9-21.9-.3-.2-.5-.4-.7-.6s-.3-.6-.3-.9 0-.6.2-.9.3-.5.6-.7c3.1-2.3 6.2-4.7 9.1-7.1.3-.2.6-.4.9-.4s.7 0 1 .1c96.2 43.9 200.4 43.9 295.5 0 .3-.1.7-.2 1-.2s.7.2.9.4c2.9 2.4 6 4.9 9.1 7.2.2.2.4.4.6.7s.2.6.2.9-.1.6-.3.9-.4.5-.6.6c-14.7 8.6-30 15.9-45.9 21.8-.2.1-.5.2-.7.4s-.3.4-.4.7-.1.5-.1.8.1.5.2.8c8.8 17 18.8 33.3 30 48.8.2.3.6.6.9.7s.8.1 1.1 0c52.9-16.2 102.6-41.3 147.1-74.2.2-.2.4-.4.5-.6s.2-.5.2-.8c12.3-126.8-20.5-236.9-86.9-334.5zm-302 267.7c-29 0-52.8-26.6-52.8-59.2s23.4-59.2 52.8-59.2c29.7 0 53.3 26.8 52.8 59.2 0 32.7-23.4 59.2-52.8 59.2m195.4 0c-29 0-52.8-26.6-52.8-59.2s23.4-59.2 52.8-59.2c29.7 0 53.3 26.8 52.8 59.2 0 32.7-23.2 59.2-52.8 59.2"></path></svg>
</a>
<a class="md-social__link" href="mailto:marek.wiewiorka@pw.edu.pl" rel="noopener" target="_blank" title="Email">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m20 8-8 5-8-5V6l8 5 8-5m0-2H4c-1.11 0-2 .89-2 2v12a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2"></path></svg>
</a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"annotate": null, "base": "..", "features": ["content.code.copy", "content.code.select", "content.code.annotate", "content.tabs.link", "content.action.edit", "content.tooltips", "toc.follow", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "navigation.footer", "navigation.tracking", "navigation.instant", "content.footnote.tooltips"], "search": "../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"default": ["latest"], "provider": "mike"}}</script>
<script src="../assets/javascripts/bundle.79ae519e.min.js"></script>
<script src="../assets/_markdown_exec_pyodide.js"></script>
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>