<!DOCTYPE html>
<html class="no-js" lang="en"><head>
<meta charset="utf-8"/>
<meta content="width=device-width,initial-scale=1" name="viewport"/>
<link href="https://biodatageeks.org/polars-bio/blog/category/benchmarks/" rel="canonical"/>
<link href="../../" rel="prev"/>
<link href="../file-formats/" rel="next"/>
<link href="../../../assets/logo.png" rel="icon"/>
<meta content="mkdocs-1.6.1, mkdocs-material-9.7.1" name="generator"/>
<title>benchmarks - polars-bio</title>
<link href="../../../assets/stylesheets/main.484c7ddc.min.css" rel="stylesheet"/>
<link crossorigin="" href="https://fonts.gstatic.com" rel="preconnect"/>
<link href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&amp;display=fallback" rel="stylesheet"/>
<style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
<link href="../../../assets/_markdown_exec_pyodide.css" rel="stylesheet"/>
<link href="../../../assets/_mkdocstrings.css" rel="stylesheet"/>
<script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-1M9PLLGBYS"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-1M9PLLGBYS",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-1M9PLLGBYS",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
<script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
<meta content="website" property="og:type"/>
<meta content="benchmarks - polars-bio" property="og:title"/>
<meta content="https://biodatageeks.org/polars-bio/assets/images/social/blog/category/benchmarks.png" property="og:image"/>
<meta content="image/png" property="og:image:type"/>
<meta content="1200" property="og:image:width"/>
<meta content="630" property="og:image:height"/>
<meta content="https://biodatageeks.org/polars-bio/blog/category/benchmarks/" property="og:url"/>
<meta content="summary_large_image" property="twitter:card"/>
<meta content="benchmarks - polars-bio" property="twitter:title"/>
<meta content="https://biodatageeks.org/polars-bio/assets/images/social/blog/category/benchmarks.png" property="twitter:image"/>
<link href="../../../assets/stylesheets/glightbox.min.css" rel="stylesheet"/><script src="../../../assets/javascripts/glightbox.min.js"></script><style id="glightbox-style">
            html.glightbox-open { overflow: initial; height: 100%; }
            .gslide-title { margin-top: 0px; user-select: text; }
            .gslide-desc { color: #666; user-select: text; }
            .gslide-image img { background: white; }
            .gscrollbar-fixer { padding-right: 15px; }
            .gdesc-inner { font-size: 0.75rem; }
            body[data-md-color-scheme="slate"] .gdesc-inner { background: var(--md-default-bg-color); }
            body[data-md-color-scheme="slate"] .gslide-title { color: var(--md-default-fg-color); }
            body[data-md-color-scheme="slate"] .gslide-desc { color: var(--md-default-fg-color); }
        </style></head>
<body dir="ltr">
<input autocomplete="off" class="md-toggle" data-md-toggle="drawer" id="__drawer" type="checkbox"/>
<input autocomplete="off" class="md-toggle" data-md-toggle="search" id="__search" type="checkbox"/>
<label class="md-overlay" for="__drawer"></label>
<div data-md-component="skip">
<a class="md-skip" href="#benchmarks">
          Skip to content
        </a>
</div>
<div data-md-component="announce">
</div>
<div data-md-color-scheme="default" data-md-component="outdated" hidden="">
</div>
<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
<nav aria-label="Header" class="md-header__inner md-grid">
<a aria-label="polars-bio" class="md-header__button md-logo" data-md-component="logo" href="../../.." title="polars-bio">
<img alt="logo" src="../../../assets/logo.png"/>
</a>
<label class="md-header__button md-icon" for="__drawer">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"></path></svg>
</label>
<div class="md-header__title" data-md-component="header-title">
<div class="md-header__ellipsis">
<div class="md-header__topic">
<span class="md-ellipsis">
            polars-bio
          </span>
</div>
<div class="md-header__topic" data-md-component="header-topic">
<span class="md-ellipsis">
            
              benchmarks
            
          </span>
</div>
</div>
</div>
<script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
<label class="md-header__button md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
</label>
<div class="md-search" data-md-component="search" role="dialog">
<label class="md-search__overlay" for="__search"></label>
<div class="md-search__inner" role="search">
<form class="md-search__form" name="search">
<input aria-label="Search" autocapitalize="off" autocomplete="off" autocorrect="off" class="md-search__input" data-md-component="search-query" name="query" placeholder="Search" required="" spellcheck="false" type="text"/>
<label class="md-search__icon md-icon" for="__search">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"></path></svg>
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</label>
<nav aria-label="Search" class="md-search__options">
<button aria-label="Clear" class="md-search__icon md-icon" tabindex="-1" title="Clear" type="reset">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"></path></svg>
</button>
</nav>
</form>
<div class="md-search__output">
<div class="md-search__scrollwrap" data-md-scrollfix="" tabindex="0">
<div class="md-search-result" data-md-component="search-result">
<div class="md-search-result__meta">
            Initializing search
          </div>
<ol class="md-search-result__list" role="presentation"></ol>
</div>
</div>
</div>
</div>
</div>
<div class="md-header__source">
<a class="md-source" data-md-component="source" href="https://github.com/biodatageeks/polars-bio" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"></path></svg>
</div>
<div class="md-source__repository">
    biodatageeks/polars-bio
  </div>
</a>
</div>
</nav>
<nav aria-label="Tabs" class="md-tabs" data-md-component="tabs">
<div class="md-grid">
<ul class="md-tabs__list">
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../..">
          
  
  
  Home

        </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../quickstart/">
        
  
  
    
  
  üèÉüèº‚Äç‚ôÇÔ∏è Quick start

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../features/">
        
  
  
    
  
  üî®Features

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../notebooks/tutorial/">
        
  
  
    
  
  üìöTutorial

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../api/">
        
  
  
    
  
  ‚öôÔ∏è API reference

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../faq/">
        
  
  
    
  
  ‚ùì FAQ

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../supplement/">
        
  
  
    
  
  üöÄ Performance

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../developers/">
        
  
  
    
  
  üõ†Ô∏è Developers guide

      </a>
</li>
<li class="md-tabs__item">
<a class="md-tabs__link" href="../../../contact/">
        
  
  
    
  
  üì° Contact

      </a>
</li>
<li class="md-tabs__item md-tabs__item--active">
<a class="md-tabs__link" href="../../">
          
  
  
  üìñ Blog

        </a>
</li>
</ul>
</div>
</nav>
</header>
<div class="md-container" data-md-component="container">
<main class="md-main" data-md-component="main">
<div class="md-main__inner md-grid">
<div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Navigation" class="md-nav md-nav--primary md-nav--lifted" data-md-level="0">
<label class="md-nav__title" for="__drawer">
<a aria-label="polars-bio" class="md-nav__button md-logo" data-md-component="logo" href="../../.." title="polars-bio">
<img alt="logo" src="../../../assets/logo.png"/>
</a>
    polars-bio
  </label>
<div class="md-nav__source">
<a class="md-source" data-md-component="source" href="https://github.com/biodatageeks/polars-bio" title="Go to repository">
<div class="md-source__icon md-icon">
<svg viewbox="0 0 512 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"></path></svg>
</div>
<div class="md-source__repository">
    biodatageeks/polars-bio
  </div>
</a>
</div>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--nested">
<input class="md-nav__toggle md-toggle" id="__nav_1" type="checkbox"/>
<label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
<span class="md-ellipsis">
    
  
    Home
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="false" aria-labelledby="__nav_1_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_1">
<span class="md-nav__icon md-icon"></span>
            
  
    Home
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../..">
<span class="md-ellipsis">
    
  
    polars-bio
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../quickstart/">
<span class="md-ellipsis">
    
  
    üèÉüèº‚Äç‚ôÇÔ∏è Quick start
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../features/">
<span class="md-ellipsis">
    
  
    üî®Features
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../notebooks/tutorial/">
<span class="md-ellipsis">
    
  
    üìöTutorial
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../api/">
<span class="md-ellipsis">
    
  
    ‚öôÔ∏è API reference
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../faq/">
<span class="md-ellipsis">
    
  
    ‚ùì FAQ
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../supplement/">
<span class="md-ellipsis">
    
  
    üöÄ Performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../developers/">
<span class="md-ellipsis">
    
  
    üõ†Ô∏è Developers guide
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../contact/">
<span class="md-ellipsis">
    
  
    üì° Contact
  

    
  </span>
</a>
</li>
</ul>
</nav>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../quickstart/">
<span class="md-ellipsis">
    
  
    üèÉüèº‚Äç‚ôÇÔ∏è Quick start
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../features/">
<span class="md-ellipsis">
    
  
    üî®Features
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../notebooks/tutorial/">
<span class="md-ellipsis">
    
  
    üìöTutorial
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../api/">
<span class="md-ellipsis">
    
  
    ‚öôÔ∏è API reference
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../faq/">
<span class="md-ellipsis">
    
  
    ‚ùì FAQ
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../supplement/">
<span class="md-ellipsis">
    
  
    üöÄ Performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../developers/">
<span class="md-ellipsis">
    
  
    üõ†Ô∏è Developers guide
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../../../contact/">
<span class="md-ellipsis">
    
  
    üì° Contact
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_10" type="checkbox"/>
<label class="md-nav__link" for="__nav_10" id="__nav_10_label" tabindex="">
<span class="md-ellipsis">
    
  
    üìñ Blog
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="true" aria-labelledby="__nav_10_label" class="md-nav" data-md-level="1">
<label class="md-nav__title" for="__nav_10">
<span class="md-nav__icon md-icon"></span>
            
  
    üìñ Blog
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item">
<a class="md-nav__link" href="../../">
<span class="md-ellipsis">
    
  
    Blog
  

    
  </span>
</a>
</li>
<li class="md-nav__item md-nav__item--active md-nav__item--nested">
<input checked="" class="md-nav__toggle md-toggle" id="__nav_10_2" type="checkbox"/>
<label class="md-nav__link" for="__nav_10_2" id="__nav_10_2_label" tabindex="0">
<span class="md-ellipsis">
    
  
    Categories
  

    
  </span>
<span class="md-nav__icon md-icon"></span>
</label>
<nav aria-expanded="true" aria-labelledby="__nav_10_2_label" class="md-nav" data-md-level="2">
<label class="md-nav__title" for="__nav_10_2">
<span class="md-nav__icon md-icon"></span>
            
  
    Categories
  

          </label>
<ul class="md-nav__list" data-md-scrollfix="">
<li class="md-nav__item md-nav__item--active">
<input class="md-nav__toggle md-toggle" id="__toc" type="checkbox"/>
<a class="md-nav__link md-nav__link--active" href="./">
<span class="md-ellipsis">
    
  
    benchmarks
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../file-formats/">
<span class="md-ellipsis">
    
  
    file formats
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../performance/">
<span class="md-ellipsis">
    
  
    performance
  

    
  </span>
</a>
</li>
<li class="md-nav__item">
<a class="md-nav__link" href="../releases/">
<span class="md-ellipsis">
    
  
    releases
  

    
  </span>
</a>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</li>
</ul>
</nav>
</div>
</div>
</div>
<div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc">
<div class="md-sidebar__scrollwrap">
<div class="md-sidebar__inner">
<nav aria-label="Table of contents" class="md-nav md-nav--secondary">
</nav>
</div>
</div>
</div>
<div class="md-content" data-md-component="content">
<div class="md-content__inner">
<header class="md-typeset">
<h1 id="benchmarks">benchmarks<a class="headerlink" href="#benchmarks" title="Permanent link"></a></h1>
</header>
<article class="md-post md-post--excerpt">
<header class="md-post__header">
<div class="md-post__meta md-meta">
<ul class="md-meta__list">
<li class="md-meta__item">
<time datetime="2026-02-20 00:00:00+00:00">Friday, February 20, 2026</time></li>
<li class="md-meta__item">
            in
            
              <a class="md-meta__link" href="../performance/">performance</a>, 
              <a class="md-meta__link" href="./">benchmarks</a></li>
<li class="md-meta__item">
            
              4 min read
            
          </li>
</ul>
</div>
</header>
<div class="md-post__content md-typeset">
<h2 id="interval-operations-benchmark-update-february-2026"><a class="toclink" href="../../2026/02/20/interval-operations-benchmark--update-february-2026/">Interval operations benchmark ‚Äî update February 2026</a></h2>
<h3 id="introduction"><a class="toclink" href="../../2026/02/20/interval-operations-benchmark--update-february-2026/#introduction">Introduction</a></h3>
<p>Back in <a href="../../2025/09/05/interval-operations-benchmark--update-september-2025/">September 2025</a> we benchmarked three libraries across three operations. A lot has changed since then. In December 2025, pyranges1 published a <a href="https://www.biorxiv.org/content/10.64898/2025.12.11.693639v1">preprint</a> describing its Rust-powered backend (<a href="https://github.com/pyranges/ruranges">ruranges</a>) and an expanded set of interval operations. On the <a href="https://github.com/biodatageeks/polars-bio">polars-bio</a> side, version 0.24.0 ships a fully rewritten range-operations engine built on upstream DataFusion UDTF providers (OverlapProvider, NearestProvider, and the new coverage/cluster/complement/merge/subtract providers from <a href="https://github.com/biodatageeks/datafusion-bio-functions">datafusion-bio-function-ranges</a>), replacing the earlier sequila-native backend.</p>
<nav class="md-post__action">
<a href="../../2026/02/20/interval-operations-benchmark--update-february-2026/">
          Continue reading
        </a>
</nav>
</div>
</article>
<article class="md-post md-post--excerpt">
<header class="md-post__header">
<div class="md-post__meta md-meta">
<ul class="md-meta__list">
<li class="md-meta__item">
<time datetime="2026-02-14 00:00:00+00:00">Saturday, February 14, 2026</time></li>
<li class="md-meta__item">
            in
            
              <a class="md-meta__link" href="../performance/">performance</a>, 
              <a class="md-meta__link" href="./">benchmarks</a>, 
              <a class="md-meta__link" href="../file-formats/">file formats</a></li>
<li class="md-meta__item">
            
              5 min read
            
          </li>
</ul>
</div>
</header>
<div class="md-post__content md-typeset">
<h2 id="benchmarking-genomic-format-readers-in-python-with-polars"><a class="toclink" href="../../2026/02/14/benchmarking-genomic-format-readers-in-python-with-polars/">Benchmarking Genomic Format Readers in Python with Polars</a></h2>
<p>Genomic analyses in Python typically start with reading BAM, VCF, or FASTQ files into memory. The choice of library for this step can have a dramatic impact on both wall-clock time and memory consumption ‚Äî especially as datasets grow to tens or hundreds of millions of records.</p>
<p><a href="https://github.com/pysam-developers/pysam">pysam</a> has long been the go-to Python library for working with these formats. It provides comprehensive bindings to <a href="https://github.com/samtools/htslib">htslib</a> and is battle-tested across thousands of projects. However, several newer libraries have emerged that leverage <a href="https://arrow.apache.org/">Apache Arrow</a> columnar format and Rust-based parsers to offer potentially better performance.</p>
<p>In this post, we benchmark four Python libraries head-to-head on real-world genomic data to find out which offers the best combination of speed and memory efficiency for reading BAM, VCF, and FASTQ files.</p>
<nav class="md-post__action">
<a href="../../2026/02/14/benchmarking-genomic-format-readers-in-python-with-polars/">
          Continue reading
        </a>
</nav>
</div>
</article>
<article class="md-post md-post--excerpt">
<header class="md-post__header">
<div class="md-post__meta md-meta">
<ul class="md-meta__list">
<li class="md-meta__item">
<time datetime="2025-09-08 00:00:00+00:00">Monday, September 8, 2025</time></li>
<li class="md-meta__item">
            in
            
              <a class="md-meta__link" href="../performance/">performance</a>, 
              <a class="md-meta__link" href="./">benchmarks</a>, 
              <a class="md-meta__link" href="../file-formats/">file formats</a></li>
<li class="md-meta__item">
            
              4 min read
            
          </li>
</ul>
</div>
</header>
<div class="md-post__content md-typeset">
<h2 id="gff-file-reading-performance-enhancements-in-polars-bio-0150"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/">GFF File Reading Performance Enhancements in polars-bio 0.15.0</a></h2>
<p>We're excited to announce significant performance improvements to GFF file reading in polars-bio <code>0.15.0</code>. This release introduces two major optimizations that dramatically improve both speed and memory efficiency when working with GFF files:</p>
<h3 id="key-enhancements"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#key-enhancements">Key Enhancements</a></h3>
<p><strong>Projection Pushdown</strong>: Only the columns you need are read from disk, reducing I/O overhead and memory usage. This is particularly beneficial when working with wide GFF files that contain many optional attributes.</p>
<p><strong>Predicate Pushdown</strong>: Row filtering is applied during the file reading process, eliminating the need to load irrelevant data into memory. This allows for lightning-fast queries on large GFF datasets.</p>
<p><strong>Fully Streamed Parallel Reads</strong>: BGZF-compressed files can now be read in parallel with true streaming, enabling out-of-core processing of massive genomic datasets without memory constraints.</p>
<h3 id="benchmark-methodology"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#benchmark-methodology">Benchmark Methodology</a></h3>
<p>To evaluate these improvements, we conducted comprehensive benchmarks comparing three popular data processing libraries:</p>
<ul>
<li><strong>Pandas</strong>: The traditional Python data analysis library</li>
<li><strong>Polars</strong>: High-performance DataFrame library with lazy evaluation</li>
<li><strong>polars-bio</strong>: Our specialized genomic data processing library built on Polars and Apache DataFusion</li>
</ul>
<p>All benchmarks were performed on a large GFF file (~7.7 million records, <a href="https://drive.google.com/file/d/1PsHqKG-gyRJy5-sNzuH3xRntw4Er--Si/view?usp=sharing">file</a> and <a href="https://drive.google.com/file/d/173RT5Afi2jAh64uCJwNRGHF4ozYU-xzX/view?usp=sharing">index</a> needed for parallel reading) with both full scan and filtered query scenarios to demonstrate real-world performance gains.</p>
<p>For pandas and polars reading, we used the following <a href="https://github.com/biodatageeks/polars-bio/blob/master/benchmarks/gff_parsers.py">methods</a> (thanks to <a href="https://github.com/urineri">@urineri</a> for the Polars code).
Since Polars decompresses compressed CSV/TSV files completely in memory as highlighted <a href="https://github.com/ghuls/polars_streaming_csv_decompression/blob/main/README.md">here</a>, we also used <code>polars_streaming_csv_decompression</code>, a great plugin developed by <a href="https://github.com/ghuls/polars_streaming_csv_decompression">@ghuls</a> to enable streaming decompression in Polars.</p>
<p>Test query used for filtered benchmarks (Polars and polars-bio):</p>
<div class="language-python highlight"><pre><span></span><code><span id="__span-2-1"><a href="#__codelineno-2-1" id="__codelineno-2-1" name="__codelineno-2-1"></a> <span class="n">result</span> <span class="o">=</span> <span class="p">(</span>
</span><span id="__span-2-2"><a href="#__codelineno-2-2" id="__codelineno-2-2" name="__codelineno-2-2"></a>        <span class="n">lf</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
</span><span id="__span-2-3"><a href="#__codelineno-2-3" id="__codelineno-2-3" name="__codelineno-2-3"></a>            <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"seqid"</span><span class="p">)</span> <span class="o">==</span> <span class="s2">"chrY"</span><span class="p">)</span>
</span><span id="__span-2-4"><a href="#__codelineno-2-4" id="__codelineno-2-4" name="__codelineno-2-4"></a>            <span class="o">&amp;</span> <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"start"</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">500000</span><span class="p">)</span>
</span><span id="__span-2-5"><a href="#__codelineno-2-5" id="__codelineno-2-5" name="__codelineno-2-5"></a>            <span class="o">&amp;</span> <span class="p">(</span><span class="n">pl</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">"end"</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">510000</span><span class="p">)</span>
</span><span id="__span-2-6"><a href="#__codelineno-2-6" id="__codelineno-2-6" name="__codelineno-2-6"></a>        <span class="p">)</span>
</span><span id="__span-2-7"><a href="#__codelineno-2-7" id="__codelineno-2-7" name="__codelineno-2-7"></a>        <span class="o">.</span><span class="n">select</span><span class="p">([</span><span class="s2">"seqid"</span><span class="p">,</span> <span class="s2">"start"</span><span class="p">,</span> <span class="s2">"end"</span><span class="p">,</span> <span class="s2">"type"</span><span class="p">])</span>
</span><span id="__span-2-8"><a href="#__codelineno-2-8" id="__codelineno-2-8" name="__codelineno-2-8"></a>        <span class="o">.</span><span class="n">collect</span><span class="p">()</span>
</span><span id="__span-2-9"><a href="#__codelineno-2-9" id="__codelineno-2-9" name="__codelineno-2-9"></a>    <span class="p">)</span>
</span></code></pre></div>
<p>The above query is very selective and returns only two rows from the entire dataset.</p>
<h3 id="results"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#results">Results</a></h3>
<p>Complete benchmark code and results are available in the <a href="https://github.com/biodatageeks/polars-bio/tree/master/benchmarks">polars-bio</a> repository.</p>
<h4 id="single-threaded-performance"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#single-threaded-performance">Single-threaded performance</a></h4>
<p><img alt="general_performance.png" src="../../figures/gff-read-optimization-2025-09/general_performance.png"/></p>
<p><strong>Key takeaways:</strong></p>
<ul>
<li><strong>polars-bio</strong> delivers comparable performance to standard <strong>Polars</strong> for full scan operations and both significantly outperform Pandas.</li>
<li>In the case of <strong>filtered queries</strong>, we can see further performance improvements with <strong>Polars</strong> and <strong>polars-bio</strong> thanks to predicate and projection pushdown optimizations. <strong>polars-bio</strong> is 2.5x faster than standard Polars.</li>
</ul>
<h4 id="memory-usage"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#memory-usage">Memory usage</a></h4>
<p><img alt="memory_comparison.png" src="../../figures/gff-read-optimization-2025-09/memory_comparison.png"/></p>
<p><strong>Key takeaways:</strong></p>
<ul>
<li><strong>Polars</strong> and <strong>polars-bio</strong> use significantly less memory than <strong>Pandas</strong> for all operations.</li>
<li><strong>polars-bio</strong> and Polars with <code>polars_streaming_csv_decompression</code> can use more than 20x less memory than vanilla Polars and more than <strong>two orders of magnitude</strong> less memory than Pandas for operations involving filtering.</li>
</ul>
<h4 id="thread-scalability"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#thread-scalability">Thread scalability</a></h4>
<p><img alt="thread_scalability.png" src="../../figures/gff-read-optimization-2025-09/thread_scalability.png"/></p>
<p><strong>Key takeaways:</strong></p>
<ul>
<li><strong>polars-bio achieves near-linear scaling</strong> up to 8 threads for full scan operations, reaching 9.5x speedup at 16 threads compared to single-threaded performance.</li>
<li><strong>Filtered operations show excellent parallelization</strong> with polars-bio reaching 11x speedup at 16 threads, significantly outperforming other libraries. There is, however, non-negligible overhead due to parallelism at 1 thread (2.25s vs 4.2s, compared to the single-threaded benchmark).</li>
<li><strong>polars-streaming shows diminishing returns</strong> at higher thread counts due to the overhead of spawning decompression program threads (in the <a href="https://github.com/pycompression/xopen/blob/main/src/xopen/__init__.py#L182">default</a> configuration, this is capped at 4), while polars-bio maintains consistent scaling benefits.</li>
</ul>
<h3 id="summary"><a class="toclink" href="../../2025/09/08/gff-file-reading-performance-enhancements-in-polars-bio-0150/#summary">Summary</a></h3>
<p>The benchmarks demonstrate that polars-bio <code>0.15.0</code> delivers significant performance improvements for GFF file processing. These optimizations, combined with near-linear thread scaling and fully streamed parallel reads, make polars-bio an ideal choice for high-performance genomic data analysis workflows.</p>
<p>If you haven't tried polars-bio yet, now is a great time to explore its capabilities for efficient genomic data processing with Python! <a href="https://docs.google.com/forms/d/e/1FAIpQLSdxE3admdmAPE_PVSsUvjJAPXuDAKN4VjHLWJWmXGNVqX6mzw/viewform?usp=dialog">Join</a> our upcoming seminar on <em>September 15, 2025</em>, to learn more about polars-bio and its applications in genomics.
<img alt="seminar.png" src="../../figures/gff-read-optimization-2025-09/seminar.png"/></p>
</div>
</article>
<article class="md-post md-post--excerpt">
<header class="md-post__header">
<div class="md-post__meta md-meta">
<ul class="md-meta__list">
<li class="md-meta__item">
<time datetime="2025-09-05 00:00:00+00:00">Friday, September 5, 2025</time></li>
<li class="md-meta__item">
            in
            
              <a class="md-meta__link" href="../performance/">performance</a>, 
              <a class="md-meta__link" href="./">benchmarks</a></li>
<li class="md-meta__item">
            
              5 min read
            
          </li>
</ul>
</div>
</header>
<div class="md-post__content md-typeset">
<h2 id="interval-operations-benchmark-update-september-2025"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/">Interval operations benchmark ‚Äî update September 2025</a></h2>
<h3 id="introduction"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#introduction">Introduction</a></h3>
<p>Benchmarking isn‚Äôt a one-and-done exercise‚Äîit‚Äôs a moving target. As tools evolve, new versions can shift performance profiles in meaningful ways, so keeping results current is just as important as the first round of measurements.</p>
<p>Recently, three novel libraries that have started to gain traction: <a href="https://github.com/pyranges/pyranges_1.x">pyranges1</a>, <a href="https://github.com/BiocPy/GenomicRanges">GenomicRanges</a> and <a href="https://github.com/biodatageeks/polars-bio">polars-bio</a>
<img alt="star-history-202595.png" src="../../figures/benchmark-sep-2025/star-history-202595.png"/></p>
<p>shipped major updates:</p>
<ul>
<li><a href="https://github.com/pyranges/pyranges_1.x">pyranges1</a> adopted a new Rust backend (<a href="https://github.com/pyranges/ruranges">ruranges</a>),</li>
<li><a href="https://github.com/BiocPy/GenomicRanges">GenomicRanges</a> switched its interval core to a Nested Containment List (<a href="https://github.com/pyranges/ncls">NCLS</a>) and added multithreaded execution,</li>
<li>polars-bio migrated to the new Polars streaming engine and added support for new interval data structures. As of version <code>0.12.0</code> it supports:<ul>
<li><a href="https://github.com/dcjones/coitrees">COITrees</a></li>
<li><a href="https://github.com/rust-bio/rust-bio/blob/master/src/data_structures/interval_tree/array_backed_interval_tree.rs">IITree</a></li>
<li><a href="https://github.com/rust-bio/rust-bio/blob/master/src/data_structures/interval_tree/avl_interval_tree.rs">AVL-tree</a></li>
<li><a href="https://github.com/sstadick/rust-lapper">rust-lapper</a></li>
<li><a href="https://github.com/kcleal/superintervals/">superintervals</a></li>
</ul>
</li>
</ul>
<p>Each of these changes has the potential to meaningfully alter performance and memory characteristics for common genomic interval tasks.</p>
<p>In this post, we revisit our benchmarks with those releases in mind. We focus on three everyday operations:</p>
<ul>
<li>overlap detection,</li>
<li>nearest feature queries</li>
<li>overlap counting.</li>
</ul>
<p>For comparability, we use the same <a href="/polars-bio/supplement/#real-dataset">AIList</a> dataset from our previous write-up, so you can see exactly how the new backends and data structures change the picture. Let‚Äôs dive in and see what‚Äôs faster, what‚Äôs leaner, and where the trade-offs now live.</p>
<h3 id="setup"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#setup">Setup</a></h3>
<h4 id="benchmark-test-cases"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#benchmark-test-cases">Benchmark test cases</a></h4>
<table>
<thead>
<tr>
<th>Dataset pairs</th>
<th>Size</th>
<th># of overlaps (1-based)</th>
</tr>
</thead>
<tbody>
<tr>
<td>1-2 &amp; 2-1</td>
<td>Small</td>
<td>54,246</td>
</tr>
<tr>
<td>7-3 &amp; 3-7</td>
<td>Medium</td>
<td>4,408,383</td>
</tr>
<tr>
<td>8-7 &amp; 7-8</td>
<td>Large</td>
<td>307,184,634</td>
</tr>
</tbody>
</table>
<h4 id="software-versions"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#software-versions">Software versions</a></h4>
<table>
<thead>
<tr>
<th>Library</th>
<th>Version</th>
</tr>
</thead>
<tbody>
<tr>
<td>polars_bio</td>
<td>0.13.1</td>
</tr>
<tr>
<td>pyranges</td>
<td>0.1.14</td>
</tr>
<tr>
<td>genomicranges</td>
<td>0.7.2</td>
</tr>
</tbody>
</table>
<h3 id="results"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#results">Results</a></h3>
<h4 id="polars-bio-interval-data-structures-performance-comparison"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#polars-bio-interval-data-structures-performance-comparison">polars-bio interval data structures performance comparison</a></h4>
<p><img alt="combined_multi_testcase.png" src="../../figures/benchmark-sep-2025/combined_multi_testcase.png"/></p>
<p>Key takeaways:</p>
<ul>
<li><strong>Superintervals</strong> seems to be the best default. Across all three test cases, it is consistently the fastest or tied for fastest, delivering 1.25‚Äì1.44x speedups over the <strong>polars-bio default (COITrees)</strong> and avoiding worst‚Äëcase behavior.</li>
<li>Lapper caveat: performs well on 1‚Äë2 and 8‚Äë7, but collapses on 7‚Äë3 (‚âà25x slower than default), so it‚Äôs risky as a general‚Äëpurpose algorithm.</li>
<li>Intervaltree/Arrayintervaltree: reliable but slower. They trail superintervals by 20‚Äì70% depending on the case.</li>
</ul>
<h4 id="all-operations-comparison"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#all-operations-comparison">All operations comparison</a></h4>
<p><img alt="all_operations_walltime_comparison.png" src="../../figures/benchmark-sep-2025/all_operations_walltime_comparison.png"/></p>
<p><img alt="bench-20250-all_operations_speedup_comparison.png" src="../../figures/benchmark-sep-2025/bench-20250-all_operations_speedup_comparison.png"/></p>
<p>Key takeaways:</p>
<ul>
<li><em>Overlap</em>: <strong>GenomicRanges</strong> wins on small inputs (1‚Äë2, 2‚Äë1) by ~2.1‚Äì2.3x, but polars‚Äëbio takes over from medium size onward and dominates on large (7‚Äë8, 8‚Äë7), where PyRanges falls far behind. Interesting case of <em>7-8</em> vs <em>8-7</em> when swapping inputs can significantly affect performance of GenomicRanges.</li>
<li><em>Nearest</em>: <strong>polars‚Äëbio</strong> leads decisively at every size; speedups over the others grow with input size (orders of magnitude on large datasets).</li>
<li><em>Count overlaps</em>: <strong>GenomicRanges</strong> edges out polars‚Äëbio on the smallest inputs, while <strong>polars‚Äëbio</strong> is faster on medium and substantially faster on large inputs.</li>
</ul>
<h4 id="all-operations-parallel-execution"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#all-operations-parallel-execution">All operations parallel execution</a></h4>
<p><img alt="bench_parallel_speedup_combined_8-7.png" src="../../figures/benchmark-sep-2025/bench_parallel_speedup_combined_8-7.png"/></p>
<p><img alt="benchmark_comparison_genomicranges_vs_polars_bio.png" class="glightbox" src="../../figures/benchmark-sep-2025/benchmark_comparison_genomicranges_vs_polars_bio.png"/></p>
<p><img alt="benchmark_speedup_comparison_genomicranges_vs_polars_bio.png" class="glightbox" src="../../figures/benchmark-sep-2025/benchmark_speedup_comparison_genomicranges_vs_polars_bio.png"/></p>
<p>Key takeaways:</p>
<ul>
<li>Thread scaling: <strong>both</strong> libraries (GenomicRanges and polars-bio) benefit from additional threads, but the absolute gap favors <strong>polars‚Äëbio</strong> for medium/large datasets across overlap, nearest, and count overlaps.</li>
<li>Small overlaps: <strong>GenomicRanges</strong> remains &gt;2x faster at 1‚Äì8 threads; on medium/large pairs its relative speed drops below 1x.</li>
<li>Nearest: <strong>polars‚Äëbio</strong> stays on the 1x reference line; <strong>GenomicRanges</strong> is typically 10‚Äì100x slower (log scale) even with more threads.</li>
<li>Count overlaps: small inputs slightly favor <strong>GenomicRanges</strong>; for larger inputs <strong>polars‚Äëbio</strong> maintains 2‚Äì10x advantage with stable scaling.</li>
</ul>
<h4 id="end-to-end-data-proecesing"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#end-to-end-data-proecesing">End to-end data proecesing</a></h4>
<p>Here we compare end-to-end performance including data loading, overlap operation, and saving results to CSV.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<ol>
<li><code>POLARS_MAX_THREADS=1</code> was set to ensure fair comparison with single-threaded PyRanges.</li>
<li>Since GenomicRanges supports Polars DataFrames as input and output, we used them instead of Pandas to again ensure fair comparison with polars-bio.</li>
<li>GenomicRanges <a href="https://biocpy.github.io/GenomicRanges/api/genomicranges.html#genomicranges.GenomicRanges.GenomicRanges.find_overlaps">find_overlaps</a> method returns hits-only table (indices of genomic intervals instead of genomic coordinates), we also benchmarked an extended version with additional lookup of intervals (<code>full rows</code>, <a href="https://github.com/biodatageeks/polars-bio-bench/blob/master/src/utils.py#L99">code</a>) for fair comparison.</li>
</ol>
</div>
<p><img alt="combined_benchmark_visualization.png" class="glightbox" src="../../figures/benchmark-sep-2025/combined_benchmark_visualization.png"/></p>
<p>Key takeaways:</p>
<ul>
<li>Wall time: <strong>GenomicRanges (hits‚Äëonly)</strong> is the fastest end‚Äëto‚Äëend here (~1.16x vs polars_bio) by avoiding full materialization of genomic intervals (unlike PyRanges and polars-bio that return pairs of genomic interval coordinates for each overlap); <strong>PyRanges</strong> is far slower; <strong>GenomicRanges</strong> (full rows, so with the output comparable with PyRanges and polars-bio) is much slower.</li>
<li>Memory: <strong>polars-bio (streaming)</strong> minimizes peak RAM (~0.7 GB) while keeping speed comparable to <strong>polars-bio</strong>. <strong>GenomicRanges</strong> (full rows) peaks at ~40 GB; hits‚Äëonly sits in the middle (~8.2 GB) as it only returns DataFrame with pairs of indices not full genomic coordinates.</li>
</ul>
<h3 id="summary"><a class="toclink" href="../../2025/09/05/interval-operations-benchmark--update-september-2025/#summary">Summary</a></h3>
<p>For small and medium datasets, all tools perform well; at large scale, <strong>polars-bio</strong> excels with better scalability and memory efficiency, achieving an ultra‚Äëlow footprint in streaming mode.</p>
</div>
</article>
<nav class="md-pagination">
</nav>
</div>
</div>
<script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
</div>
<button class="md-top md-icon" data-md-component="top" hidden="" type="button">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"></path></svg>
  Back to top
</button>
</main>
<footer class="md-footer">
<nav aria-label="Footer" class="md-footer__inner md-grid">
<a aria-label="Previous: Blog" class="md-footer__link md-footer__link--prev" href="../../">
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"></path></svg>
</div>
<div class="md-footer__title">
<span class="md-footer__direction">
                Previous
              </span>
<div class="md-ellipsis">
                Blog
              </div>
</div>
</a>
<a aria-label="Next: file formats" class="md-footer__link md-footer__link--next" href="../file-formats/">
<div class="md-footer__title">
<span class="md-footer__direction">
                Next
              </span>
<div class="md-ellipsis">
                file formats
              </div>
</div>
<div class="md-footer__button md-icon">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"></path></svg>
</div>
</a>
</nav>
<div class="md-footer-meta md-typeset">
<div class="md-footer-meta__inner md-grid">
<div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" rel="noopener" target="_blank">
      Material for MkDocs
    </a>
</div>
<div class="md-social">
<a class="md-social__link" href="https://discord.gg/bpxQ4Yxhk5" rel="noopener" target="_blank" title="Discord">
<svg viewbox="0 0 576 512" xmlns="http://www.w3.org/2000/svg"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M492.5 69.8c-.2-.3-.4-.6-.8-.7-38.1-17.5-78.4-30-119.7-37.1-.4-.1-.8 0-1.1.1s-.6.4-.8.8c-5.5 9.9-10.5 20.2-14.9 30.6-44.6-6.8-89.9-6.8-134.4 0-4.5-10.5-9.5-20.7-15.1-30.6-.2-.3-.5-.6-.8-.8s-.7-.2-1.1-.2C162.5 39 122.2 51.5 84.1 69c-.3.1-.6.4-.8.7C7.1 183.5-13.8 294.6-3.6 404.2c0 .3.1.5.2.8s.3.4.5.6c44.4 32.9 94 58 146.8 74.2.4.1.8.1 1.1 0s.7-.4.9-.7c11.3-15.4 21.4-31.8 30-48.8.1-.2.2-.5.2-.8s0-.5-.1-.8-.2-.5-.4-.6-.4-.3-.7-.4c-15.8-6.1-31.2-13.4-45.9-21.9-.3-.2-.5-.4-.7-.6s-.3-.6-.3-.9 0-.6.2-.9.3-.5.6-.7c3.1-2.3 6.2-4.7 9.1-7.1.3-.2.6-.4.9-.4s.7 0 1 .1c96.2 43.9 200.4 43.9 295.5 0 .3-.1.7-.2 1-.2s.7.2.9.4c2.9 2.4 6 4.9 9.1 7.2.2.2.4.4.6.7s.2.6.2.9-.1.6-.3.9-.4.5-.6.6c-14.7 8.6-30 15.9-45.9 21.8-.2.1-.5.2-.7.4s-.3.4-.4.7-.1.5-.1.8.1.5.2.8c8.8 17 18.8 33.3 30 48.8.2.3.6.6.9.7s.8.1 1.1 0c52.9-16.2 102.6-41.3 147.1-74.2.2-.2.4-.4.5-.6s.2-.5.2-.8c12.3-126.8-20.5-236.9-86.9-334.5zm-302 267.7c-29 0-52.8-26.6-52.8-59.2s23.4-59.2 52.8-59.2c29.7 0 53.3 26.8 52.8 59.2 0 32.7-23.4 59.2-52.8 59.2m195.4 0c-29 0-52.8-26.6-52.8-59.2s23.4-59.2 52.8-59.2c29.7 0 53.3 26.8 52.8 59.2 0 32.7-23.2 59.2-52.8 59.2"></path></svg>
</a>
<a class="md-social__link" href="mailto:marek.wiewiorka@pw.edu.pl" rel="noopener" target="_blank" title="Email">
<svg viewbox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m20 8-8 5-8-5V6l8 5 8-5m0-2H4c-1.11 0-2 .89-2 2v12a2 2 0 0 0 2 2h16a2 2 0 0 0 2-2V6a2 2 0 0 0-2-2"></path></svg>
</a>
</div>
</div>
</div>
</footer>
</div>
<div class="md-dialog" data-md-component="dialog">
<div class="md-dialog__inner md-typeset"></div>
</div>
<script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": ["content.code.copy", "content.code.select", "content.code.annotate", "content.tabs.link", "content.action.edit", "content.tooltips", "toc.follow", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "navigation.footer", "navigation.tracking", "navigation.instant", "content.footnote.tooltips"], "search": "../../../assets/javascripts/workers/search.2c215733.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"default": ["latest"], "provider": "mike"}}</script>
<script src="../../../assets/javascripts/bundle.79ae519e.min.js"></script>
<script src="../../../assets/_markdown_exec_pyodide.js"></script>
<script id="init-glightbox">const lightbox = GLightbox({"touchNavigation": true, "loop": false, "zoomable": true, "draggable": true, "openEffect": "zoom", "closeEffect": "zoom", "slideEffect": "slide"});
document$.subscribe(()=>{ lightbox.reload(); });
</script></body></html>